// orders/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const editOrder = require('./editOrder');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  iptCheckerClientId: 'messenger/client',
  importPath: './',
  zintExe: 'zint',
  syncPsStatus: {
    generator: false,
    updater: false
  },
  syncDrillStatus: {
    generator: false,
    updater: false
  }
};

exports.models = [
  require('./models/changeNumber'),
  require('./models/component'),
  require('./models/emptyOrder'),
  require('./models/invalidOrder'),
  require('./models/mechOrder'),
  require('./models/order'),
  require('./models/orderEto'),
  require('./models/orderIntake'),
  require('./models/orderZlf1'),
  require('./models/pkhdComponent'),
  require('./models/pkhdStrategy'),
  require('./models/productNote'),
  require('./models/productTag'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('paintShop/models/paintShopOrder'),
  require('delayReasons/models/delayReason'),
  require('wmes-drilling/models/drillingOrder'),
  require('wmes-compRel/models/compRelEntry')
];

exports.recordTopics = {
  info: ['mechOrders.synced']
};

exports.republishTopics = [
  'orders.updated.*',
  'orders.quantityDone.*',
  'orders.invalid.**',
  'orders.productNotes.*',
  'orders.productTags.*'
];

exports.userPrivileges = [
  'ORDERS:VIEW',
  'ORDERS:MANAGE',
  'ORDERS:COMMENT'
];

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'settings': require('./operationGroups'),
  'mongoose': [
    require('./psStatus'),
    require('./drillStatus'),
    require('./compRel')
  ],
  'mongoose user express settings': require('./routes')
};

exports.start = (app, module) =>
{
  module.editOrder = editOrder.bind(null, app, module);
};
