// orders/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const editOrder = require('./editOrder');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  iptCheckerClientId: 'messenger/client',
  importPath: './',
  zintExe: 'zint'
};

exports.models = [
  require('./models/component'),
  require('./models/emptyOrder'),
  require('./models/invalidOrder'),
  require('./models/mechOrder'),
  require('./models/order'),
  require('./models/orderEto'),
  require('./models/orderIntake'),
  require('./models/orderZlf1'),
  require('./models/pkhdComponent'),
  require('./models/pkhdStrategy'),
  require('./models/productNote'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('paintShop/models/paintShopOrder'),
  require('delayReasons/models/delayReason')
];

exports.recordTopics = {
  info: ['mechOrders.synced']
};

exports.republishTopics = [
  'orders.updated.*',
  'orders.quantityDone.*',
  'orders.invalid.**',
  'orders.productMessages.*'
];

exports.userPrivileges = ['ORDERS:VIEW', 'ORDERS:MANAGE', 'ORDERS:COMMENT'];

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'settings': require('./operationGroups'),
  'mongoose': require('./paintShop'),
  'mongoose user express settings': require('./routes')
};

exports.start = (app, module) =>
{
  module.editOrder = editOrder.bind(null, app, module);
};
