// orderBomMatchers/models/orderBomMatcher.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'OrderBomMatcher';

exports.setUp = (app, mongoose) =>
{
  const orderBomMatcherSchema = new mongoose.Schema({
    active: {
      type: Boolean,
      default: true
    },
    description: {
      type: String,
      required: true,
      trim: true
    },
    matchers: {
      mrp: [String],
      nc12: [String],
      name: [String],
      line: [String]
    },
    components: [{
      pattern: String,
      notPattern: String,
      description: String,
      optional: {
        type: Boolean,
        default: false
      },
      missing: {
        type: Boolean,
        default: false
      },
      unique: {
        type: Boolean,
        default: true
      },
      first: {
        type: Boolean,
        default: false
      },
      single: {
        type: Boolean,
        default: true
      },
      labelPattern: {
        type: String,
        required: true,
        trim: true
      },
      nc12Index: [Number],
      snIndex: [Number]
    }]
  }, {
    id: false
  });

  orderBomMatcherSchema.statics.TOPIC_PREFIX = 'orderBomMatchers';

  mongoose.model(exports.name, orderBomMatcherSchema);
};
