// planning/models/planChange.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'PlanChange';

exports.setUp = (app, mongoose) =>
{
  const planChangeSchema = new mongoose.Schema({
    plan: Date,
    date: Date,
    user: {},
    data: {}
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  planChangeSchema.statics.TOPIC_PREFIX = 'planning.changes';
  planChangeSchema.statics.BROWSE_LIMIT = 1000;

  planChangeSchema.index({plan: 1});

  planChangeSchema.methods.toCreatedMessage = function(plan, isNew, message)
  {
    if (!message)
    {
      message = this.toJSON();
    }

    message.isNew = isNew;

    (message.data.changedLines || []).forEach(changedLine =>
    {
      let planLine = plan.lines.find(line => line._id === changedLine._id);

      if (planLine.toObject)
      {
        planLine = planLine.toObject();
      }

      Object.assign(changedLine, planLine);
    });

    return message;
  };

  mongoose.model(exports.name, planChangeSchema);
};
