// wmes-luca-frontend/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  messengerClientId: 'messenger/client',
  productionId: 'production'
};

exports.models = [
  require('./models/lucaEvent')
];

exports.republishTopics = [
  'luca.events.saved'
];

exports.userPrivileges = [
  'LUCA:VIEW',
  'LUCA:MANAGE'
];

exports.optionalModules = {
  'user production messengerClient settings express': require('./routes')
};

exports.onModuleSetUp = (app, {module, setUpModule}) =>
{
  if (setUpModule.name === module.config.expressId)
  {
    setUpModule.config.noSessionPatterns.push(req => /^LUCA/i.test(req.headers['user-agent']));
  }
};

exports.start = (app) =>
{
  app.broker.subscribe('production.synced.*', onProductionSynced);
  app.broker.subscribe('kanban.entries.updated', onKanbanEntryUpdated);

  function onProductionSynced(changes)
  {
    if (changes.types.includes('changeOrder') || changes.types.includes('correctOrder'))
    {
      app.broker.publish('luca.orderChanged', {
        lineId: changes.prodLine
      });
    }
  }

  function onKanbanEntryUpdated(change)
  {
    if (change.property === 'locations' || change.property === 'workCenter')
    {
      app.broker.publish('luca.kanbanChanged', change);
    }
  }
};
