// qi/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  httpServerId: 'httpServer',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  reportsId: 'reports',
  orgUnitsId: 'orgUnits',
  mailSenderId: 'mail/sender',
  morId: 'mor',
  wkhtmltopdfExe: 'wkhtmltopdf',
  attachmentsDest: null
};

exports.models = [
  require('./models/qiActionStatus'),
  require('./models/qiErrorCategory'),
  require('./models/qiFault'),
  require('./models/qiKind'),
  require('./models/qiResult'),
  require('orders/models/order'),
  require('prodShiftOrders/models/prodShiftOrder')
];

exports.optionalModules = {
  'mongoose': require('./counter'),
  'mongoose orgUnits mailSender mor': require('./notifier'),
  'mongoose express user settings reports orgUnits': require('./routes')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    kinds: 'QiKind',
    faults: 'QiFault',
    errorCategories: 'QiErrorCategory',
    actionStatuses: 'QiActionStatus'
  };

  module.tmpAttachments = {};
};
