// kanban/routes/importSap.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');

module.exports = function importSapRoute(app, module, req, res, next)
{
  const sapGui = app[module.config.sapGuiId];

  if (!sapGui)
  {
    return next(app.createError('Import feature is disabled.', 'DISABLED', 500));
  }

  if (module.importTimer)
  {
    return next(app.createError('Import in progress.', 'IN_PROGRESS', 400));
  }

  const {settings, user, KanbanSupplyArea} = module;

  module.importTimer = setTimeout(() => module.importTimer = null, 15 * 60 * 1000);

  step(
    function()
    {
      KanbanSupplyArea.find({}, {name: 1}).lean().exec(this.parallel());

      settings.findValues('kanban.', this.parallel());
    },
    function(err, supplyAreas, kanbanSettings)
    {
      if (err)
      {
        return next(err);
      }

      const kanbanJob = {
        name: 'kanban',
        scriptTimeout: 10 * 60 * 1000,
        repeatOnFailure: 0,
        supplyAreas: _.uniq(supplyAreas.map(d => d.name)),
        pkhdStorageType: kanbanSettings['import.pkhdStorageType'] || '151',
        maktLanguage: kanbanSettings['import.maktLanguage'] || 'PL',
        mlgtWarehouseNo: kanbanSettings['import.mlgtWarehouseNo'] || 'KZ1',
        mlgtStorageType: kanbanSettings['import.mlgtStorageType'] || '851',
        waitForResult: false
      };

      sapGui.runRemoteJob(kanbanJob, err =>
      {
        if (err)
        {
          clearTimeout(module.importTimer);
          module.importTimer = null;

          return next(err);
        }

        app.broker.publish('kanban.import.started');

        module.info(`SAP import started.`, {
          user: user.createUserInfo(req.session.user, req)
        });

        res.sendStatus(204);
      });
    }
  );
};
