// wmes-osh/routes/accidents/add.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

module.exports = async (app, module, req, res, next) =>
{
  const {
    OshAccident
  } = module;

  const input = _.pick(req.body, [
    'division',
    'workplace',
    'department',
    'building',
    'location',
    'station',
    'eventDate',
    'subject',
    'description',
    'attachments'
  ]);

  if (input.attachments && Array.isArray(input.attachments.added))
  {
    input.attachments = module.attachments.claimPendingUploads(input.attachments.added);
  }
  else
  {
    input.attachments = [];
  }

  try
  {
    const creator = module.createUserInfo(req.session.user, req);
    const createdAt = new Date();

    const accident = new OshAccident({
      ...input,
      status: 'finished',
      createdAt,
      updatedAt: createdAt,
      finishedAt: createdAt,
      duration: 0,
      creator,
      updater: creator,
      finisher: creator,
      coordinators: []
    });

    await OshAccident.assignNextRid(accident);

    accident.updateUsers([], accident.creator);

    await accident.save();

    app.broker.publish(`${OshAccident.TOPIC_PREFIX}.added`, {
      user: accident.creator,
      model: accident
    });

    res.json(accident);
  }
  catch (err)
  {
    next(err);
  }
};
