// wmes-osh/routes/accidents/export.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');

module.exports = async (app, module, req, res, next) =>
{
  const {
    express,
    nameCache,
    OshAccident
  } = module;

  const names = await nameCache.load([
    'divisions',
    'workplaces',
    'departments',
    'buildings',
    'locations',
    'stations'
  ]);

  const columns = {
    rid: 13,
    createdAt: 'datetime',
    creator: 20,
    coordinators: 20,
    division: 20,
    workplace: 20,
    department: 20,
    building: 20,
    location: 20,
    station: 20,
    eventDate: 'datetime+utc',
    subject: 40,
    description: 40,
    comments: 60
  };

  req.rql.fields = {
    changes: 1
  };

  Object.keys(columns).forEach(field => req.rql.fields[field] = 1);

  express.crud.exportRoute(app, {
    filename: 'WMES_OSH_ACCIDENTS',
    freezeRows: 1,
    freezeColumns: 1,
    columns,
    model: OshAccident,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    const comments = [];

    doc.changes.forEach(change =>
    {
      if (!change.comment)
      {
        return;
      }

      const time = moment(change.date).format('YYYY-MM-DD HH:mm:ss');
      const user = change.user.label;
      const text = change.comment.replace(/\r\n/g, '\n').replace('\n', '\r\n');

      comments.push(`[${time}] ${user}: ${text}`);
    });

    return {
      rid: doc.rid,
      status: doc.status,
      createdAt: doc.createdAt,
      creator: doc.creator.label,
      coordinators: doc.coordinators.map(u => u.label).join('; '),
      division: names.single('divisions', doc.division),
      workplace: names.single('workplaces', doc.division),
      department: names.single('departments', doc.department),
      building: names.single('buildings', doc.building),
      location: names.single('locations', doc.location),
      station: names.single('stations', doc.station),
      eventDate: doc.eventDate,
      subject: doc.subject,
      description: doc.description,
      comments: comments.join('\r\n')
    };
  }
};
