// production/logEntryHandlers/addDowntime.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const util = require('./util');

module.exports = function(app, productionModule, prodLine, logEntry, done)
{
  const mongoose = app[productionModule.config.mongooseId];
  const ProdDowntime = mongoose.model('ProdDowntime');

  const prodDowntime = new ProdDowntime(logEntry.data);

  step(
    function findProdShiftOrderStep()
    {
      if (prodDowntime.prodShiftOrder)
      {
        productionModule.getProdData('order', prodDowntime.prodShiftOrder, this.parallel());
      }
    },
    function saveProdDowntimeStep(err, prodShiftOrder)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to find order for a new downtime.', {
          prodDowntimeId: logEntry.data._id,
          prodLogEntryId: logEntry._id
        });

        return this.done(done, err);
      }

      this.prodShiftOrder = prodShiftOrder;

      prodDowntime.workerCount = prodShiftOrder ? prodShiftOrder.workerCount : 1;
      prodDowntime.orderData = ProdDowntime.getOrderData(prodShiftOrder);
      prodDowntime.save(this.next());
    },
    function recalcOrderDurationsStep(err)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to save a new downtime.', {
          prodDowntimeId: logEntry.data._id,
          prodLogEntryId: logEntry._id
        });

        return this.done(done, err);
      }

      if (this.prodShiftOrder)
      {
        this.prodShiftOrder.recalcDurations(true, this.next());
      }
    },
    util.createRecalcShiftTimesStep(productionModule, logEntry),
    function finalizeStep(err)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to recalc durations for order after adding a downtime.', {
          prodShiftOrderId: prodDowntime.prodShiftOrder,
          prodDowntimeId: prodDowntime._id,
          prodLogEntryId: logEntry._id
        });
      }

      this.prodShiftOrder = null;

      return done(err);
    }
  );
};
