// production/logEntryHandlers/changeOrder.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const util = require('./util');

module.exports = function(app, productionModule, prodLine, logEntry, done)
{
  const mongoose = app[productionModule.config.mongooseId];
  const ProdShiftOrder = mongoose.model('ProdShiftOrder');

  if (logEntry.data.master === undefined)
  {
    copyPersonnelData();
  }
  else
  {
    handleOfflineEntry();
  }

  function copyPersonnelData()
  {
    productionModule.getProdData('shift', logEntry.prodShift, function(err, prodShift)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to find prod shift to copy personnel data.', {
          prodShiftId: logEntry.prodShift,
          prodLogEntryId: logEntry._id
        });
      }
      else if (prodShift)
      {
        logEntry.data.master = prodShift.master;
        logEntry.data.leader = prodShift.leader;
        logEntry.data.operator = prodShift.operator;
        logEntry.data.operators = prodShift.operators;
      }

      return handleOfflineEntry();
    });
  }

  function handleOfflineEntry()
  {
    if (util.isOfflineEntry(logEntry))
    {
      util.fillOrderData(app, productionModule, logEntry, createProdShiftOrder);
    }
    else
    {
      createProdShiftOrder();
    }
  }

  function createProdShiftOrder(err)
  {
    if (err)
    {
      return done(err);
    }

    step(
      function()
      {
        const newProdShiftOrder = new ProdShiftOrder(logEntry.data);

        newProdShiftOrder.copyOperationData();
        newProdShiftOrder.save(this.next());
      },
      function(err, newProdShiftOrder)
      {
        if (err && err.code !== 11000)
        {
          productionModule.error(err, 'Failed to save a new prod shift order.', {
            prodLogEntryId: logEntry._id
          });

          return this.skip(err);
        }

        productionModule.getProdData('shift', logEntry.prodShift, this.parallel());
        productionModule.getProdShiftOrders(logEntry.prodShift, this.parallel());

        if (newProdShiftOrder)
        {
          setImmediate(this.parallel(), null, newProdShiftOrder);
        }
        else
        {
          productionModule.getProdData('order', logEntry.prodShiftOrder, this.parallel());
        }
      },
      function(err, prodShift, allProdShiftOrders, prodShiftOrder)
      {
        if (err)
        {
          return this.skip(err);
        }

        if (prodShiftOrder)
        {
          productionModule.setProdData(prodShiftOrder);
        }

        if (prodLine.isNew || !prodShift || !prodShiftOrder)
        {
          return this.skip();
        }

        const oldNextOrders = prodShift.getNextOrders();
        const newNextOrders = oldNextOrders.filter(next => next.orderNo !== prodShiftOrder.orderId);

        if (newNextOrders.length !== oldNextOrders.length)
        {
          prodShift.nextOrder = newNextOrders;
        }

        prodShift.recalcOrderData(allProdShiftOrders);
        prodShift.save(this.group());

        prodLine.set({
          prodShiftOrder: prodShiftOrder._id,
          prodDowntime: null
        });

        prodLine.save(this.group());
      },
      function(err)
      {
        if (err)
        {
          productionModule.error(err, 'Failed to save prod line after changing the prod shift order.', {
            prodLineId: prodLine._id,
            prodShiftOrderId: logEntry.prodShiftOrder,
            prodLogEntryId: logEntry._id
          });

          return this.skip(err);
        }
      },
      done
    );
  }
};
