// production/logEntryHandlers/checkSpigot.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');

module.exports = function(app, productionModule, prodLine, logEntry, done)
{
  step(
    function getModelsStep()
    {
      productionModule.getProdData('order', logEntry.prodShiftOrder, this.parallel());
      productionModule.getProdData('downtime', logEntry.data.prodDowntime, this.parallel());
    },
    function checkSpigotStep(err, prodShiftOrder, prodDowntime)
    {
      if (err)
      {
        productionModule.error(err, '[checkSpigot] Failed to find models for order.', {
          prodShiftOrderId: logEntry.prodShiftOrder,
          prodLogEntryId: logEntry._id
        });

        return this.skip();
      }

      if (!prodShiftOrder)
      {
        productionModule.warn('[checkSpigot] Order not found.', {
          prodShiftOrderId: logEntry.prodShiftOrder,
          prodLogEntryId: logEntry._id
        });

        return this.skip();
      }

      if (!prodDowntime)
      {
        productionModule.warn('[checkSpigot] Downtime not found.', {
          prodShiftOrderId: logEntry.prodShiftOrder,
          prodDowntimeId: logEntry.data.prodDowntime,
          prodLogEntryId: logEntry._id
        });

        return this.skip();
      }

      if (logEntry.data.nc12 !== logEntry.data.component.nc12)
      {
        productionModule.info('[checkSpigot] Invalid spigot.', {
          prodShiftOrderId: logEntry.prodShiftOrder,
          prodLogEntryId: logEntry._id,
          expected: logEntry.data.component.nc12,
          actual: logEntry.data.nc12
        });

        return this.skip();
      }

      const newSpigot = {
        prodDowntime: logEntry.data.prodDowntime,
        component: logEntry.data.component,
        initial: true,
        final: false
      };

      if (prodShiftOrder.spigot)
      {
        if (prodShiftOrder.spigot.final)
        {
          productionModule.warn('[checkSpigot] Final check already performed.', {
            prodLogEntryId: logEntry._id
          });

          return this.skip();
        }

        newSpigot.final = true;
      }

      prodShiftOrder.spigot = newSpigot;
      prodShiftOrder.save(this.parallel());

      prodDowntime.changes.push({
        date: logEntry.createdAt,
        user: prodShiftOrder.operator || logEntry.creator,
        data: {
          component: newSpigot.component
        },
        comment: 'WMES:SPIGOT:' + (newSpigot.final ? 'FINAL' : 'INITIAL')
      });
      prodDowntime.save(this.parallel());
    },
    function checkErrorsStep(err)
    {
      if (err)
      {
        productionModule.error(err, '[checkSpigot] Failed to update models.', {
          prodLogEntryId: logEntry._id
        });

        return this.skip();
      }
    },
    done
  );
};
