// production/logEntryHandlers/finishDowntimeAlert.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');

module.exports = function(app, productionModule, prodLine, logEntry, done)
{
  const logData = logEntry.data;

  step(
    function()
    {
      productionModule.getProdData('downtime', logData.downtimeId, this.next());
    },
    function(err, prodDowntime)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to find prod downtime to finish alert.', {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });

        return this.done(done, err);
      }

      if (!prodDowntime)
      {
        productionModule.warn(`Couldn't find prod downtime to finish alert.`, {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });

        return this.done(done);
      }

      const alert = _.find(prodDowntime.alerts, '_id', logData.alertId);

      if (!alert)
      {
        productionModule.warn(`Tried to finish alert that wasn't started.`, {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });

        return done();
      }

      if (!alert.active)
      {
        productionModule.warn('Tried to finish alert that was already finished.', {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });

        return done();
      }

      alert.active = false;

      prodDowntime.changes.push({
        date: logEntry.createdAt,
        user: logEntry.creator,
        data: {
          alert: {
            _id: alert._id,
            name: alert.name
          }
        },
        comment: 'WMES:ALERTS:FINISH'
      });

      prodDowntime.save(this.next());
    },
    function(err)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to save prod downtime after finishing alert.', {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });
      }

      return done(err);
    }
  );
};
