// production/logEntryHandlers/notifyDowntimeAlert.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');

module.exports = function(app, productionModule, prodLine, logEntry, done)
{
  const logData = logEntry.data;

  step(
    function()
    {
      productionModule.getProdData('downtime', logData.downtimeId, this.next());
    },
    function(err, prodDowntime)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to find prod downtime to record alert action.', {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });

        return this.done(done, err);
      }

      if (!prodDowntime)
      {
        productionModule.warn(`Couldn't find prod downtime to record alert action.`, {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });

        return this.done(done);
      }

      let alert = _.find(prodDowntime.alerts, '_id', logData.alertId);

      if (alert && !alert.active)
      {
        productionModule.warn('Tried to record action of an already finished alert.', {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });

        return done();
      }

      const isNew = !alert;

      if (isNew)
      {
        alert = {
          _id: logData.alertId,
          name: logData.alertName,
          action: logData.action,
          active: true
        };

        prodDowntime.alerts.push(alert);
      }
      else
      {
        alert.action = logData.action;
      }

      if (isNew || !_.isEmpty(logData.recipients))
      {
        prodDowntime.changes.push({
          date: logEntry.createdAt,
          user: logEntry.creator,
          data: {
            alert: {
              _id: alert._id,
              name: alert.name,
              recipients: logData.recipients
            }
          },
          comment: 'WMES:ALERTS:' + (isNew ? 'START' : 'NOTIFY')
        });
      }

      prodDowntime.save(this.next());
    },
    function(err)
    {
      if (err)
      {
        productionModule.error(err, `Failed to save prod downtime after recording alert action.`, {
          prodDowntimeId: logData.downtimeId,
          prodDowntimeAlertId: logData.alertId,
          prodLogEntryId: logEntry._id
        });
      }

      return done(err);
    }
  );
};
