// production/logEntryHandlers/startDowntime.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');

module.exports = function(app, productionModule, prodLine, logEntry, done)
{
  const ProdDowntime = app[productionModule.config.mongooseId].model('ProdDowntime');

  if (logEntry.data.date === undefined || logEntry.data.operators === undefined)
  {
    productionModule.getProdData('shift', logEntry.prodShift, function(err, prodShift)
    {
      if (err)
      {
        productionModule.error(err, 'Failed to get prod shift to start a downtime.', {
          prodShiftId: logEntry.prodShift,
          prodLogEntryId: logEntry._id
        });

        return done(err);
      }

      if (prodShift)
      {
        logEntry.data.date = prodShift.date;
        logEntry.data.shift = prodShift.shift;
        logEntry.data.operators = prodShift.operators;
      }

      createProdDowntime();
    });
  }
  else
  {
    createProdDowntime();
  }

  function createProdDowntime()
  {
    const prodDowntime = new ProdDowntime(logEntry.data);

    step(
      function findProdShiftOrderStep()
      {
        if (logEntry.data.prodShiftOrder)
        {
          productionModule.getProdData('order', logEntry.data.prodShiftOrder, this.next());
        }
      },
      function saveProdDowntimeStep(err, prodShiftOrder)
      {
        if (err)
        {
          productionModule.error(err, 'Failed to find order of a new downtime.', {
            prodLogEntryId: logEntry._id
          });

          return this.done(done, err);
        }

        prodDowntime.workerCount = prodShiftOrder ? prodShiftOrder.workerCount : 1;
        prodDowntime.orderData = ProdDowntime.getOrderData(prodShiftOrder);

        prodDowntime.save(this.next());
      },
      function updateProdLineStep(err)
      {
        if (err && err.code !== 11000)
        {
          productionModule.error(err, 'Failed to save a new prod downtime.', {
            prodLogEntryId: logEntry._id
          });

          return this.done(done, err);
        }

        if (!err)
        {
          productionModule.setProdData(prodDowntime);
        }

        if (prodLine.isNew)
        {
          return this.done(done);
        }

        prodLine.prodDowntime = prodDowntime._id;

        prodLine.save(this.next());
      },
      function finalizeStep(err)
      {
        if (err)
        {
          productionModule.error(err, 'Failed to save prod line after changing the prod downtime.', {
            prodLineId: prodLine._id,
            prodLogEntryId: logEntry._id,
            prodDowntimeId: prodDowntime._id
          });
        }

        done(err);
      }
    );
  }
};
