// wmes-wiring/routes/exportOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module, req, res, next) =>
{
  const {express, WiringOrder} = module;

  const options = {
    filename: 'WMES-WIRING-ORDERS',
    freezeRows: 1,
    freezeColumns: 1,
    columns: {
      status: 9,
      nc12: 13,
      name: 45,
      qtyTodo: {type: 'integer', width: 8},
      qtyDone: {type: 'integer', width: 8},
      mrp: 4,
      date: 'date+utc',
      startedAt: {type: 'datetime+utc', width: 18},
      finishedAt: {type: 'datetime+utc', width: 18},
      orders: 40
    },
    serializeRow: (wiringOrder) =>
    {
      return {
        status: wiringOrder.status,
        nc12: wiringOrder.nc12,
        name: wiringOrder.name,
        qtyTodo: wiringOrder.qty,
        qtyDone: wiringOrder.qtyDone,
        mrp: wiringOrder.mrp,
        date: wiringOrder.date,
        startedAt: wiringOrder.startedAt,
        finishedAt: wiringOrder.finishedAt,
        orders: wiringOrder.orders.map(leadingOrder => leadingOrder._id).join(', ')
      };
    },
    model: WiringOrder
  };

  express.crud.exportRoute(app, options, req, res, next);
};
