// paintShop/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  fteId: 'fte',
  settingsId: 'settings',
  updaterId: 'updater',
  reportsId: 'reports',
  generator: false,
  loadSecretKey: null,
  generatorLogPath: null
};

exports.models = [
  require('./models/paintShopDropZone'),
  require('./models/paintShopEvent'),
  require('./models/paintShopLoad'),
  require('./models/paintShopOrder'),
  require('./models/paintShopPaint'),
  require('orders/models/order'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('planning/models/plan'),
  require('planning/models/planSettings')
];

exports.userPrivileges = [
  'PAINT_SHOP:VIEW',
  'PAINT_SHOP:MANAGE',
  'PAINT_SHOP:PAINTER',
  'PAINT_SHOP:DROP_ZONES'
];

exports.republishTopics = [
  'paintShop.events.saved',
  'paintShop.orders.changed.*',
  'paintShop.orders.updated.*',
  'paintShop.dropZones.updated.*',
  'paintShop.paints.*',
  'paintShop.load.updated'
];

exports.recordTopics = {
  warning: ['paintShop.paints.deleted']
};

exports.optionalModules = {
  'mongoose settings': require('./generator'),
  'mongoose fte user express updater reports settings': require('paintShop/routes')
};
