// paintShop/routes.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const moment = require('moment');
const step = require('h5.step');
const loadReport = require('./loadReport');

module.exports = function setUpPaintShopRoutes(app, module)
{
  const updaterModule = app[module.config.updaterId];
  const reportsModule = app[module.config.reportsId];
  const settings = app[module.config.settingsId];
  const express = app[module.config.expressId];
  const userModule = app[module.config.userId];
  const mongoose = app[module.config.mongooseId];
  const PaintShopEvent = mongoose.model('PaintShopEvent');
  const PaintShopOrder = mongoose.model('PaintShopOrder');
  const PaintShopDropZone = mongoose.model('PaintShopDropZone');
  const PaintShopPaint = mongoose.model('PaintShopPaint');
  const PaintShopLoad = mongoose.model('PaintShopLoad');

  const canView = userModule.auth('LOCAL', 'PAINT_SHOP:VIEW', 'PLANNING:VIEW');
  const canUpdate = userModule.auth('LOCAL', 'PAINT_SHOP:PAINTER', 'PAINT_SHOP:MANAGE');
  const canManage = userModule.auth('PAINT_SHOP:MANAGE');
  const canManageDropZones = userModule.auth('PAINT_SHOP:DROP_ZONES');

  // Apps
  ['ps-queue', 'ps-load'].forEach(appId =>
  {
    express.get(`/${appId}`, userModule.auth('LOCAL', 'PAINT_SHOP:VIEW'), (req, res) =>
    {
      res.format({
        'text/html': function()
        {
          res.render('index', updaterModule.getAppTemplateData(appId, req));
        }
      });
    });
  });

  // Settings
  express.get(
    '/paintShop/settings',
    canView,
    (req, res, next) =>
    {
      req.rql.selector = {
        name: 'regex',
        args: ['_id', '^paintShop\\.']
      };

      return next();
    },
    express.crud.browseRoute.bind(null, app, settings.Setting)
  );

  express.put('/paintShop/settings/:id', canManage, settings.updateRoute);

  // Queues
  express.post('/paintShop/:id;generate', canManage, generateRoute);

  // Events
  express.get(
    '/paintShop/events',
    canView,
    express.crud.browseRoute.bind(null, app, PaintShopEvent)
  );

  // Orders
  express.get(
    '/paintShop/orders',
    canView,
    prepareCurrentDate,
    express.crud.browseRoute.bind(null, app, PaintShopOrder)
  );

  express.get(
    '/paintShop/orders;export.:format?',
    canView,
    express.crud.exportRoute.bind(null, app, {
      filename: 'WMES-PAINT_SHOP-ORDERS',
      freezeRows: 1,
      freezeColumns: 1,
      columns: {
        no: 7,
        type: 11,
        orderNo: 10,
        status: 9,
        nc12: 13,
        name: 45,
        quantity: {type: 'integer', width: 8},
        unit: 4,
        mrp: 4,
        date: 'date+utc',
        assemblyTime: {type: 'datetime+utc', width: 18},
        startedAt: {type: 'datetime+utc', width: 18},
        finishedAt: {type: 'datetime+utc', width: 18}
      },
      serializeRow: exportPaintShopOrder,
      model: PaintShopOrder
    })
  );

  express.get('/paintShop/orders/:id', canView, express.crud.readRoute.bind(null, app, PaintShopOrder));

  express.patch('/paintShop/orders/:id', canUpdate, updatePaintShopOrderRoute);

  // Drop zones
  express.get(
    '/paintShop/dropZones',
    canView,
    express.crud.browseRoute.bind(null, app, PaintShopDropZone)
  );

  express.post('/paintShop/dropZones/:date/:mrp', canManageDropZones, updateDropZoneRoute);

  // Paints
  express.get('/paintShop/paints', canView, express.crud.browseRoute.bind(null, app, PaintShopPaint));

  express.post('/paintShop/paints', canManage, express.crud.addRoute.bind(null, app, PaintShopPaint));

  express.get('/paintShop/paints/:id', canView, express.crud.readRoute.bind(null, app, PaintShopPaint));

  express.put('/paintShop/paints/:id', canManage, express.crud.editRoute.bind(null, app, PaintShopPaint));

  express.delete('/paintShop/paints/:id', canManage, express.crud.deleteRoute.bind(null, app, PaintShopPaint));

  // Load
  express.post('/paintShop/load/update', updateLoadRoute);

  express.get('/paintShop/load/stats', canView, getLoadStatsRoute);

  express.get('/paintShop/load/recent', canView, getRecentLoadRoute);

  express.get(
    '/paintShop/load/report',
    canView,
    reportsModule.helpers.sendCachedReport.bind(null, 'paintShop/load'),
    loadReportRoute
  );

  function prepareCurrentDate(req, res, next)
  {
    req.rql.selector.args.forEach(function(term)
    {
      if (term.name !== 'eq' || term.args[0] !== 'date')
      {
        return;
      }

      const date = term.args[1];

      if (date === 'current')
      {
        const m = moment();

        if (m.hours() < 17)
        {
          m.startOf('day').subtract(1, 'days');
        }
        else
        {
          m.startOf('day').add(1, 'days');
        }

        term.args[1] = moment.utc(m.format('YYYY-MM-DD'), 'YYYY-MM-DD').toDate();
      }
      else if (/^[0-9]+-[0-9]+-[0-9]+$/.test(date))
      {
        term.args[1] = moment.utc(date, 'YYYY-MM-DD').toDate();
      }
    });

    next();
  }

  function updatePaintShopOrderRoute(req, res, next)
  {
    const input = req.body;

    step(
      function()
      {
        PaintShopOrder.findById(req.params.id).exec(this.next());
      },
      function(err, psOrder)
      {
        if (err)
        {
          return this.skip(err);
        }

        if (!psOrder)
        {
          return this.skip(app.createError('NOT_FOUND', 404));
        }

        psOrder.act(input, this.next());
      },
      function(err, changes)
      {
        if (err)
        {
          return next(err);
        }

        if (!changes)
        {
          return res.sendStatus(204);
        }

        res.json(changes);

        app.broker.publish(`paintShop.orders.updated.${req.params.id}`, changes);

        const user = userModule.createUserInfo(req.session.user, req);

        if (!req.session.user.loggedIn && input.user)
        {
          user.id = input.user.id;
          user.label = input.user.label;
        }

        PaintShopEvent.record({
          type: input.action,
          time: new Date(),
          user,
          order: req.params.id,
          data: _.omit(input, ['action'])
        });
      }
    );
  }

  function generateRoute(req, res)
  {
    app.broker.publish('paintShop.generator.requested', {
      date: req.params.id
    });

    res.sendStatus(204);
  }

  function updateDropZoneRoute(req, res, next)
  {
    const {date, mrp} = req.params;

    if (!_.isString(date) || !moment(date, 'YYYY-MM-DD').isValid() || !_.isString(mrp) || _.isEmpty(mrp))
    {
      return next(app.createError('INPUT', 400));
    }

    const state = !!req.body.state;

    PaintShopDropZone.collection.updateOne({_id: {date, mrp}}, {$set: {state}}, {upsert: true}, err =>
    {
      if (err)
      {
        return next(err);
      }

      const dropZone = {
        _id: {date, mrp},
        state
      };

      res.json(dropZone);

      app.broker.publish(`paintShop.dropZones.updated.${date}`, dropZone);
    });
  }

  function exportPaintShopOrder(leadingOrder)
  {
    const rows = [{
      no: `${leadingOrder.no}.`,
      type: 'leading',
      orderNo: leadingOrder.order,
      status: leadingOrder.status,
      nc12: leadingOrder.nc12,
      quantity: leadingOrder.qty,
      unit: 'PCE',
      name: leadingOrder.name,
      mrp: leadingOrder.mrp,
      assemblyTime: new Date(leadingOrder.startTime),
      startedAt: leadingOrder.startedAt,
      finishedAt: leadingOrder.finishedAt
    }];

    leadingOrder.childOrders.forEach((childOrder, childOrderI) =>
    {
      rows.push({
        no: `${leadingOrder.no}.${childOrderI + 1}.`,
        type: 'child',
        orderNo: childOrder.order,
        status: leadingOrder.status,
        nc12: childOrder.nc12,
        quantity: childOrder.qty,
        unit: 'PCE',
        name: childOrder.name,
        mrp: leadingOrder.mrp,
        assemblyTime: new Date(leadingOrder.startTime),
        startedAt: leadingOrder.startedAt,
        finishedAt: leadingOrder.finishedAt
      });

      childOrder.components.forEach((component, componentI) =>
      {
        rows.push({
          no: `${leadingOrder.no}.${childOrderI + 1}.${componentI + 1}.`,
          type: 'component',
          orderNo: childOrder.order,
          status: leadingOrder.status,
          nc12: component.nc12,
          quantity: Math.ceil(component.qty),
          unit: component.unit,
          name: component.name,
          mrp: leadingOrder.mrp,
          assemblyTime: new Date(leadingOrder.startTime),
          startedAt: leadingOrder.startedAt,
          finishedAt: leadingOrder.finishedAt
        });
      });
    });

    return rows;
  }

  function updateLoadRoute(req, res, next)
  {
    const {secretKey, items} = req.body;

    if (secretKey !== module.config.loadSecretKey)
    {
      return next(app.createError('INVALID_SECRET_KEY', 400));
    }

    if (!Array.isArray(items) || !items.length)
    {
      return next(app.createError('INVALID_ITEMS', 400));
    }

    const docs = items
      .map(item => ({_id: new Date(item._id), d: item.d}))
      .filter(doc => !isNaN(doc._id.getTime()) && doc.d >= 0);

    if (!docs.length)
    {
      return next(app.createError('INVALID_ITEMS', 400));
    }

    step(
      function()
      {
        PaintShopLoad.collection.insertMany(docs, {ordered: false}, this.next());
      },
      function(err)
      {
        if (err)
        {
          module.warn(`Failed to update load: ${err.message}`);
        }

        PaintShopLoad.getStats(this.next());
      },
      function(err, stats)
      {
        if (err)
        {
          return next(err);
        }

        res.sendStatus(204);

        app.broker.publish('paintShop.load.updated', {
          stats,
          items: docs
        });
      }
    );
  }

  function getLoadStatsRoute(req, res, next)
  {
    PaintShopLoad.getStats((err, stats) =>
    {
      if (err)
      {
        return next(err);
      }

      res.json(stats);
    });
  }

  function getRecentLoadRoute(req, res, next)
  {
    const limit = Math.min(2560, Math.max(1, req.rql.limit || 100));

    PaintShopLoad.find({}, {_id: 0, d: 1}).sort({_id: -1}).limit(limit).lean().exec((err, docs) =>
    {
      if (err)
      {
        return next(err);
      }

      res.json({
        totalCount: docs.length,
        collection: docs.map(d => Math.round(d.d / 1000)).reverse()
      });
    });
  }

  function loadReportRoute(req, res, next)
  {
    const query = req.query;
    const options = {
      fromTime: reportsModule.helpers.getTime(query.from) || null,
      toTime: reportsModule.helpers.getTime(query.to) || null,
      interval: reportsModule.helpers.getInterval(query.interval, 'day')
    };

    reportsModule.helpers.generateReport(
      app,
      reportsModule,
      loadReport,
      'paintShop/load',
      req.reportHash,
      options,
      function(err, reportJson)
      {
        if (err)
        {
          return next(err);
        }

        res.type('json');
        res.send(reportJson);
      }
    );
  }
};
