// planning/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');
const {ObjectId} = require('mongoose').Types;

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  userId: 'user',
  expressId: 'express',
  ordersId: 'orders',
  settingsId: 'settings',
  reportsId: 'reports',
  generators: []
};

exports.models = [
  require('settings/models/setting'),
  require('orders/models/component'),
  require('orders/models/order'),
  require('orders/models/orderEto'),
  require('paintShop/models/paintShopOrder'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('reports/models/dailyMrpCount'),
  require('./models/plan'),
  require('./models/planChange'),
  require('./models/planSettings'),
  require('./models/planOrderGroup'),
  require('./models/whOrderStatus')
];

exports.republishTopics = ['planning.**'];

exports.userPrivileges = ['PLANNING:VIEW', 'PLANNING:MANAGE', 'PLANNING:PLANNER', 'PLANNING:WHMAN'];

exports.optionalModules = {
  'mongoose': [
    require('./generator'),
    require('./mrpAutoLocker'),
    require('./activeMrps')
  ],
  'mongoose user express settings reports': require('./routes')
};

exports.start = (app, module) =>
{
  let requestTimer = null;

  app.broker.subscribe('app.started', () =>
  {
    createSpecialOrderGroups();
  });

  app.broker.subscribe('orders.synced', m =>
  {
    const now = moment();
    const planDate = moment.utc(now.format('YYYY-MM-DD'), 'YYYY-MM-DD');

    if (now.hours() >= 6)
    {
      planDate.add(1, 'days');
    }

    const PlanSettings = app[module.config.mongooseId].model('PlanSettings');

    PlanSettings.find({_id: {$gte: planDate}}).sort({_id: 1}).limit(1).exec((err, settings) =>
    {
      if (err)
      {
        module.error(err, `Failed to find plan settings for generator request.`);
      }

      const freezeFirstShiftOrders = settings.length === 1 && settings[0].shouldFreezeFirstShiftOrders(30);

      if (freezeFirstShiftOrders || m.created || m.updated || m.removed)
      {
        app.broker.publish('planning.generator.requested', {
          freezeFirstShiftOrders
        });
      }
    });
  });

  app.broker.subscribe('settings.updated.production.lineAutoDowntimes', () =>
  {
    app.broker.publish('planning.generator.requested', {
      reloadAutoDowntimes: true
    });
  });

  app.broker.subscribe('settings.updated.production.taktTime.coeffs', () =>
  {
    clearTimeout(requestTimer);
    requestTimer = setTimeout(app.broker.publish.bind(app.broker, 'planning.generator.requested', {}), 15000);
  });

  app.broker.subscribe('subdivisions.edited', () =>
  {
    app.broker.publish('planning.generator.requested', {
      reloadAutoDowntimes: true
    });
  });

  app.broker.subscribe('planning.settings.updated', m =>
  {
    if (Array.isArray(m.changes) && m.changes.length)
    {
      app.broker.publish('planning.generator.requested', {
        date: m.date
      });
    }
  });

  app.broker.subscribe('orderDocuments.eto.synced', m =>
  {
    if (module.eto)
    {
      module.eto.add(m.nc12);
    }
  });

  app.broker.subscribe('planning.generator.requested', function(message)
  {
    if (!message.date)
    {
      clearTimeout(requestTimer);
      requestTimer = null;
    }
  });

  async function createSpecialOrderGroups()
  {
    const {PlanOrderGroup} = module;

    try
    {
      const orderGroups = [
        {
          _id: new ObjectId('000000000000000000000000'),
          name: 'Bez grupy',
          description: 'Tutaj trafią zlecenia niepasujące do żadnej innej grupy.',
          productInclude: [],
          productExclude: [],
          bomInclude: [],
          bomExclude: [],
          names: {}
        }
      ];

      await PlanOrderGroup.collection.insertMany(orderGroups);
    }
    catch (err)
    {
      if (err.code === 11000)
      {
        return;
      }

      module.error(err, 'Failed to create special order groups.');
    }
  }
};
