// planning/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  userId: 'user',
  expressId: 'express',
  ordersId: 'orders',
  settingsId: 'settings',
  reportsId: 'reports',
  generator: true
};

exports.models = [
  require('settings/models/setting'),
  require('orders/models/order'),
  require('orders/models/orderEto'),
  require('paintShop/models/paintShopOrder'),
  require('./models/plan'),
  require('./models/planChange'),
  require('./models/planSettings'),
  require('./models/whOrderStatus')
];

exports.optionalModules = {
  'mongoose': require('./generator'),
  'mongoose user express settings reports': require('./routes')
};

exports.start = (app, module) =>
{
  app.broker.subscribe('orders.synced', m =>
  {
    const now = moment();
    const planDate = moment.utc(now.format('YYYY-MM-DD'), 'YYYY-MM-DD');

    if (now.hours() >= 6)
    {
      planDate.add(1, 'days');
    }

    const PlanSettings = app[module.config.mongooseId].model('PlanSettings');

    PlanSettings.find({_id: {$gte: planDate}}).sort({_id: 1}).limit(1).exec((err, settings) =>
    {
      if (err)
      {
        module.error(`Failed to find plan settings for generator request: ${err.message}`);
      }

      const freezeFirstShiftOrders = settings.length === 1 && settings[0].shouldFreezeFirstShiftOrders(30);

      if (freezeFirstShiftOrders || m.created || m.updated || m.removed)
      {
        app.broker.publish('planning.generator.requested', {
          freezeFirstShiftOrders
        });
      }
    });
  });

  app.broker.subscribe('settings.updated.production.lineAutoDowntimes', () =>
  {
    app.broker.publish('planning.generator.requested', {
      reloadAutoDowntimes: true
    });
  });

  app.broker.subscribe('subdivisions.edited', () =>
  {
    app.broker.publish('planning.generator.requested', {
      reloadAutoDowntimes: true
    });
  });

  app.broker.subscribe('planning.settings.updated', m =>
  {
    if (Array.isArray(m.changes) && m.changes.length)
    {
      app.broker.publish('planning.generator.requested', {
        date: m.date
      });
    }
  });

  app.broker.subscribe('orderDocuments.eto.synced', m =>
  {
    if (module.eto)
    {
      module.eto.add(m.nc12);
    }
  });
};
