// wh/models/oldWhPendingComponents.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'OldWhPendingComponents';

exports.setUp = (app, mongoose) =>
{
  const pendingComponentsSchema = new mongoose.Schema({
    date: Date,
    set: Number,
    line: String
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  pendingComponentsSchema.statics.TOPIC_PREFIX = 'old.wh.pending.components';
  pendingComponentsSchema.statics.BROWSE_LIMIT = 0;

  pendingComponentsSchema.statics.deleteBySet = async function(date, set, session)
  {
    const WhPendingComponents = this;

    const result = {
      deleted: [],
      messages: []
    };

    const pendingDeliveries = await WhPendingComponents.find({set, date}).lean().exec();

    if (!pendingDeliveries.length)
    {
      return result;
    }

    const $in = [];

    pendingDeliveries.forEach(pendingDelivery =>
    {
      $in.push(pendingDelivery._id);
      result.deleted.push(pendingDelivery);
    });

    await WhPendingComponents.collection.deleteMany({_id: {$in}}, {session});

    result.messages.push({
      topic: `${WhPendingComponents.TOPIC_PREFIX}`,
      message: {
        deleted: result.deleted
      }
    });

    return result;
  };

  mongoose.model(exports.name, pendingComponentsSchema);
};
