// wh/models/oldWhPendingDelivery.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'OldWhPendingDelivery';

exports.setUp = (app, mongoose) =>
{
  const pendingDeliverySchema = new mongoose.Schema({
    date: Date,
    set: Number,
    line: String
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  pendingDeliverySchema.statics.TOPIC_PREFIX = 'old.wh.pendingDeliveries';
  pendingDeliverySchema.statics.BROWSE_LIMIT = 100;

  pendingDeliverySchema.statics.deleteBySet = async function(date, set, session)
  {
    const WhPendingDelivery = this;

    const result = {
      deleted: [],
      messages: []
    };

    const pendingDeliveries = await WhPendingDelivery.find({set, date}).lean().exec();

    if (!pendingDeliveries.length)
    {
      return result;
    }

    const $in = [];

    pendingDeliveries.forEach(pendingDelivery =>
    {
      $in.push(pendingDelivery._id);
      result.deleted.push(pendingDelivery);
    });

    await WhPendingDelivery.collection.deleteMany({_id: {$in}}, {session});

    result.messages.push({
      topic: `${WhPendingDelivery.TOPIC_PREFIX}`,
      message: {
        deleted: result.deleted
      }
    });

    return result;
  };

  mongoose.model(exports.name, pendingDeliverySchema);
};
