// wh/models/oldWhEvent.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'OldWhEvent';

exports.setUp = (app, mongoose) =>
{
  const whEventSchema = new mongoose.Schema({
    type: {
      type: String,
      required: true
    },
    time: {
      type: Date,
      required: true
    },
    user: {
      type: {},
      required: true
    },
    order: {
      type: String,
      default: null,
      ref: 'OldWhOrder'
    },
    data: {}
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  whEventSchema.statics.TOPIC_PREFIX = 'old.wh.events';
  whEventSchema.statics.BROWSE_LIMIT = 100;

  whEventSchema.index({time: -1});
  whEventSchema.index({type: 1, time: -1});
  whEventSchema.index({order: 1, time: -1});

  whEventSchema.post('save', function(doc)
  {
    app.broker.publish(`${whEventSchema.statics.TOPIC_PREFIX}.saved`, doc.toJSON());
  });

  whEventSchema.statics.record = function(event)
  {
    (new this(event)).save(err =>
    {
      if (err)
      {
        app.error(err, 'Failed to record event.', {
          module: 'wh',
          event
        });
      }
    });
  };

  mongoose.model(exports.name, whEventSchema);
};
