// wmes-gft-frontend/models/gftTester.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'GftTester';

exports.setUp = (app, mongoose) =>
{
  const programSchema = new mongoose.Schema({
    leds: {
      type: Number,
      min: 1,
      max: 1000,
      required: true
    },
    outputs: [{
      type: Number,
      min: 1,
      max: 16
    }]
  }, {
    _id: false
  });

  const schema = new mongoose.Schema({
    active: {
      type: Boolean,
      required: true
    },
    name: {
      type: String,
      required: true,
      trim: true,
      maxLength: 100
    },
    line: {
      type: String,
      required: true,
      trim: true,
      maxLength: 30
    },
    station: {
      type: Number,
      required: true,
      min: 0,
      max: 20
    },
    host: {
      type: String,
      required: true,
      trim: true,
      maxLength: 100
    },
    port: {
      type: Number,
      required: true,
      min: 1,
      max: 0xFFFF
    },
    program: [
      programSchema
    ]
  }, {
    id: false
  });

  schema.statics.TOPIC_PREFIX = 'gft.testers';
  schema.statics.BROWSE_LIMIT = 100;

  schema.index({host: 1, port: 1}, {unique: true});

  return schema;
};
