// prodDowntimeAlerts/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const setUpServer = require('./server');
const setUpClient = require('./client');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  fteId: 'fte',
  messengerServerId: null,
  messengerClientId: null,
  mailSenderId: 'mail/sender',
  smsSenderId: 'sms/sender'
};

exports.recordTopics = {
  blacklist: ['prodDowntimeAlerts.added']
};

exports.userPrivileges = ['PROD_DOWNTIME_ALERTS:VIEW', 'PROD_DOWNTIME_ALERTS:MANAGE'];

exports.models = [
  require('./models/prodDowntimeAlert'),
  require('user/models/user'),
  require('aors/models/aor'),
  require('downtimeReasons/models/downtimeReason'),
  require('prodFlows/models/prodFlow'),
  require('workCenters/models/workCenter'),
  require('prodLines/models/prodLine'),
  require('prodLogEntries/models/prodLogEntry'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('prodDowntimes/models/prodDowntime')
];

exports.optionalModules = {
  'mongoose user express': require('./routes'),
  'mongoose fte messengerServer': setUpServer,
  'mongoose fte messengerClient': setUpClient
};

exports.start = (app, module) =>
{
  app.onModuleReady(
    [
      module.config.mongooseId,
      module.config.fteId
    ],
    function()
    {
      if (!module.config.messengerServerId && !module.config.messengerClientId)
      {
        setUpServer(app, module);
        setUpClient(app, module);
      }
    }
  );
};
