// prodDowntimeAlerts/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const setUpRoutes = require('./routes');
const setUpServer = require('./server');
const setUpClient = require('./client');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  fteId: 'fte',
  messengerServerId: null,
  messengerClientId: null,
  mailSenderId: 'mail/sender',
  smsSenderId: 'sms/sender'
};

exports.models = [
  require('./models/prodDowntimeAlert'),
  require('user/models/user'),
  require('aors/models/aor'),
  require('downtimeReasons/models/downtimeReason'),
  require('prodFlows/models/prodFlow'),
  require('workCenters/models/workCenter'),
  require('prodLines/models/prodLine'),
  require('prodLogEntries/models/prodLogEntry'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('prodDowntimes/models/prodDowntime')
];

exports.start = function startProdDowntimeAlertsModule(app, module)
{
  app.onModuleReady(
    [
      module.config.mongooseId,
      module.config.userId,
      module.config.expressId
    ],
    setUpRoutes.bind(null, app, module)
  );

  app.onModuleReady(
    [
      module.config.mongooseId,
      module.config.fteId,
      module.config.messengerServerId
    ],
    setUpServer.bind(null, app, module)
  );

  app.onModuleReady(
    [
      module.config.mongooseId,
      module.config.fteId,
      module.config.messengerClientId
    ],
    setUpClient.bind(null, app, module)
  );

  app.onModuleReady(
    [
      module.config.mongooseId,
      module.config.fteId
    ],
    function()
    {
      if (!module.config.messengerServerId && !module.config.messengerClientId)
      {
        setUpServer(app, module);
        setUpClient(app, module);
      }
    }
  );
};
