// wmes-osh/routes/observations/add.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const locks = require('util/locks');

module.exports = async (app, module, req, res, next) =>
{
  const {
    OshObservation
  } = module;

  const releaseLocks = [];

  const input = _.pick(req.body, [
    'observationKind',
    'subject',
    'userWorkplace',
    'userDepartment',
    'company',
    'companyName',
    'workplace',
    'department',
    'building',
    'location',
    'station',
    'date',
    'behaviors',
    'workConditions',
    'attachments'
  ]);

  if (input.attachments && Array.isArray(input.attachments.added))
  {
    input.attachments = module.attachments.claimPendingUploads(input.attachments.added);
  }
  else
  {
    input.attachments = [];
  }

  let session = null;

  try
  {
    session = await OshObservation.startSession();
    session.startTransaction();

    const creator = module.createUserInfo(req.session.user, req);
    const createdAt = new Date();

    const observation = new OshObservation({
      ...input,
      status: 'inProgress',
      createdAt,
      updatedAt: createdAt,
      creator,
      updater: creator
    });

    observation.$session(session);

    const [coordinators] = await Promise.all([
      OshObservation.resolveCoordinators(observation),
      OshObservation.assignNextRid(observation)
    ]);

    observation.coordinators = coordinators;

    observation.updateUsers([], observation.creator);

    await observation.save();

    const messages = [{
      topic: `${OshObservation.TOPIC_PREFIX}.added`,
      message: {
        user: observation.creator,
        model: observation
      }
    }];
    const models = new Set();
    const resolutions = {
      nearMiss: new Map(),
      kaizen: new Map()
    };
    const relation = observation.getRelation();

    observation.behaviors.concat(observation.workConditions).forEach(o =>
    {
      if (o.resolution._id)
      {
        resolutions[o.resolution.type].set(o.resolution._id, o);
      }
    });

    for (const relationType of Object.keys(resolutions))
    {
      if (!resolutions[relationType].size)
      {
        continue;
      }

      const Entry = module.TYPE_TO_MODEL[relationType];
      const ids = Array.from(resolutions[relationType].keys());

      releaseLocks.push(...await Promise.all(
        ids.map(id => locks.createAsync(`${Entry.TOPIC_PREFIX}.${id}`))
      ));

      const entries = await Entry
        .find({_id: {$in: ids}})
        .session(session)
        .exec();

      for (const entry of entries)
      {
        models.add(observation);
        models.add(Entry.addRelation({
          model: entry,
          relation,
          updater: observation.creator,
          updatedAt: observation.createdAt,
          messages
        }));

        resolutions[relationType].get(entry._id).implementer = entry.creator;
      }
    }

    if (models.size)
    {
      await Promise.all(Array.from(models).map(model => model ? model.save() : null));
    }

    if (session)
    {
      await session.commitTransaction();
    }

    messages.forEach(({topic, message}) => app.broker.publish(topic, message));

    res.json(observation);
  }
  catch (err)
  {
    next(err);

    if (session)
    {
      try
      {
        await session.abortTransaction();
      }
      catch (err)
      {
        if (err.message !== 'Cannot call abortTransaction twice')
        {
          module.error(err, `Failed to abort a transaction.`);
        }
      }
    }
  }
  finally
  {
    if (session)
    {
      session.endSession();
      session = null;
    }

    releaseLocks.forEach(releaseLock => releaseLock ? releaseLock() : null);
  }
};
