// wmes-osh/routes/observations/resolutions.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = async (app, module, req, res, next) =>
{
  const {
    OshObservation,
    OshNearMiss,
    OshKaizen
  } = module;

  try
  {
    const observation = await OshObservation
      .findOne({
        [/^[0-9]+$/.test(req.params.id) ? '_id' : 'rid']: req.params.id
      })
      .select({
        'behaviors.resolution': 1,
        'workConditions.resolution': 1
      })
      .lean()
      .exec();

    if (!observation)
    {
      return next(app.createError('Entry not found.', 'NOT_FOUND', 404));
    }

    const rids = [];

    observation.behaviors.concat(observation.workConditions).forEach(o =>
    {
      if (o.resolution.rid)
      {
        rids.push(o.resolution.rid);
      }
    });

    const [nearMisses, kaizens] = await Promise.all([
      OshNearMiss
        .find({rid: {$in: rids}})
        .select({
          rid: 1,
          subject: 1,
          status: 1,
          creator: 1,
          createdAt: 1
        })
        .lean()
        .exec(),
      OshKaizen
        .find({rid: {$in: rids}})
        .select({
          rid: 1,
          subject: 1,
          status: 1,
          creator: 1,
          createdAt: 1
        })
        .lean()
        .exec()
    ]);

    const resolutions = nearMisses.concat(kaizens);

    res.json({
      totalCount: resolutions.length,
      collection: resolutions
    });
  }
  catch (err)
  {
    next(err);
  }
};
