// wmes-ct-frontend/models/ctBalancingPce.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'CtBalancingPce';

exports.setUp = (app, mongoose) =>
{
  const balancingPceSchema = new mongoose.Schema({
    createdBy: {},
    createdAt: Date,
    line: {
      type: String,
      required: true
    },
    station: {
      type: Number,
      required: true,
      min: 1,
      max: 10
    },
    order: {},
    startedAt: {
      type: Date,
      required: true
    },
    finishedAt: {
      type: Date,
      required: true
    },
    duration: Number
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  balancingPceSchema.statics.TOPIC_PREFIX = 'ct.balancing.pces';
  balancingPceSchema.statics.BROWSE_LIMIT = 100;

  balancingPceSchema.index({startedAt: -1});
  balancingPceSchema.index({'order._id': 1});
  balancingPceSchema.index({'order.nc12': 1});
  balancingPceSchema.index({line: 1, station: 1});

  mongoose.model(exports.name, balancingPceSchema);
};
