// wmes-gft-frontend/routes/getRecentOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const resolveProductName = require('util/resolveProductName');

module.exports = async (app, module, req, res) =>
{
  const {
    ProdShiftOrder
  } = module;

  const psos = await ProdShiftOrder
    .find({
      prodLine: req.params.line
    })
    .select({
      startedAt: 1,
      orderId: 1,
      'orderData.nc12': 1,
      'orderData.name': 1,
      'orderData.description': 1
    })
    .sort({
      prodLine: 1,
      startedAt: -1
    })
    .limit(10);

  const collection = [];
  const ids = new Set();

  psos.forEach(pso =>
  {
    if (ids.has(pso.orderId))
    {
      return;
    }

    ids.add(pso.orderId);

    collection.push({
      orderNo: pso.orderId,
      startedAt: pso.startedAt,
      productCode: pso.orderData && pso.orderData.nc12 || '',
      productName: resolveProductName(pso.orderData)
    });
  });

  res.json({
    totalCount: collection.length,
    collection
  });
};
