// wmes-gft-frontend/routes/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const axios = require('axios');

module.exports = (app, module) =>
{
  const {
    user,
    express,
    Order
  } = module;

  const canView = user.auth('USER');

  express.get('/gft/orders/:orderNo', canView, getOrderDataRoute);

  async function getOrderDataRoute(req, res)
  {
    const sapOrder = await Order
      .findById(req.params.orderNo)
      .select({
        nc12: 1,
        name: 1,
        description: 1
      })
      .lean()
      .exec();

    if (!sapOrder)
    {
      throw app.createError('Order not found.', 'NOT_FOUND', 404);
    }

    const data = {
      orderNo: req.params.orderNo,
      productFamily: '',
      productCode: sapOrder.nc12,
      productName: sapOrder.name,
      commercialDesignation: sapOrder.description,
      ltuneCode: '',
      driverType: '',
      numberOfDrivers: '',
      numberOfLeds: '',
      lampColor: '',
      pcbName: '',
      pcbCode: '',
      numberOfPcbs: 0
    };

    const orderRes = await axios.get(`http://ketrzyn.ipt.intra.lighting.com/api/order/${data.orderNo}`, {
      validateStatus: null,
      proxy: false
    });

    if (orderRes.status !== 200)
    {
      throw app.createError(`Invalid order response status: ${orderRes.status}`, 'IPT_FAILURE', 500);
    }

    orderRes.data.operations.forEach(op =>
    {
      if (op.operation === 'ProgramDeviceWithXMLFile')
      {
        if (op.parameters.numberOfDrivers)
        {
          data.numberOfDrivers = op.parameters.numberOfDrivers;
        }
      }
      else if (op.operation === 'PrintStandardLabel')
      {
        if (op.parameters.product.productFamily)
        {
          data.productFamily = op.parameters.product.productFamily;
        }

        if (op.parameters.product.driverType)
        {
          data.driverType = op.parameters.product.driverType;
          data.numberOfDrivers = data.driverType
            .split('_')
            .filter(part => !!part.length && part !== '0')
            .length;
        }

        if (op.parameters.technical.ltuneCode && op.parameters.technical.ltuneCode !== 'NA')
        {
          data.ltuneCode = op.parameters.technical.ltuneCode;
        }

        if (op.parameters.technical.numberOfLeds > 0)
        {
          data.numberOfLeds = op.parameters.technical.numberOfLeds;
        }

        if (op.parameters.technical.productFamily)
        {
          data.productFamily = op.parameters.technical.productFamily;
        }
      }
    });

    if (data.ltuneCode)
    {
      const itisRes = await axios.get(`http://import.ipt.intra.lighting.com/itis/${data.ltuneCode}`, {
        validateStatus: null,
        proxy: false
      });

      if (itisRes.status !== 200)
      {
        throw app.createError(`Invalid itis response status: ${itisRes.status}`, 'IPT_FAILURE', 500);
      }

      if (itisRes.data.data.LAMP_COLOUR)
      {
        data.lampColor = +itisRes.data.data.LAMP_COLOUR;
      }

      if (itisRes.data.data.DRIVERS_CODE_KEY)
      {
        data.driverType = itisRes.data.data.DRIVERS_CODE_KEY;
        data.numberOfDrivers = data.driverType
          .split(':')
          .filter(part => !!part.length && part !== '0').length;
      }
    }

    res.type('application/json');
    res.end(JSON.stringify(data, null, 2));
  }
};
