// wmes-osh/models/oshAccident.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const userInfoSchema = require('user/models/userInfoSchema');
const attachmentSchema = require('./attachmentSchema');
const userSchema = require('./userSchema');
const changeSchema = require('./changeSchema');
const decorateEntrySchema = require('./decorateEntrySchema');

exports.name = 'OshAccident';

exports.setUp = (app, mongoose) =>
{
  const STATUSES = ['finished'];

  const schema = new mongoose.Schema({
    rid: String,
    ridInc: Number,
    status: {
      type: String,
      enum: STATUSES
    },
    statusComment: {
      type: String,
      default: ''
    },
    statusUpdater: userInfoSchema,
    createdAt: {
      type: Date,
      required: true
    },
    updatedAt: {
      type: Date,
      required: true
    },
    creator: userInfoSchema,
    updater: userInfoSchema,
    finisher: userInfoSchema,
    coordinators: [userInfoSchema],
    division: {
      type: Number,
      ref: 'OshDivision',
      required: true
    },
    workplace: {
      type: Number,
      ref: 'OshWorkplace',
      required: true
    },
    department: {
      type: Number,
      ref: 'OshDepartment',
      required: true
    },
    building: {
      type: Number,
      ref: 'OshBuilding',
      default: 0
    },
    location: {
      type: Number,
      ref: 'OshLocation',
      default: 0
    },
    station: {
      type: Number,
      ref: 'OshStation',
      default: 0
    },
    eventDate: {
      type: Date,
      required: true
    },
    subject: {
      type: String,
      required: true,
      trim: true,
      maxLength: 150
    },
    description: {
      type: String,
      required: true,
      trim: true,
      maxLength: 1000
    },
    attachments: [attachmentSchema],
    users: [userSchema],
    changes: [changeSchema]
  }, {
    id: false,
    minimize: false
  });

  schema.statics.TOPIC_PREFIX = 'osh.accidents';
  schema.statics.PRIVILEGE_PREFIX = 'OSH:ACCIDENTS';
  schema.statics.RID_PREFIX = 'W';
  schema.statics.STATUSES = STATUSES;
  schema.statics.RELATION_TYPE = 'accident';
  schema.statics.RELATIONS_TYPE = 'accidents';

  decorateEntrySchema(app, mongoose, exports.name, schema);

  schema.methods.canDelete = function(userId)
  {
    return this.isCreator(userId)
      && (Date.now() - this.createdAt) < 3600000;
  };

  return schema;
};
