// wmes-osh/models/oshDivision.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const autoIncrement = require('mongoose-plugin-autoinc-fix');
const lastModifiedPaths = require('h5-mongoose/plugins/lastModifiedPaths');
const userInfoSchema = require('user/models/userInfoSchema');

exports.name = 'OshDivision';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    active: {
      type: Boolean,
      required: true
    },
    shortName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 50
    },
    longName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 200
    },
    workplace: {
      type: Number,
      ref: 'OshWorkplace',
      required: true
    },
    manager: userInfoSchema,
    coordinators: [userInfoSchema]
  }, {
    id: false,
    minimize: false
  });

  schema.virtual('buildings', {
    ref: 'OshBuilding',
    localField: '_id',
    foreignField: 'divisions'
  });

  schema.statics.TOPIC_PREFIX = 'osh.divisions';
  schema.statics.BROWSE_LIMIT = 1000;

  schema.plugin(autoIncrement.plugin, {
    model: exports.name,
    field: '_id',
    startAt: 1,
    incrementBy: 1
  });

  lastModifiedPaths(schema);

  return schema;
};
