// wmes-osh/models/oshDivision.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const autoIncrement = require('mongoose-plugin-autoinc-fix');

exports.name = 'OshDivision';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    active: {
      type: Boolean,
      required: true
    },
    shortName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 50
    },
    longName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 200
    },
    minObservers: {
      type: Number,
      min: 0,
      max: 100,
      default: 0
    }
  }, {
    id: false,
    minimize: false
  });

  schema.virtual('workplaces', {
    ref: 'OshWorkplace',
    localField: '_id',
    foreignField: 'division'
  });

  schema.statics.TOPIC_PREFIX = 'osh.divisions';
  schema.statics.BROWSE_LIMIT = 1000;

  schema.plugin(autoIncrement.plugin, {
    model: exports.name,
    field: '_id',
    startAt: 1,
    incrementBy: 1
  });

  return schema;
};
