// wmes-osh/models/oshEmployment.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'OshEmployment';

exports.setUp = (app, mongoose) =>
{
  const departmentSchema = new mongoose.Schema({
    division: {type: Number, required: true, min: 0, max: 9999},
    workplace: {type: Number, required: true, min: 0, max: 9999},
    department: {type: Number, required: true, min: 0, max: 9999},
    internal: {type: Number, required: true, min: 0, max: 99999},
    external: {type: Number, required: true, min: 0, max: 99999},
    absent: {type: Number, required: true, min: 0, max: 99999},
    total: {type: Number, required: true, min: 0, max: 99999},
    observers: {type: Number, required: true, min: 0, max: 99999}
  }, {
    _id: false,
    minimize: false
  });

  const schema = new mongoose.Schema({
    _id: Date,
    departments: [departmentSchema]
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  schema.statics.TOPIC_PREFIX = 'osh.employments';
  schema.statics.BROWSE_LIMIT = 100;

  schema.index({date: -1});

  return schema;
};
