// wmes-osh/models/oshKind.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const autoIncrement = require('mongoose-plugin-autoinc-fix');
const coordinatorsSchema = require('./coordinatorsSchema');

exports.name = 'OshKind';

exports.setUp = (app, mongoose) =>
{
  const TYPES = ['osh', 'env', 'inf', 'other'];

  const schema = new mongoose.Schema({
    active: {
      type: Boolean,
      required: true
    },
    type: {
      type: String,
      enum: TYPES,
      required: true
    },
    entryTypes: [{
      type: String,
      enum: ['nearMiss', 'kaizen', 'action']
    }],
    shortName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 50
    },
    longName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 200
    },
    description: {
      type: String,
      trim: true,
      maxLength: 1000
    },
    coordinators: [coordinatorsSchema]
  }, {
    id: false,
    minimize: false
  });

  schema.statics.TOPIC_PREFIX = 'osh.kinds';
  schema.statics.BROWSE_LIMIT = 1000;
  schema.statics.TYPES = TYPES;

  schema.plugin(autoIncrement.plugin, {
    model: exports.name,
    field: '_id',
    startAt: 1,
    incrementBy: 1
  });

  return schema;
};
