// wmes-osh/models/oshNearMiss.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const userInfoSchema = require('user/models/userInfoSchema');
const attachmentSchema = require('./attachmentSchema');
const userSchema = require('./userSchema');
const changeSchema = require('./changeSchema');
const relationSchema = require('./relationSchema');
const decorateEntrySchema = require('./decorateEntrySchema');

exports.name = 'OshNearMiss';

exports.setUp = (app, mongoose) =>
{
  const STATUSES = ['new', 'inProgress', 'finished', 'paused', 'cancelled'];

  const schema = new mongoose.Schema({
    rid: String,
    ridInc: Number,
    status: {
      type: String,
      enum: STATUSES
    },
    statusComment: {
      type: String,
      default: ''
    },
    statusUpdater: userInfoSchema,
    createdAt: {
      type: Date,
      required: true
    },
    updatedAt: {
      type: Date,
      required: true
    },
    startedAt: {
      type: Date,
      default: null
    },
    plannedAt: {
      type: Date,
      default: null
    },
    finishedAt: {
      type: Date,
      default: null
    },
    duration: {
      type: Number,
      min: 0
    },
    creator: userInfoSchema,
    updater: userInfoSchema,
    acceptor: userInfoSchema,
    finisher: userInfoSchema,
    implementer: userInfoSchema,
    coordinators: [userInfoSchema],
    division: {
      type: Number,
      ref: 'OshDivision',
      required: true
    },
    workplace: {
      type: Number,
      ref: 'OshWorkplace',
      required: true
    },
    department: {
      type: Number,
      ref: 'OshDepartment',
      required: true
    },
    building: {
      type: Number,
      ref: 'OshBuilding',
      required: true
    },
    location: {
      type: Number,
      ref: 'OshLocation',
      required: true
    },
    station: {
      type: Number,
      ref: 'OshStation',
      default: null
    },
    kind: {
      type: Number,
      ref: 'OshKind',
      required: true
    },
    eventDate: {
      type: Date,
      required: true
    },
    eventCategory: {
      type: Number,
      ref: 'OshEventCategory',
      required: true
    },
    reasonCategory: {
      type: Number,
      ref: 'OshReasonCategory'
    },
    priority: {
      type: Number,
      default: 1,
      min: 0,
      max: 4
    },
    materialLoss: {
      type: Boolean,
      default: false
    },
    subject: {
      type: String,
      required: true,
      trim: true,
      maxLength: 150
    },
    problem: {
      type: String,
      required: true,
      trim: true,
      maxLength: 1000
    },
    reason: {
      type: String,
      trim: true,
      maxLength: 1000
    },
    suggestion: {
      type: String,
      trim: true,
      maxLength: 1000
    },
    resolution: relationSchema,
    relations: [relationSchema],
    attachments: [attachmentSchema],
    users: [userSchema],
    changes: [changeSchema]
  }, {
    id: false,
    minimize: false
  });

  schema.statics.TOPIC_PREFIX = 'osh.nearMisses';
  schema.statics.PRIVILEGE_PREFIX = 'OSH:NEAR_MISSES';
  schema.statics.RID_PREFIX = 'Z';
  schema.statics.STATUSES = STATUSES;
  schema.statics.RELATION_TYPE = 'nearMiss';

  decorateEntrySchema(app, mongoose, exports.name, schema);

  return schema;
};
