// wmes-osh/models/oshObservationKind.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const autoIncrement = require('mongoose-plugin-autoinc-fix');

exports.name = 'OshObservationKind';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    active: {
      type: Boolean,
      required: true
    },
    shortName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 50
    },
    longName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 200
    },
    description: {
      type: String,
      trim: true,
      maxLength: 10000
    },
    behaviors: [{
      type: Number,
      ref: 'OshObservationCategory'
    }],
    workConditions: [{
      type: Number,
      ref: 'OshObservationCategory'
    }]
  }, {
    id: false,
    minimize: false
  });

  schema.statics.TOPIC_PREFIX = 'osh.observationKinds';
  schema.statics.BROWSE_LIMIT = 1000;

  schema.plugin(autoIncrement.plugin, {
    model: exports.name,
    field: '_id',
    startAt: 1,
    incrementBy: 1
  });

  return schema;
};
