// wmes-osh/models/oshPayout.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const userInfoSchema = require('user/models/userInfoSchema');

exports.name = 'OshPayout';

exports.setUp = (app, mongoose) =>
{
  const companySchema = new mongoose.Schema({
    id: String,
    label: String
  }, {
    _id: false
  });

  const recipientSchema = new mongoose.Schema({
    id: String,
    label: String,
    personnelId: String,
    company: {},
    department: String,
    count: {},
    amount: {},
    entries: [String]
  }, {
    _id: false
  });

  const schema = new mongoose.Schema({
    createdAt: Date,
    creator: userInfoSchema,
    types: [String],
    companies: [companySchema],
    description: String,
    options: {},
    recipients: [recipientSchema],
    count: {},
    amount: {}
  }, {
    id: false,
    minimize: false
  });

  schema.statics.TOPIC_PREFIX = 'osh.payouts';
  schema.statics.BROWSE_LIMIT = 100;

  schema.index({createdAt: -1});
  schema.index({types: 1, createdAt: -1});
  schema.index({'companies.id': 1, createdAt: -1});
  schema.index({'recipients.id': 1, createdAt: -1});
  schema.index({'recipients.entries': 1});

  return schema;
};
