// wmes-osh/models/oshWorkplace.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const autoIncrement = require('mongoose-plugin-autoinc-fix');
const lastModifiedPaths = require('h5-mongoose/plugins/lastModifiedPaths');

exports.name = 'OshWorkplace';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    active: {
      type: Boolean,
      required: true
    },
    shortName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 50
    },
    longName: {
      type: String,
      trim: true,
      required: true,
      maxLength: 200
    },
    division: {
      type: Number,
      ref: 'OshDivision',
      required: true
    }
  }, {
    id: false,
    minimize: false
  });

  schema.virtual('departments', {
    ref: 'OshDepartment',
    localField: '_id',
    foreignField: 'workplace'
  });

  schema.statics.TOPIC_PREFIX = 'osh.workplaces';
  schema.statics.BROWSE_LIMIT = 1000;

  schema.plugin(autoIncrement.plugin, {
    model: exports.name,
    field: '_id',
    startAt: 1,
    incrementBy: 1
  });

  lastModifiedPaths(schema);

  return schema;
};
