// h5-twilio/routes.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const {Twilio, webhook} = require('twilio');

module.exports = function setUpTwilioRoutes(app, twilioModule)
{
  const express = app[twilioModule.config.expressId];
  const mongoose = app[twilioModule.config.mongooseId];
  const TwilioRequest = mongoose.model('TwilioRequest');
  const TwilioResponse = mongoose.model('TwilioResponse');

  express.options('/twilio', function(req, res)
  {
    res.set('Access-Control-Allow-Origin', '*');
    res.set('Access-Control-Allow-Headers', 'Content-Type');
    res.end();
  });

  express.post('/twilio', function(req, res, next)
  {
    res.set('Access-Control-Allow-Origin', '*');
    res.set('Access-Control-Allow-Headers', 'Content-Type');

    const body = req.body;

    if (!_.isObject(body)
      || !_.isFunction(twilioModule[body.operation])
      || !_.isObject(body.options))
    {
      return next(app.createError('INVALID_INPUT', 400));
    }

    if (twilioModule.config.secretKey !== null && body.secretKey !== twilioModule.config.secretKey)
    {
      return next(app.createError('INVALID_SECRET_KEY', 401));
    }

    twilioModule[body.operation](body.options, function(err)
    {
      if (err)
      {
        twilioModule.error(err, 'Operation failed.', {
          operation: body.operation,
          options: body.options
        });

        return next(err);
      }

      twilioModule.info('Executed operation', {
        operation: body.operation,
        options: body.options
      });

      res.sendStatus(204);
    });
  });

  express.get('/twilio/:id', function(req, res, next)
  {
    TwilioRequest.findById(req.params.id).exec((err, twilioRequest) =>
    {
      if (err)
      {
        return next(err);
      }

      if (!twilioRequest)
      {
        return next(app.createError('NOT_FOUND', 400));
      }

      const twimlResponse = twilioRequest.operation === 'say'
        ? new Twilio.twiml.VoiceResponse()
        : new Twilio.twiml.MessagingResponse();

      twilioRequest.buildTwiml(twimlResponse);

      res.type('text/xml');
      res.send(twimlResponse.toString());

      TwilioRequest.updateStatus(twilioRequest._id, 'requested', (err) =>
      {
        if (err)
        {
          twilioModule.error(err, 'Failed to update a status of request [requested]', {requestId: twilioRequest._id});
        }
      });
    });
  });

  express.post('/twilio/:id', webhook(twilioModule.config.authToken), (req, res) =>
  {
    res.sendStatus(204);

    if (req.body.CallStatus)
    {
      TwilioRequest.updateStatus(req.params.id, req.body.CallStatus, (err) =>
      {
        if (err)
        {
          twilioModule.error(err, 'Failed to update a status of request.', {
            requestId: req.params.id,
            callStatus: req.body.CallStatus
          });
        }
      });
    }

    const twilioResponse = new TwilioResponse({
      request: req.params.id,
      createdAt: new Date(),
      payload: req.body
    });

    twilioResponse.save((err) =>
    {
      if (err)
      {
        twilioModule.error(err, 'Failed to save a response.', {requestId: req.params.id});
      }
    });
  });
};
