// wmes-osh/routes/rewards/export.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = async (app, module, req, res, next) =>
{
  const {
    express,
    OshReward
  } = module;

  const columns = {
    rid: 13,
    createdAt: 'datetime',
    creator: 20,
    subject: 40,
    recipient: 20,
    amount: 'decimal',
    paid: 'boolean',
    paidAt: 'datetime',
    payer: 40
  };

  req.rql.fields = {};

  Object.keys(columns).forEach(field => req.rql.fields[field] = 1);

  express.crud.exportRoute(app, {
    filename: 'WMES_OSH_REWARDS',
    freezeRows: 1,
    freezeColumns: 1,
    columns,
    model: OshReward,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    return {
      rid: doc.rid,
      createdAt: doc.createdAt,
      creator: doc.creator.label,
      subject: doc.subject,
      recipient: doc.recipient.label,
      amount: doc.amount,
      paid: doc.paid,
      paidAt: doc.paidAt,
      payer: doc.payer ? doc.payer.label : ''
    };
  }
};
