// settings/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user'
};

exports.models = [
  require('./models/setting')
];

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'express': require('./routes')
};

exports.start = (app, module) =>
{
  const {Setting} = module;

  module.findById = (_id, done) =>
  {
    module.find({_id: _id}, (err, docs) =>
    {
      done(err, Array.isArray(docs) && docs.length ? docs[0] : null);
    });
  };

  module.find = (conditions, done) =>
  {
    Setting.find(conditions, {value: 1}).lean().exec(done);
  };

  module.findValues = (conditions, ns, done) =>
  {
    if (typeof conditions === 'string' && typeof ns === 'function')
    {
      done = ns;
      ns = conditions;
      conditions = {_id: new RegExp('^' + _.escapeRegExp(ns))};
    }

    module.find(conditions, (err, settings) =>
    {
      if (err)
      {
        return done(err);
      }

      const values = {};

      settings.forEach(setting =>
      {
        values[setting._id.replace(ns, '')] = setting.value;
      });

      done(null, values);
    });
  };

  module.update = (_id, newValue, updater, done) =>
  {
    const updatedAt = new Date();
    const update = {
      $set: {
        value: newValue,
        updater: updater,
        updatedAt: updatedAt
      }
    };

    Setting.updateOne({_id}, update, {upsert: true}, function(err)
    {
      if (err)
      {
        module.error(err, `Failed to update a setting.`, {setting: {_id, newValue, updater}});
      }
      else
      {
        app.broker.publish(`settings.updated.${_id}`, {
          _id,
          value: newValue,
          updater,
          updatedAt
        });
      }

      done(err);
    });
  };

  module.multiUpdate = (settings, updater, done) =>
  {
    step(
      function()
      {
        Object.keys(settings).forEach(_id =>
        {
          module.update(_id, settings[_id], updater, this.group());
        });
      },
      done
    );
  };

  module.updateRoute = (req, res, next) =>
  {
    const {_id, value} = req.body;

    if (typeof _id !== 'string' || _id.trim().length === 0)
    {
      return next(app.createError(`Invalid setting ID.`, 'INPUT', 400));
    }

    if (value === undefined)
    {
      return next(app.createError(`Invalid setting value.`, 'INPUT', 400));
    }

    const userModule = app[module.config.userId];
    const updater = userModule ? userModule.createUserInfo(req.session.user, req) : null;

    module.update(_id, value, updater, (err) =>
    {
      if (err)
      {
        next(err);
      }
      else
      {
        res.sendStatus(204);
      }
    });
  };

  module.browseRoute = (pattern) =>
  {
    return function(req, res, next)
    {
      const express = app[module.config.expressId];

      if (!express)
      {
        return next(app.createError('No express module.', 'NO_MODULE', 500));
      }

      if (/^[A-Za-z0-9.-]+$/.test(pattern))
      {
        pattern = `^${pattern}\\.`;
      }

      req.rql.selector = {
        name: 'regex',
        args: ['_id', pattern]
      };

      express.crud.browseRoute(app, Setting, req, res, next);
    };
  };
};
