// reports/routes/report9.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const moment = require('moment');
const helpers = require('./helpers');
const report9 = require('../report9');

module.exports = function report9Route(app, reportsModule, req, res, next)
{
  const settingsModule = app[reportsModule.config.settingsId];
  const options = {
    fromTime: moment().startOf('month').subtract(6, 'months').valueOf(),
    toTime: moment().startOf('month').valueOf(),
    settings: {}
  };

  step(
    function getSettingsStep()
    {
      settingsModule.findValues({_id: /^reports\.cag/}, 'reports.cag.', this.next());
    },
    function generateReportStep(err, settings)
    {
      if (err)
      {
        return this.skip(err);
      }

      options.settings = settings;

      helpers.generateReport(
        app,
        reportsModule,
        report9,
        '9',
        req.reportHash,
        options,
        this.next()
      );
    },
    function sendResultStep(err, reportJson)
    {
      if (err)
      {
        return next(err);
      }

      res.type('json');
      res.send(reportJson);
    }
  );
};
