// reports/routes/report1.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const helpers = require('./helpers');
const report1 = require('../report1');

module.exports = (app, reportsModule, req, res, next) =>
{
  const orgUnitsModule = app[reportsModule.config.orgUnitsId];
  const prodTasksModule = app[reportsModule.config.prodTasksId];
  const downtimeReasonsModule = app[reportsModule.config.downtimeReasonsId];

  if (!orgUnitsModule)
  {
    return next(app.createError('No required module: orgUnits', 'MODULE', 500));
  }

  if (!prodTasksModule)
  {
    return next(app.createError('No required module: prodTasks', 'MODULE', 500));
  }

  if (!downtimeReasonsModule)
  {
    return next(app.createError('No required module: downtimeReasons', 'MODULE', 500));
  }

  const query = req.query;
  const orgUnit = orgUnitsModule.getByTypeAndId(query.orgUnitType, query.orgUnitId);

  if (orgUnit === null && (query.orgUnitType || query.orgUnitId))
  {
    return res.sendStatus(400);
  }

  const division = orgUnit ? orgUnitsModule.getDivisionFor(orgUnit) : null;

  if (orgUnit !== null && !division)
  {
    return res.sendStatus(400);
  }

  let subdivisions = orgUnit ? orgUnitsModule.getSubdivisionsFor(orgUnit) : null;
  let subdivisionType = query.subdivisionType;

  if (subdivisionType === 'prod' || subdivisionType === 'press')
  {
    subdivisions = filterSubdivisionByType(
      orgUnitsModule, subdivisions, subdivisionType === 'press' ? 'press' : 'assembly'
    );
  }
  else
  {
    subdivisionType = null;
  }

  const subdivisionTypes = {};

  _.forEach(orgUnitsModule.getAllByType('subdivision'), function(subdivision)
  {
    subdivisionTypes[subdivision._id] = subdivision.type;
  });

  const options = {
    fromTime: helpers.getTime(query.from),
    toTime: helpers.getTime(query.to),
    interval: helpers.getInterval(query.interval, 'hour'),
    orgUnitType: orgUnit ? query.orgUnitType : null,
    orgUnitId: orgUnit ? query.orgUnitId : null,
    division: helpers.idToStr(division),
    subdivisions: helpers.idToStr(subdivisions),
    subdivisionType: subdivisionType,
    subdivisionTypes: subdivisionTypes,
    prodFlows: helpers.idToStr(orgUnitsModule.getProdFlowsFor(orgUnit)),
    prodTasks: helpers.getProdTasksWithTags(prodTasksModule.models),
    orgUnits: helpers.getOrgUnitsForFte(orgUnitsModule, query.orgUnitType, orgUnit),
    downtimeReasons: getDowntimeReasons(downtimeReasonsModule.models),
    prodNumConstant: reportsModule.prodNumConstant,
    ignoredOrgUnits: helpers.decodeOrgUnits(orgUnitsModule, query.ignoredOrgUnits)
  };

  if (isNaN(options.fromTime) || isNaN(options.toTime))
  {
    return next(new Error('INVALID_TIME'));
  }

  helpers.generateReport(app, reportsModule, report1, '1', req.reportHash, options, function(err, reportJson)
  {
    if (err)
    {
      return next(err);
    }

    res.type('json');
    res.send(reportJson);
  });
};

function filterSubdivisionByType(orgUnitsModule, subdivisions, subdivisionType)
{
  if (subdivisions === null)
  {
    subdivisions = orgUnitsModule.getAllByType('subdivision');
  }

  subdivisions = subdivisions.filter(function(subdivision)
  {
    return subdivision.type === subdivisionType;
  });

  return subdivisions;
}

function getDowntimeReasons(allDowntimeReasons)
{
  const downtimeReasons = {
    breaks: {},
    schedule: {}
  };

  _.forEach(allDowntimeReasons, function(downtimeReason)
  {
    if (downtimeReason.type === 'break')
    {
      downtimeReasons.breaks[downtimeReason._id] = true;
    }
    else
    {
      downtimeReasons.schedule[downtimeReason._id] = downtimeReason.scheduled;
    }
  });

  return downtimeReasons;
}
