// reports/dailyMrpCounter.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');
const {formatDate} = require('util/dateFormatter');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose'
};

exports.models = [
  require('./models/dailyMrpCount'),
  require('prodShiftOrders/models/prodShiftOrder')
];

exports.requiredModules = 'mongoose';

exports.start = (app, module) =>
{
  app.broker.subscribe('orgUnits.rebuilt', onAppStarted).setLimit(1);

  app.broker.subscribe('shiftChanged', newShift => setTimeout(onShiftChanged, 1337, newShift));

  app.broker.subscribe('prodShiftOrders.created.*', onPsoChanged);
  app.broker.subscribe('prodShiftOrders.updated.*', onPsoChanged);
  app.broker.subscribe('prodShiftOrders.deleted.*', onPsoChanged);

  function onAppStarted()
  {
    const localMoment = moment();

    if (localMoment.hours() < 14)
    {
      localMoment.subtract(1, 'days');
    }

    localMoment.startOf('day');

    const utcMoment = moment.utc(localMoment.format('YYYY-MM-DD'), 'YYYY-MM-DD');

    recount(utcMoment.toDate());
  }

  function onShiftChanged(newShift)
  {
    const localMoment = moment(newShift.date).startOf('day');

    if (newShift.no === 1)
    {
      localMoment.subtract(1, 'days');
    }

    const utcMoment = moment.utc(localMoment.format('YYYY-MM-DD'), 'YYYY-MM-DD');

    recount(utcMoment.toDate());
  }

  function onPsoChanged({_id})
  {
    module.ProdShiftOrder.findById(_id).select({date: 1}).lean().exec((err, pso) =>
    {
      if (err)
      {
        return module.error(err, 'Failed to find ProdShiftOrder.', {prodShiftOrder: _id});
      }

      if (!pso)
      {
        return module.warn('ProdShiftOrder does not exist.', {prodShiftOrder: _id});
      }

      const localMoment = moment(pso.date);
      const utcMoment = moment.utc(localMoment.format('YYYY-MM-DD'), 'YYYY-MM-DD');

      recount(utcMoment.toDate());
    });
  }

  function recount(date)
  {
    module.DailyMrpCount.recount(date, err =>
    {
      if (err)
      {
        module.error(err, `Failed to recount.`);
      }
      else
      {
        module.info(`Recounted ${formatDate(date)}.`);
      }
    });
  }
};
