// reports/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const businessDays = require('./businessDays');
const helpers = require('./routes/helpers');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  orgUnitsId: 'orgUnits',
  prodTasksId: 'prodTasks',
  downtimeReasonsId: 'downtimeReasons',
  prodFunctionsId: 'prodFunctions',
  morId: 'mor',
  messengerClientId: null,
  messengerType: 'request',
  javaBatik: null,
  reports: {},
  nc12ToCagsJsonPath: 'nc12_to_cags.json'
};

exports.userPrivileges = [
  'REPORTS:VIEW', 'REPORTS:MANAGE',
  'REPORTS:1:VIEW', 'REPORTS:2:VIEW', 'REPORTS:3:VIEW', 'REPORTS:4:VIEW',
  'REPORTS:5:VIEW', 'REPORTS:6:VIEW', 'REPORTS:7:VIEW', 'REPORTS:8:VIEW',
  'REPORTS:9:VIEW'
];

exports.models = [
  require('./models/clipOrderCache'),
  require('./models/clipOrderCount'),
  require('./models/nonBusinessDay'),
  require('user/models/user')
];

exports.optionalModules = {
  'mongoose user settings downtimeReasons orgUnits express': require('./routes')
};

exports.setUp = (app, module) =>
{
  Object.keys(module.config.reports).forEach(id =>
  {
    const {models} = module.config.reports[id];

    if (Array.isArray(models))
    {
      exports.models = exports.models.concat(models);
    }
  });
};

exports.start = (app, module) =>
{
  let totalCountCache = {};

  module.helpers = helpers;
  module.prodNumConstant = 8;
  module.mapReduceResults = {};

  module.getCachedTotalCount = (key) =>
  {
    return totalCountCache[key];
  };

  module.setCachedTotalCount = (key, totalCount) =>
  {
    totalCountCache[key] = totalCount;
  };

  module.generateResponse = (report, options, req, res, next) =>
  {
    helpers.generateReport(app, module, report, report.id, req.reportHash, options, (err, reportJson) =>
    {
      if (err)
      {
        return next(err);
      }

      res.type('json');
      res.send(reportJson);
    });
  };

  app.onModuleReady(module.config.mongooseId, () =>
  {
    businessDays.reload(app, module, {recount: 30});
  });

  app.onModuleReady(module.config.settingsId, () =>
  {
    app[module.config.settingsId].findById('reports.prodNumConstant.coeff', (err, setting) =>
    {
      if (err)
      {
        return module.error(err, 'Failed to find the prodNumConstant setting.');
      }

      if (setting)
      {
        const value = parseFloat(setting.value);

        if (!isNaN(value) && value > 0)
        {
          module.prodNumConstant = value;
        }
      }
    });
  });

  app.broker.subscribe('settings.updated.reports.prodNumConstant.coeff', (message) =>
  {
    if (typeof message.value === 'number' && message.value > 0)
    {
      module.prodNumConstant = message.value;

      helpers.clearCachedReports(['1', '2', 'clip', '4']);
    }
  });

  app.broker.subscribe('settings.updated.reports.lean.**', () =>
  {
    helpers.clearCachedReports('8');
  });

  app.broker.subscribe('settings.updated.reports.clip.**', () =>
  {
    helpers.clearCachedReports('clip');
  });

  app.broker.subscribe('warehouse.shiftMetrics.updated', () =>
  {
    helpers.clearCachedReports('6');
  });

  app.broker.subscribe('clipOrderCount.created', () =>
  {
    helpers.clearCachedReports(['2', '7']);
  });

  app.broker.subscribe('cags.*', () =>
  {
    helpers.clearCachedReports('9');
  });

  app.broker.subscribe('cagGroups.*', () =>
  {
    helpers.clearCachedReports('9');
  });

  app.broker.subscribe('cags.*.synced', () =>
  {
    helpers.clearCachedReports('9');
  });

  app.broker.subscribe('orders.synced', () =>
  {
    totalCountCache = {};
  });

  app.broker.subscribe('orders.updated.*', ({_id, change}) =>
  {
    if (change.newValues.change)
    {
      helpers.clearCachedReports('clip');
    }

    const $set = {};

    if (change.comment)
    {
      $set.comment = change.comment;
    }

    ['delayReason', 'delayComponent', 'm4'].forEach(prop =>
    {
      if (change.newValues[prop] !== undefined)
      {
        $set[prop] = change.newValues[prop];
      }
    });

    if (!Object.keys($set).length)
    {
      return;
    }

    app[module.config.mongooseId].model('ClipOrderCache').updateMany({'_id.no': _id}, {$set});
  });
};
