// qi/routes/countReport.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const countReport = require('../countReport');

module.exports = (app, {reports}, {query, reportHash}, res, next) =>
{
  const options = {
    fromTime: reports.helpers.getTime(query.from) || null,
    toTime: reports.helpers.getTime(query.to) || null,
    interval: reports.helpers.getInterval(query.interval),
    productFamilies: _.isEmpty(query.productFamilies) ? '' : query.productFamilies,
    kinds: _.isEmpty(query.kinds) ? [] : query.kinds.split(','),
    errorCategories: _.isEmpty(query.errorCategories) ? [] : query.errorCategories.split(','),
    faultCodes: _.isEmpty(query.faultCodes) ? [] : query.faultCodes.split(','),
    inspector: query.inspector || null
  };

  reports.helpers.generateReport(
    app,
    reports,
    countReport,
    'qi/count',
    reportHash,
    options,
    (err, reportJson) =>
    {
      if (err)
      {
        return next(err);
      }

      res.type('json');
      res.send(reportJson);
    }
  );
};
