// qi/routes/editResult.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');
const prepareAttachments = require('./prepareAttachments');

module.exports = (app, module, req, res, next) =>
{
  const {
    user,
    QiResult
  } = module;

  const sessionUser = req.session.user;
  const input = {
    comment: req.body.comment
  };

  if (user.isAllowedTo(sessionUser, 'QI:RESULTS:MANAGE'))
  {
    Object.assign(input, req.body);
    prepareAttachments(module.tmpAttachments, input);
  }
  else
  {
    const leader = user.isAllowedTo(sessionUser, [['FN:leader'], ['FN:wh']]);
    const inspector = user.isAllowedTo(sessionUser, 'QI:INSPECTOR') || leader;
    const specialist = user.isAllowedTo(sessionUser, 'QI:SPECIALIST');
    const master = user.isAllowedTo(sessionUser, 'FN:master');
    const qualityEngineer = user.isAllowedTo(sessionUser, 'FN:quality_engineer');

    if (inspector)
    {
      Object.assign(input, _.omit(req.body, [
        'errorCategory',
        'correctiveActions'
      ]));
      prepareAttachments(module.tmpAttachments, input);
    }

    if (specialist)
    {
      Object.assign(input, _.pick(req.body, [
        'errorCategory',
        'immediateActions',
        'rootCause',
        'correctiveActions'
      ]));
    }

    if (master || leader)
    {
      Object.assign(input, _.pick(req.body, [
        'immediateActions',
        'rootCause',
        'correctiveActions'
      ]));
    }

    if (qualityEngineer)
    {
      Object.assign(input, _.pick(req.body, [
        'standard'
      ]));
    }
  }

  const updater = user.createUserInfo(sessionUser, req);

  step(
    function findResultStep()
    {
      QiResult.findById(req.params.id).exec(this.next());
    },
    function applyChangesStep(err, qiResult)
    {
      if (err)
      {
        return this.skip(err);
      }

      if (!qiResult)
      {
        return this.skip(app.createError('NOT_FOUND', 404));
      }

      const nokOwner = qiResult.nokOwner && qiResult.nokOwner.id === sessionUser._id;
      const leader = qiResult.leader && qiResult.leader.id === sessionUser._id;
      const corrector = (qiResult.correctiveActions || []).some(
        action => action.who.some(who => who.id === sessionUser._id)
      );

      if (nokOwner || leader || corrector)
      {
        Object.assign(input, _.pick(req.body, [
          'correctiveActions'
        ]));
      }

      const changed = qiResult.applyChanges(input, updater);

      if (changed)
      {
        qiResult.save(this.next());
      }
    },
    function sendResponseStep(err, qiResult)
    {
      if (err)
      {
        return next(err);
      }

      if (qiResult)
      {
        res.json(qiResult);

        app.broker.publish('qi.results.edited', {
          model: qiResult,
          user: updater
        });
      }
      else
      {
        res.json({_id: req.params.id});
      }
    }
  );
};
