// qi/routes/sendAttachment.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const fs = require('fs');
const path = require('path');
const step = require('h5.step');
const contentDisposition = require('content-disposition');
const gm = require('gm');

module.exports = (app, module, req, res, next) =>
{
  const {QiResult} = module;

  const fields = {};
  const changeIndex = /^-?[0-9]+$/.test(req.query.change) ? parseInt(req.query.change, 10) : null;
  let attachmentProperty = req.params.attachment;

  if (/NOK$/.test(attachmentProperty))
  {
    attachmentProperty = 'nokFile';
  }
  else if (/OK$/.test(attachmentProperty))
  {
    attachmentProperty = 'okFile';
  }

  fields[attachmentProperty] = 1;

  if (changeIndex !== null)
  {
    fields.changes = 1;
  }

  step(
    function findResultStep()
    {
      QiResult.findById(req.params.result, fields).lean().exec(this.next());
    },
    function handleFindResultResultStep(err, qiResult)
    {
      if (err)
      {
        return this.skip(err);
      }

      if (!qiResult)
      {
        return this.skip(app.createError('NOT_FOUND', 404));
      }

      if (changeIndex === null)
      {
        this.attachment = qiResult[attachmentProperty];
      }
      else
      {
        const change = qiResult.changes[Math.abs(changeIndex)];

        if (change && change.data[attachmentProperty])
        {
          const isOld = changeIndex < 0 || Object.is(changeIndex, -0);

          this.attachment = change.data[attachmentProperty][isOld ? 0 : 1];
        }
      }

      if (!this.attachment)
      {
        return this.skip(app.createError('NOT_FOUND', 404));
      }

      if (!req.query.min
        || !/^image/.test(this.attachment.type)
        || this.attachment.size < 512000)
      {
        return this.skip(null, false);
      }
    },
    function findMinFileStep()
    {
      fs.stat(path.join(module.config.attachmentsDest, this.attachment.path + '.min.jpg'), this.next());
    },
    function createMinFileStep(err)
    {
      if (!err)
      {
        return this.skip(null, true);
      }

      gm(path.join(module.config.attachmentsDest, this.attachment.path))
        .resize(1200, null, '>')
        .write(path.join(module.config.attachmentsDest, this.attachment.path + '.min.jpg'), this.next());
    },
    function handleCreateMinFileResultStep(err)
    {
      if (!err)
      {
        return this.skip(null, true);
      }

      module.error(err, `Failed to generate a min image of [${attachmentProperty}] of [${req.params.result}].`);

      return this.skip(null, false);
    },
    function sendAttachmentStep(err, min)
    {
      if (err)
      {
        return next(err);
      }

      let fileType;
      let filePath;

      if (min)
      {
        fileType = 'image/jpeg';
        filePath = path.join(module.config.attachmentsDest, this.attachment.path + '.min.jpg');
      }
      else
      {
        fileType = this.attachment.type;
        filePath = path.join(module.config.attachmentsDest, this.attachment.path);
      }

      res.type(fileType);
      res.append('Content-Disposition', contentDisposition(this.attachment.name, {
        type: req.query.download ? 'attachment' : 'inline'
      }));
      res.sendFile(filePath);

      this.attachment = null;
    }
  );
};
