// qi/routes/todo.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const shifts = require('util/shifts');

module.exports = async (app, module, req, res) =>
{
  const {production, QiResult} = module;

  const currentShift = shifts.getCurrentShiftInfo(false);
  const twoWeeksAgo = currentShift.moment.clone().subtract(14, 'days').toDate();

  const lastLineResults = await QiResult
    .aggregate([
      {$match: {
        source: 'prod',
        inspectedAt: {$gte: twoWeeksAgo}
      }},
      {$group: {
        _id: '$line',
        count: {$sum: 1},
        ok: {$sum: {$cond: ['$ok', 1, 0]}},
        nok: {$sum: {$cond: ['$ok', 0, 1]}},
        lastAt: {$max: '$createdAt'},
        divisions: {$addToSet: '$division'}
      }},
      {$sort: {
        lastAt: 1
      }}
    ])
    .exec();

  const lines = {};

  lastLineResults.forEach(r =>
  {
    lines[r._id] = r;
  });

  const prodLineStates = await production.getProdLineStates();

  prodLineStates.forEach(pls =>
  {
    if (!lines[pls.prodLine._id])
    {
      lines[pls.prodLine._id] = {
        _id: pls.prodLine._id,
        count: 0,
        ok: 0,
        nok: 0,
        lastAt: null,
        divisions: []
      };
    }

    const line = lines[pls.prodLine._id];

    line.state = pls.state;
  });

  res.json(Object.values(lines));
};
