// qi/routes/uploadAttachments.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const fs = require('fs');
const _ = require('lodash');

module.exports = (app, qiModule, req, res) =>
{
  const attachments = {};

  _.forEach(req.files, function(file)
  {
    const id = file.filename.replace(/\..*?$/, '');

    qiModule.tmpAttachments[id] = {
      data: {
        _id: id,
        type: file.mimetype,
        path: file.filename,
        name: file.originalname,
        size: file.size,
        description: file.fieldname
      },
      timer: setTimeout(removeAttachmentFile, 30000, id, file.path)
    };

    attachments[file.fieldname] = id;
  });

  res.json(attachments);

  function removeAttachmentFile(id, filePath)
  {
    delete qiModule.tmpAttachments[id];

    fs.unlink(filePath + '.min.jpg', () => {});
    fs.unlink(filePath, function(err)
    {
      if (err && err.code !== 'ENOENT')
      {
        qiModule.error(err, `Failed to remove an unused attachment.`, {filePath});
      }
      else
      {
        qiModule.info('Removed an unused attachment.', {filePath});
      }
    });
  }
};
