// opinionSurveys/cleanup.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const exec = require('child_process').exec;
const path = require('path');
const _ = require('lodash');
const step = require('h5.step');
const fs = require('fs-extra');

module.exports = (app, module) =>
{
  const mongoose = app[module.config.mongooseId];
  const OpinionSurveyOmrResult = mongoose.model('OpinionSurveyOmrResult');

  let cleaningScanTemplates = false;

  app.broker.subscribe('opinionSurveys.surveys.edited', (message) =>
  {
    cleanupSurveys(message.model._id, false);
  });

  app.broker.subscribe('opinionSurveys.surveys.deleted', (message) =>
  {
    cleanupSurveys(message.model._id, true);
  });

  app.broker.subscribe('opinionSurveys.scanTemplates.edited', cleanupScanTemplates);
  app.broker.subscribe('opinionSurveys.scanTemplates.deleted', cleanupScanTemplates);

  app.broker.subscribe('opinionSurveys.responses.edited', updateOmrResults.bind(null, 'fixed'));
  app.broker.subscribe('opinionSurveys.responses.deleted', updateOmrResults.bind(null, 'ignored'));

  function cleanupSurveys(surveyId, custom)
  {
    fs.readdir(module.config.surveysPath, (err, files) => // eslint-disable-line handle-callback-err
    {
      (files || []).forEach(file =>
      {
        if (!custom && file.includes('.custom'))
        {
          return;
        }

        if (file.startsWith(surveyId))
        {
          fs.unlink(path.join(module.config.surveysPath, file), () => {});
        }
      });
    });
  }

  function cleanupScanTemplates()
  {
    if (cleaningScanTemplates)
    {
      return;
    }

    cleaningScanTemplates = true;

    step(
      function getUsedImagesStep()
      {
        app[module.config.mongooseId].model('OpinionSurveyScanTemplate').distinct('image').exec(this.next());
      },
      function mapUsedImagesStep(err, images)
      {
        if (err)
        {
          return this.skip(err);
        }

        this.usedImages = {};

        for (let i = 0; i < images.length; ++i)
        {
          this.usedImages[images[i] + '.jpg'] = true;
        }
      },
      function readDirStep()
      {
        fs.readdir(module.config.templatesPath, this.next());
      },
      function getStatsStep(err, files)
      {
        if (err)
        {
          return this.skip(err);
        }

        for (let i = 0; i < files.length; ++i)
        {
          const file = files[i];

          if (!this.usedImages[file])
          {
            filterScanTemplate(file, this.group());
          }
        }
      },
      function removeUnusedImagesStep(err, filePaths)
      {
        if (err)
        {
          return this.skip(err);
        }

        if (!Array.isArray(filePaths))
        {
          return;
        }

        for (let i = 0; i < filePaths.length; ++i)
        {
          const filePath = filePaths[i];

          if (filePath)
          {
            fs.unlink(filePaths[i], this.group());
          }
        }
      },
      function finalizeStep(err)
      {
        if (err)
        {
          module.error(err, 'Failed to cleanup scan templates.');
        }
      }
    );
  }

  function filterScanTemplate(file, done)
  {
    const filePath = path.join(module.config.templatesPath, file);

    fs.stat(filePath, function(err, stats)
    {
      if (err || !stats.isFile() || (Date.now() - stats.ctime.getTime()) < 3600000)
      {
        return done(null, null);
      }

      return done(null, filePath);
    });
  }

  function updateOmrResults(newStatus, message)
  {
    const response = message.model;

    step(
      function()
      {
        OpinionSurveyOmrResult.find({status: 'unrecognized', response: response._id}).exec(this.next());
      },
      function(err, omrResults)
      {
        if (err)
        {
          return this.skip(err);
        }

        for (let i = 0; i < omrResults.length; ++i)
        {
          omrResults[i].status = newStatus;
          omrResults[i].save(this.group());
        }
      },
      function(err, omrResults)
      {
        if (err)
        {
          module.error(
            err,
            `Failed to change status of OMR result to [${newStatus}] after response [${response._id}] update.`
          );
        }
        else
        {
          _.forEach(omrResults, function(omrResult)
          {
            app.broker.publish('opinionSurveys.omrResults.edited', {
              user: null,
              model: omrResult
            });

            cleanUpProcessingFiles(omrResult);
          });
        }
      }
    );
  }

  function cleanUpProcessingFiles(result)
  {
    const processingDirPath = path.join(module.config.processingPath, result._id);
    const fromInputFilePath = result.omrOutput
      ? path.join(processingDirPath, result.scanTemplate._id.toString(), 'input.jpg')
      : path.join(processingDirPath, result.inputFileName);
    const toInputFilePath = path.join(module.config.responsesPath, result._id + '.jpg');

    fs.move(fromInputFilePath, toInputFilePath, {overwrite: true}, function()
    {
      removeDir(processingDirPath);
    });
  }

  function removeDir(dir)
  {
    exec(process.platform === 'win32' ? `rmdir /S /Q "${dir}"` : `rm -rf "${dir}"`, () => {});
  }
};
