// paintShop/models/paintShopLoad.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const shifts = require('util/shifts');

exports.name = 'PaintShopLoad';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    _id: {
      ts: Date, // Timestamp
      c: Number // Counter
    },
    d: Number, // Duration,
    r: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'PaintShopLoadReason',
      default: null
    }
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  schema.statics.TOPIC_PREFIX = 'paintShop.load';
  schema.statics.BROWSE_LIMIT = 100;

  schema.index({'_id.ts': -1});
  schema.index({'_id.c': 1, '_id.ts': -1});

  schema.statics.getStats = function(counter, done)
  {
    const PaintShopLoad = this;
    const now = Date.now();
    const oneHour = 60 * 60 * 1000;
    const timeRanges = [
      {id: 'shift', from: shifts.getCurrentShiftInfo().startTime},
      {id: '10m', from: now - 10 * 60 * 1000},
      {id: '1h', from: now - oneHour},
      {id: '8h', from: now - 8 * oneHour},
      {id: '1d', from: now - 24 * oneHour},
      {id: '7d', from: now - 7 * 24 * oneHour},
      {id: '30d', from: now - 30 * 24 * oneHour}
    ];

    step(
      function()
      {
        const conditions = {};
        const sort = {};

        if (counter > 0)
        {
          conditions['_id.c'] = counter;
          sort['_id.c'] = 1;
        }

        sort['_id.ts'] = -1;

        PaintShopLoad
          .findOne(conditions)
          .sort(sort)
          .exec(this.group());

        timeRanges.forEach(timeRange =>
        {
          const $match = {
            '_id.ts': {$gte: new Date(timeRange.from)}
          };

          if (counter > 0)
          {
            $match['_id.c'] = counter;
          }

          const pipeline = [
            {$match},
            {$group: {
              _id: null,
              sum: {$sum: 1},
              avg: {$avg: '$d'},
              min: {$min: '$d'},
              max: {$max: '$d'}
            }}
          ];

          PaintShopLoad.aggregate(pipeline, this.group());
        });
      },
      function(err, results)
      {
        if (err)
        {
          return done(err);
        }

        const stats = {
          counter,
          last: results.shift()
        };

        timeRanges.forEach(timeRange =>
        {
          const stat = results.shift()[0] || {
            sum: 0,
            avg: 0,
            min: 0,
            max: 0
          };

          delete stat._id;

          stat.avg = Math.round(stat.avg);

          stats[timeRange.id] = stat;
        });

        done(null, stats);
      }
    );
  };

  schema.statics.updateLoad = function(counter, items, done)
  {
    const PaintShopLoad = this;

    step(
      function()
      {
        PaintShopLoad.collection.insertMany(items, {ordered: false}, this.next());
      },
      function(err)
      {
        if (err && err.code !== 11000)
        {
          return this.skip(err);
        }

        PaintShopLoad.getStats(counter, this.next());
      },
      function(err, stats)
      {
        if (err)
        {
          return done(err);
        }

        app.broker.publish('paintShop.load.updated', {
          counter,
          stats,
          items
        });

        done();
      }
    );
  };

  return schema;
};
