// paintShop/models/paintShopEvent.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'PaintShopEvent';

exports.setUp = (app, mongoose) =>
{
  const paintShopEventSchema = new mongoose.Schema({
    type: {
      type: String,
      required: true
    },
    time: {
      type: Date,
      required: true
    },
    user: {
      type: {},
      required: true
    },
    order: {
      type: String,
      default: null,
      ref: 'PaintShopOrder'
    },
    data: {}
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  paintShopEventSchema.statics.TOPIC_PREFIX = 'paintShop.events';
  paintShopEventSchema.statics.BROWSE_LIMIT = 100;

  paintShopEventSchema.index({time: -1});
  paintShopEventSchema.index({type: 1, time: -1});
  paintShopEventSchema.index({order: 1, time: -1});

  paintShopEventSchema.post('save', function(doc)
  {
    app.broker.publish('paintShop.events.saved', doc.toJSON());
  });

  paintShopEventSchema.statics.record = function(event)
  {
    (new this(event)).save(err =>
    {
      if (err)
      {
        app.error(err, 'Failed to record event.', {module: 'paintShop', event});
      }
    });
  };

  mongoose.model(exports.name, paintShopEventSchema);
};
