// prodShiftOrders/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const validateOverlappingOrders = require('./validateOverlappingOrders');
const addProdShiftOrder = require('./addProdShiftOrder');
const editProdShiftOrder = require('./editProdShiftOrder');
const deleteProdShiftOrder = require('./deleteProdShiftOrder');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  orgUnitsId: 'orgUnits',
  productionId: 'production',
  fteId: 'fte'
};

exports.models = [
  require('./models/prodShiftOrder')
];

exports.republishTopics = ['prodShiftOrders.**'];

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'express user orgUnits production': require('./routes'),
  'mongoose fte': require('./noOperationsFix')
};

exports.start = (app, module) =>
{
  module.validateOverlappingOrders = validateOverlappingOrders.bind(null, app, module);
  module.addProdShiftOrder = addProdShiftOrder.bind(null, app, module);
  module.editProdShiftOrder = editProdShiftOrder.bind(null, app, module);
  module.deleteProdShiftOrder = deleteProdShiftOrder.bind(null, app, module);
};
