// cags/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  planUploadPath: './'
};

exports.models = [
  require('./models/cag'),
  require('./models/cagGroup'),
  require('./models/cagPlan')
];

exports.republishTopics = [
  'cags.nc12.synced', 'cags.nc12.syncFailed',
  'cags.plan.synced', 'cags.plan.syncFailed'
];

exports.optionalModules = {
  'mongoose user express': require('./routes')
};

exports.start = (app, module) =>
{
  app.broker.subscribe('cagGroups.added', onCagGroupChanged);
  app.broker.subscribe('cagGroups.edited', onCagGroupChanged);

  function onCagGroupChanged(message)
  {
    const mongoose = app[module.config.mongooseId];

    if (!mongoose)
    {
      return;
    }

    const CagGroup = mongoose.model('CagGroup');
    const conditions = {
      _id: {
        $ne: message.model._id
      }
    };
    const update = {
      $pullAll: {
        cags: message.model.cags
      }
    };

    CagGroup.updateMany(conditions, update, err =>
    {
      if (err)
      {
        module.error(err, `Failed to update CAGs in groups.`);
      }
    });
  }
};
