// behaviorObsCards/routes.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const countReport = require('./countReport');

module.exports = (app, module) =>
{
  const {user, reports, express, BehaviorObsCard, KaizenBehaviour} = module;

  const canView = user.auth('LOCAL', 'USER');
  const canManage = user.auth('USER');

  express.get(
    '/behaviorObsCards',
    canView,
    prepareForBrowse,
    express.crud.browseRoute.bind(null, app, BehaviorObsCard)
  );

  express.get(
    '/behaviorObsCards;export.:format?',
    canView,
    prepareForBrowse,
    prepareForExport,
    express.crud.exportRoute.bind(null, app, {
      filename: 'WMES-BEHAVIOR_OBS',
      freezeRows: 1,
      freezeColumns: 1,
      columns: {
        rid: 10,
        date: 'date',
        shift: {type: 'integer', width: 5},
        observer: 30,
        superior: 30,
        section: 20,
        line: 20,
        position: 20,
        company: 20,
        kind: 15,
        option1: 15,
        option2: 10,
        text1: 35,
        text2: 35,
        category: 30
      },
      serializeRow: exportBehaviorObsCard,
      model: BehaviorObsCard
    })
  );

  express.get('/behaviorObsCards;rid', canView, findByRidRoute);

  express.post(
    '/behaviorObsCards',
    canView,
    prepareForAdd,
    express.crud.addRoute.bind(null, app, BehaviorObsCard)
  );

  express.get('/behaviorObsCards/:id', canView, express.crud.readRoute.bind(null, app, {
    model: BehaviorObsCard,
    idProperty: req => /^[0-9]+$/.test(req.params.id) ? 'rid' : '_id'
  }));

  express.put(
    '/behaviorObsCards/:id',
    canManage,
    prepareForEdit,
    express.crud.editRoute.bind(null, app, BehaviorObsCard)
  );

  express.delete('/behaviorObsCards/:id', canManage, express.crud.deleteRoute.bind(null, app, BehaviorObsCard));

  express.get(
    '/behaviorObsCards/reports/count',
    canView,
    reports.helpers.sendCachedReport.bind(null, 'behaviorObsCards/count'),
    countReportRoute
  );

  function prepareForBrowse(req, res, next)
  {
    if (!req.session || !req.session.user)
    {
      return next();
    }

    req.rql.selector.args.forEach(function(term)
    {
      if (term.name === 'lang')
      {
        req.params.lang = term.args[0];
      }

      if (term.name === 'eq' && term.args[0] === 'users' && term.args[1] === 'mine')
      {
        term.args[1] = req.session.user._id;
      }
    });

    next();
  }

  function prepareForExport(req, res, next)
  {
    req.dictionaries = {
      behaviors: new Map()
    };

    KaizenBehaviour.find().lean().exec((err, behaviors) =>
    {
      if (err)
      {
        return next(err);
      }

      if (!behaviors.length)
      {
        return next();
      }

      if (!req.params.lang)
      {
        req.params.lang = Object.keys(behaviors[0].lang)[0];
      }

      behaviors.forEach(behavior =>
      {
        const lang = behavior.lang[req.params.lang];

        if (lang)
        {
          req.dictionaries.behaviors.set(behavior._id, lang.name);
        }
      });

      next();
    });
  }

  function prepareForAdd(req, res, next)
  {
    const body = req.body;

    body.createdAt = new Date();
    body.creator = user.createUserInfo(req.session.user, req);
    body.creator.id = body.creator.id.toString();

    return next();
  }

  function prepareForEdit(req, res, next)
  {
    const body = req.body;

    body.updatedAt = new Date();
    body.updater = user.createUserInfo(req.session.user, req);
    body.updater.id = body.updater.id.toString();

    return next();
  }

  function findByRidRoute(req, res, next)
  {
    const rid = parseInt(req.query.rid, 10);

    if (isNaN(rid) || rid <= 0)
    {
      return res.sendStatus(400);
    }

    BehaviorObsCard.findOne({rid: rid}, {_id: 1}).lean().exec(function(err, doc)
    {
      if (err)
      {
        return next(err);
      }

      if (doc)
      {
        return res.json(doc._id);
      }

      return res.sendStatus(404);
    });
  }

  function exportBehaviorObsCard(doc, req)
  {
    const rows = [];
    const superior = doc.superior ? doc.superior.label : '';

    _.forEach(doc.observations, function(o)
    {
      rows.push({
        rid: doc.rid,
        date: doc.date,
        shift: doc.shift,
        observer: doc.observer.label,
        superior,
        section: doc.section,
        line: doc.line,
        position: doc.position,
        company: doc.companyName,
        kind: 'observation',
        option1: o.safe ? '' : o.easy ? 'easy' : 'hard',
        option2: o.safe ? 'safe' : 'risky',
        text1: o.safe ? '' : o.observation,
        text2: o.safe ? '' : o.cause,
        category: req.dictionaries.behaviors.has(o.id)
          ? req.dictionaries.behaviors.get(o.id)
          : o.behavior
      });
    });

    _.forEach(doc.risks, function(o)
    {
      rows.push({
        rid: doc.rid,
        date: doc.date,
        shift: doc.shift,
        observer: doc.observer.label,
        superior,
        section: doc.section,
        line: doc.line,
        position: doc.position,
        company: doc.companyName,
        kind: 'risk',
        option1: o.easy ? 'easy' : 'hard',
        option2: '',
        text1: o.risk,
        text2: o.cause,
        category: ''
      });
    });

    _.forEach(doc.difficulties, function(o)
    {
      rows.push({
        rid: doc.rid,
        date: doc.date,
        shift: doc.shift,
        observer: doc.observer.label,
        superior,
        section: doc.section,
        line: doc.line,
        position: doc.position,
        company: doc.companyName,
        kind: 'difficulty',
        option1: o.behavior ? 'behavior' : 'workConditions',
        option2: '',
        text1: o.problem,
        text2: o.solution,
        category: ''
      });
    });

    return rows;
  }

  function countReportRoute(req, res, next)
  {
    const query = req.query;
    const options = {
      fromTime: reports.helpers.getTime(query.from) || null,
      toTime: reports.helpers.getTime(query.to) || null,
      interval: reports.helpers.getInterval(query.interval, 'month'),
      sections: _.isEmpty(query.sections) ? [] : query.sections.split(','),
      observerSections: _.isEmpty(query.observerSections) ? [] : query.observerSections.split(','),
      superior: _.isEmpty(query.superior) ? null : query.superior,
      company: _.isEmpty(query.company) ? [] : query.company.split(','),
      shift: +query.shift || 0
    };

    reports.helpers.generateReport(
      app,
      reports,
      countReport,
      'behaviorObsCards/count',
      req.reportHash,
      options,
      function(err, reportJson)
      {
        if (err)
        {
          return next(err);
        }

        res.type('json');
        res.send(reportJson);
      }
    );
  }
};
