// mail/downloader.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const fs = require('fs');
const path = require('path');
const _ = require('lodash');

exports.DEFAULT_CONFIG = {
  savePath: '/var/tmp',
  matcher: (email) => true, // eslint-disable-line no-unused-vars
  timestamp: false
};

exports.start = (app, module) =>
{
  app.broker.subscribe('mail.received', (mail) =>
  {
    if (typeof module.config.matcher === 'function' && !module.config.matcher(mail))
    {
      return;
    }

    if (!Array.isArray(mail.attachments) || mail.attachments.length === 0)
    {
      return module.info(`Ignoring an e-mail without any attachments.`, {subject: mail.subject});
    }

    let timestamp = Math.round(Date.now() / 1000);

    if (mail.headers.date)
    {
      const time = new Date(mail.headers.date).getTime();

      if (!isNaN(time))
      {
        timestamp = Math.round(time / 1000);
      }
    }

    module.info(`Saving e-mail attachments.`, {
      count: mail.attachments.length,
      subject: mail.subject
    });

    _.forEach(mail.attachments, (attachment) =>
    {
      const savePath = createSavePath(timestamp, attachment.fileName);

      fs.writeFile(savePath, attachment.content, (err) =>
      {
        if (err)
        {
          module.error(err, `Failed to save an attachment.`, {fileName: attachment.fileName});
        }
        else
        {
          module.info(`Saved attachment.`, {fileName: attachment.fileName});
        }
      });
    });
  });

  function createSavePath(timestamp, fileName)
  {
    if (module.config.timestamp)
    {
      fileName = `${timestamp}@${fileName}`;
    }

    return path.resolve(module.config.savePath, fileName);
  }
};
