// wmes-osh/routes/payouts/export.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = async (app, module, req, res, next) =>
{
  const {
    express,
    OshPayout
  } = module;

  const columns = {
    createdAt: 'datetime',
    creator: 20,
    description: 30,
    recipient: 30,
    personnelId: 10,
    company: 20,
    department: 30,
    kaizenCount: 'integer',
    kaizenAmount: 'decimal',
    observationCount: 'integer',
    observationAmount: 'decimal',
    payoutId: 20
  };

  express.crud.exportRoute(app, {
    filename: 'WMES_OSH_PAYOUTS',
    freezeRows: 1,
    columns,
    model: OshPayout,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    const payoutId = doc._id.toString();

    return doc.recipients.map(recipient =>
    {
      return {
        createdAt: doc.createdAt,
        creator: doc.creator.label,
        description: doc.description,
        recipient: recipient.label,
        personnelId: recipient.personnelId,
        company: recipient.company.text,
        department: recipient.department,
        kaizenCount: recipient.count.kaizen,
        kaizenAmount: Math.round(recipient.amount.kaizen * 100) / 100,
        observationCount: recipient.count.observation,
        observationAmount: Math.round(recipient.amount.observation * 100) / 100,
        payoutId
      };
    });
  }
};
