// opinionSurveys/models/opinionSurveyResponse.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'OpinionSurveyResponse';

exports.setUp = (app, mongoose) =>
{
  const opinionSurveyAnswerSchema = new mongoose.Schema({
    question: {
      type: String,
      required: true,
      ref: 'OpinionSurveyQuestion'
    },
    answer: {
      type: String,
      required: true
    }
  }, {
    _id: false
  });

  const opinionSurveyResponseSchema = new mongoose.Schema({
    _id: {
      type: String,
      required: true
    },
    survey: {
      type: String,
      required: true,
      ref: 'OpinionSurvey'
    },
    createdAt: {
      type: Date,
      required: true
    },
    creator: {},
    comment: {
      type: String,
      default: ''
    },
    employer: {
      type: String,
      default: null,
      ref: 'OpinionSurveyEmployer'
    },
    division: {
      type: String,
      default: null,
      ref: 'OpinionSurveyDivision'
    },
    superior: {
      type: String,
      default: null
    },
    answers: [opinionSurveyAnswerSchema]
  }, {
    id: false
  });

  opinionSurveyResponseSchema.statics.TOPIC_PREFIX = 'opinionSurveys.responses';

  opinionSurveyResponseSchema.index({survey: 1});
  opinionSurveyResponseSchema.index({employer: 1});
  opinionSurveyResponseSchema.index({division: 1});
  opinionSurveyResponseSchema.index({superior: 1});

  mongoose.model(exports.name, opinionSurveyResponseSchema);
};
