// wmes-dummyPaint/models/dpOrder.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'DpOrder';

exports.setUp = (app, mongoose) =>
{
  const orderSchema = new mongoose.Schema({
    _id: String,
    nc12: String,
    name: String,
    leadingNo: String,
    leadingNc12: String,
    leadingName: String,
    salesNo: String,
    salesItem: String,
    paintSource: String,
    productFamily: String,
    paintFamily: String,
    paintCategory: String,
    paintCode: String,
    dummyNc12: String,
    paintNc12: String,
    paintName: String,
    changed: Boolean,
    error: String,
    stage: String,
    job: String,
    createdAt: Date,
    updatedAt: Date,
    updatedBy: {}
  }, {
    id: false,
    minimize: false
  });

  orderSchema.statics.TOPIC_PREFIX = 'dummyPaint.orders';
  orderSchema.statics.BROWSE_LIMIT = 100;

  orderSchema.index({job: 1});
  orderSchema.index({createdAt: 1});
  orderSchema.index({stage: 1, createdAt: 1});
  orderSchema.index({dummyNc12: 1, stage: 1, createdAt: 1});
  orderSchema.index({leadingNo: 1, job: 1});
  orderSchema.index({leadingNc12: 1, job: 1});
  orderSchema.index({salesNo: 1, job: 1});
  orderSchema.index({changed: 1, dummyNc12: 1, createdAt: 1});

  mongoose.model(exports.name, orderSchema);
};
