// wmes-dummyPaint/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  sapGuiId: 'sapGui',
  mailSenderId: 'mail/sender',
  baseUrl: null
};

exports.models = [
  require('./models/dpCode'),
  require('./models/dpFamily'),
  require('./models/dpJob'),
  require('./models/dpOrder'),
  require('./models/dpPaint'),
  require('./models/dpWorker'),
  require('user/models/user')
];

exports.userPrivileges = [
  'DUMMY_PAINT:VIEW',
  'DUMMY_PAINT:MANAGE',
  'DUMMY_PAINT:WORKER'
];

exports.republishTopics = [
  'dummyPaint.paints.added', 'dummyPaint.paints.edited', 'dummyPaint.paints.deleted',
  'dummyPaint.codes.added', 'dummyPaint.codes.edited', 'dummyPaint.codes.deleted',
  'dummyPaint.families.added', 'dummyPaint.families.edited', 'dummyPaint.families.deleted',
  'dummyPaint.workers.updated',
  'dummyPaint.jobs.updated',
  'dummyPaint.orders.updated'
];

exports.recordTopics = {
  debug: [
    'dummyPaint.paints.added', 'dummyPaint.paints.edited',
    'dummyPaint.codes.added', 'dummyPaint.codes.edited',
    'dummyPaint.families.added', 'dummyPaint.families.edited'
  ],
  warning: [
    'dummyPaint.paints.deleted',
    'dummyPaint.codes.deleted'
  ]
};

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'user express settings': require('./routes'),
  'settings sapGui': require('./jobs'),
  'mailSender': require('./notifier')
};

exports.start = (app, module) =>
{
  if (!module.config.baseUrl && app.options.baseUrl)
  {
    module.config.baseUrl = app.options.baseUrl;
  }

  module.DICTIONARIES = {
    codes: {model: module.DpCode},
    paints: {model: module.DpPaint, load: false},
    families: {model: module.DpFamily}
  };
};
