// prodTasks/models/prodTask.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'ProdTask';

exports.setUp = (app, mongoose) =>
{
  const prodTaskSchema = new mongoose.Schema({
    name: {
      type: String,
      required: true,
      trim: true
    },
    tags: [String],
    fteDiv: {
      type: Boolean,
      default: false
    },
    clipColor: {
      type: String,
      default: '#eeee00'
    },
    inProd: {
      type: Boolean,
      default: true
    },
    parent: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'ProdTask',
      default: null
    }
  }, {
    id: false
  });

  prodTaskSchema.index({tags: 1});

  prodTaskSchema.statics.TOPIC_PREFIX = 'prodTasks';
  prodTaskSchema.statics.BROWSE_LIMIT = 1000;

  prodTaskSchema.statics.getForSubdivision = function(subdivisionId, done)
  {
    const subdivision = app.subdivisions.modelsById[subdivisionId];

    if (!subdivision)
    {
      return done(null, []);
    }

    const subdivisionTags = subdivision.prodTaskTags;

    if (!Array.isArray(subdivisionTags) || !subdivisionTags.length)
    {
      return done(null, []);
    }

    this.find({tags: {$in: subdivisionTags}}, {name: 1, fteDiv: 1, parent: 1})
      .sort({name: 1})
      .lean()
      .exec(done);
  };

  mongoose.model(exports.name, prodTaskSchema);
};
