// wmes-luma2-frontend/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings'
};

exports.models = [
  require('wmes-luma2-backend/models/luma2Event'),
  require('wmes-luma2-backend/models/luma2Line')
];

exports.republishTopics = [
  'luma2.events.saved',
  'luma2.lines.added', 'luma2.lines.edited', 'luma2.lines.deleted'
];

exports.recordTopics = {
  debug: ['luma2.lines.added', 'luma2.lines.edited'],
  warning: ['luma2.lines.deleted']
};

exports.userPrivileges = [
  'LUMA2:VIEW', 'LUMA2:MANAGE'
];

exports.optionalModules = {
  'mongoose user settings express': require('./routes')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    lines: 'Luma2Line'
  };

  app.broker.subscribe('production.synced.*', onProductionSynced);

  function onProductionSynced(changes)
  {
    if (changes.types.includes('changeOrder'))
    {
      app.broker.publish('luma2.lines.orderChanged', {
        line: changes.prodLine
      });
    }
  }
};
