// wmes-toolcal/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const path = require('path');
const {ensureDir} = require('fs-extra');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  userId: 'user',
  settingsId: 'settings',
  expressId: 'express',
  mailSenderId: 'mail/sender',
  uploadsDest: './data/uploads/toolcal'
};

exports.models = [
  require('./models/toolCalType'),
  require('./models/toolCalTool'),
  require('user/models/user')
];

exports.republishTopics = ['toolcal.**'];

exports.recordTopics = {
  debug: ['toolcal.types.added', 'toolcal.types.edited'],
  warning: ['toolcal.*.deleted']
};

exports.userPrivileges = [
  'TOOLCAL:VIEW',
  'TOOLCAL:MANAGE',
  'TOOLCAL:DICTIONARIES:VIEW',
  'TOOLCAL:DICTIONARIES:MANAGE'
];

exports.optionalModules = {
  'mongoose user settings express': require('./routes'),
  'mongoose user settings mailSender': require('./notifier')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    types: 'ToolCalType'
  };

  module.tmpAttachments = new Map();

  module.config.uploadsDest = path.resolve(module.config.uploadsDest);

  ensureDir(module.config.uploadsDest, 0o775, () => {});
};
