// wmes-ct-backend/models/ctPce.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'CtPce';

exports.setUp = (app, mongoose) =>
{
  const pceSchema = new mongoose.Schema({
    line: {
      type: String,
      required: true
    },
    station: {
      type: Number,
      required: true,
      min: 1,
      max: 7
    },
    cart: String,
    order: {},
    pce: {
      type: Number,
      min: 1,
      required: true
    },
    startedAt: {
      type: Date,
      required: true
    },
    finishedAt: {
      type: Date,
      required: true
    },
    durations: {
      total: Number,
      work: Number,
      downtime: Number,
      scheduled: Number
    }
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  pceSchema.statics.TOPIC_PREFIX = 'ct.pces';
  pceSchema.statics.BROWSE_LIMIT = 100;

  pceSchema.index({startedAt: -1});
  pceSchema.index({'order._id': 1});
  pceSchema.index({'order.nc12': 1});
  pceSchema.index({'order.pso': 1});
  pceSchema.index({line: 1, station: 1});

  pceSchema.post('save', function()
  {
    app.broker.publish(`${pceSchema.statics.TOPIC_PREFIX}.saved`, this.toJSON());
  });

  mongoose.model(exports.name, pceSchema);
};
