// kaizen/models/kaizenSection.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const userInfoSchema = require('user/models/userInfoSchema');

exports.name = 'KaizenSection';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    _id: {
      type: String,
      required: true,
      match: /^[A-Za-z0-9-]+$/
    },
    active: {
      type: Boolean,
      default: true
    },
    entryTypes: {
      type: [String],
      enum: ['kaizenOrders', 'suggestions', 'observations', 'minutes', 'audits', 'talks']
    },
    name: {
      type: String,
      required: true,
      trim: true
    },
    position: {
      type: Number,
      default: 0
    },
    subdivisions: [mongoose.Schema.Types.ObjectId],
    confirmers: [userInfoSchema],
    coordinators: [userInfoSchema],
    auditors: [userInfoSchema],
    controlLists: [mongoose.Schema.Types.ObjectId]
  }, {
    id: false
  });

  schema.statics.TOPIC_PREFIX = 'kaizen.sections';
  schema.statics.BROWSE_LIMIT = 1000;

  return schema;
};
