// kaizen/models/kaizenCategory.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'KaizenCategory';

exports.setUp = (app, mongoose) =>
{
  const coordSectionSchema = new mongoose.Schema({
    section: String,
    funcs: [String],
    mor: {
      type: String,
      enum: ['none', 'all', 'mrp']
    }
  }, {
    _id: false,
    minimize: false
  });

  const kaizenCategorySchema = new mongoose.Schema({
    _id: {
      type: String,
      required: true,
      match: /^[A-Za-z0-9-]+$/
    },
    active: {
      type: Boolean,
      default: true
    },
    name: {
      type: String,
      required: true,
      trim: true
    },
    description: {
      type: String,
      trim: true,
      default: ''
    },
    inNearMiss: {
      type: Boolean,
      required: true,
      default: false
    },
    inSuggestion: {
      type: Boolean,
      required: true,
      default: false
    },
    position: {
      type: Number,
      default: 0
    },
    coordSections: [coordSectionSchema]
  }, {
    id: false
  });

  kaizenCategorySchema.statics.TOPIC_PREFIX = 'kaizen.categories';
  kaizenCategorySchema.statics.BROWSE_LIMIT = 1000;

  mongoose.model(exports.name, kaizenCategorySchema);
};
