// kaizen/models/kaizenProductFamily.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const userInfoSchema = require('user/models/userInfoSchema');

exports.name = 'KaizenProductFamily';

exports.setUp = (app, mongoose) =>
{
  const coordSectionSchema = new mongoose.Schema({
    section: String,
    filterSections: [String],
    excludeSections: Boolean,
    coordinators: Boolean,
    funcs: [String],
    mor: {
      type: String,
      enum: ['none', 'all', 'mrp']
    },
    users: [userInfoSchema]
  }, {
    _id: false,
    minimize: false
  });

  const productFamilySchema = new mongoose.Schema({
    _id: {
      type: String,
      required: true,
      match: /^[A-Za-z0-9-]+$/
    },
    name: {
      type: String,
      required: true,
      trim: true
    },
    position: {
      type: Number,
      default: 0
    },
    active: {
      type: Boolean,
      default: true
    },
    mrps: [String],
    coordSections: [coordSectionSchema]
  }, {
    id: false,
    minimize: false
  });

  productFamilySchema.statics.TOPIC_PREFIX = 'kaizen.productFamilies';
  productFamilySchema.statics.BROWSE_LIMIT = 1000;

  mongoose.model(exports.name, productFamilySchema);
};
