// wh/state/cancelOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, user, session}) =>
{
  const {
    OldWhEvent: WhEvent,
    OldWhOrder: WhOrder,
    OldWhSetCart: WhSetCart,
    OldWhLine: WhLine
  } = module;

  const conditions = {
    status: {$ne: 'cancelled'}
  };

  if (data.orders)
  {
    if (!Array.isArray(data.orders) || data.orders.length !== 1)
    {
      throw app.createError('Invalid orders.', 'INPUT', 400);
    }

    conditions._id = {$in: data.orders};
  }
  else
  {
    conditions.set = data.set;
    conditions.date = data.date;
  }

  const whOrders = await WhOrder
    .find(conditions)
    .session(session)
    .exec();

  if (!whOrders.length)
  {
    return;
  }

  const time = new Date();
  const $set = {
    status: 'cancelled',
    finishedAt: time
  };
  const events = [];

  for (const whOrder of whOrders)
  {
    Object.assign(whOrder, $set);

    events.push({
      type: 'orderCancelled',
      time,
      user,
      order: whOrder._id,
      data: {
        date: data.date,
        orders: data.orders || null,
        set: data.set || null
      }
    });
  }

  const results = [];

  const [eventsResults] = await Promise.all([
    WhEvent.record(events, session),
    WhOrder.collection.updateMany({_id: {$in: whOrders.map(o => o._id)}}, {$set}, {session})
  ]);

  results.push(eventsResults);

  if (data.orders)
  {
    const setCartsResults = await WhSetCart.deleteByWhOrder(whOrders[0]);
    const [lineResults, setResults] = await Promise.all([
      WhLine.recountPickupByWhOrder(whOrders[0]),
      whOrders[0].finalizeSet(setCartsResults)
    ]);

    results.push(setCartsResults, lineResults, setResults);
  }
  else
  {
    const [setCartsResults, lineResults] = await Promise.all([
      WhSetCart.deleteBySet(data.date, data.set, session),
      WhLine.recountPickupByWhOrder(whOrders[0])
    ]);

    results.push(setCartsResults, lineResults);
  }

  await session.commitTransaction();

  const leanWhOrders = whOrders.map(whOrder => whOrder.toJSON());

  return {
    response: {
      orders: leanWhOrders
    },
    results,
    messages: [{
      topic: `${WhOrder.TOPIC_PREFIX}.updated`,
      message: {updated: leanWhOrders}
    }]
  };
};
