// wh/state/editDeliveredOrder.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, user, session}) =>
{
  const {
    OldWhDeliveredOrder: WhDeliveredOrder,
    OldWhLine: WhLine,
    OldWhEvent: WhEvent
  } = module;

  const deliveredOrder = await WhDeliveredOrder
    .findById(data._id)
    .lean()
    .session(session)
    .exec();

  if (!deliveredOrder)
  {
    throw app.createError('Order not found.', 'INPUT', 400);
  }

  const qtyTodo = Math.max(1, Math.floor(data.qtyTodo));
  const qtyDone = Math.min(qtyTodo, Math.max(0, Math.floor(data.qtyDone)));
  const update = {
    _id: deliveredOrder._id,
    qtyDone,
    qtyTodo,
    qtyRemaining: qtyTodo - qtyDone
  };

  await WhLine.lockLines([deliveredOrder.line]);

  await WhDeliveredOrder.collection.updateOne({_id: deliveredOrder._id}, {$set: update}, {session});

  const event = {
    type: 'deliveredOrderEdited',
    user,
    time: new Date(),
    data: {
      orders: [{
        whOrder: deliveredOrder.whOrder,
        sapOrder: deliveredOrder.sapOrder,
        date: deliveredOrder.date,
        set: deliveredOrder.set
      }],
      lines: [deliveredOrder.line],
      oldQtyDone: deliveredOrder.qtyDone,
      oldQtyTodo: deliveredOrder.qtyTodo,
      newQtyDone: qtyDone,
      newQtyTodo: qtyTodo
    }
  };

  const [lineResult, eventsResult] = await Promise.all([
    WhLine.recount(deliveredOrder.line, true, true, session),
    WhEvent.record(event, session)
  ]);

  await session.commitTransaction();

  return {
    response: update,
    results: [lineResult, eventsResult],
    messages: [{
      topic: `${WhDeliveredOrder.TOPIC_PREFIX}.updated`,
      message: {updated: [update]}
    }]
  };
};
