// wh/state/editDeliveredOrder.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, session}) =>
{
  const {
    OldWhDeliveredOrder: WhDeliveredOrder,
    OldWhLine: WhLine
  } = module;

  const deliveredOrder = await WhDeliveredOrder
    .findById(data._id)
    .lean()
    .session(session)
    .exec();

  if (!deliveredOrder)
  {
    throw app.createError('Order not found.', 'INPUT', 400);
  }

  const qtyTodo = Math.max(1, Math.floor(data.qtyTodo));
  const qtyDone = Math.min(qtyTodo, Math.max(0, Math.floor(data.qtyDone)));
  const update = {
    _id: deliveredOrder._id,
    qtyDone,
    qtyTodo,
    qtyRemaining: qtyTodo - qtyDone
  };

  await WhLine.lockLines([deliveredOrder.line]);

  await WhDeliveredOrder.collection.updateOne({_id: deliveredOrder._id}, {$set: update}, {session});

  const lineResult = await WhLine.recount(deliveredOrder.line, true, true, session);

  await session.commitTransaction();

  return {
    response: update,
    results: [lineResult],
    messages: [{
      topic: `${WhDeliveredOrder.TOPIC_PREFIX}.updated`,
      message: {updated: [update]}
    }]
  };
};
