// wh/state/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

/* eslint-disable curly */

'use strict';

const _ = require('lodash');
const {ObjectId} = require('mongoose').Types;
const locks = require('util/locks');
const actions = require('./actions');
const printLabels = require('./printLabels');
const printPsPicklist = require('./printPsPicklist');
const handleSerialNumber = require('./handleSerialNumber');
const handleFinishOrder = require('./handleFinishOrder');
const createPickupStatus = require('./pickupStatus');
const updateStartedPlan = require('./updateStartedPlan');
const updatePlanInfo = require('./updatePlanInfo');

module.exports = (app, module) =>
{
  if (module.config.generator)
  {
    return;
  }

  const {
    settings,
    OldWhOrder: WhOrder,
    OldWhLine: WhLine,
    OldWhLineSnapshot: WhLineSnapshot
  } = module;

  const doUpdateStartedPlan = updateStartedPlan.bind(null, app, module);
  const doUpdatePlanInfo = updatePlanInfo.bind(null, app, module);
  const scheduleNextPlanInfoUpdate = _.debounce(doUpdatePlanInfo, 15000);

  let releases = [];
  let queue = null;
  const recountingLines = new Set();
  const workingLines = new Map();

  app.broker.subscribe('app.started', onAppStarted).setLimit(1);
  app.broker.subscribe('settings.updated.wh.planning.**', onSettingUpdated);
  app.broker.subscribe('production.taktTime.snChecked.**', onSnChecked);
  app.broker.subscribe('production.synced.**', onProductionSynced);
  app.broker.subscribe('production.stateChanged.**', onProductionStateChanged);
  app.broker.subscribe('orders.updated.*', onOrderUpdated);
  app.broker.subscribe('planning.changes.created', onPlanChanged);
  app.broker.subscribe('shiftChanged', onShiftChanged);
  app.broker.subscribe(`${WhLine}.updated`, onLineUpdated);

  module.state = {
    settings: {},
    act: (input, done) =>
    {
      if (!actions[input.action])
      {
        return done(app.createError(`Unknown action.`, 'UNKNOWN_ACTION', 400, {input}));
      }

      if (queue !== null)
      {
        queue.push({input, done});

        return;
      }

      queue = [{input, done}];

      setImmediate(executeNextAction);
    },
    lock,
    printLabels: printLabels.bind(null, app, module),
    printPsPicklist: printPsPicklist.bind(null, app, module),
    scheduleLineRecount,
    abortTransaction,
    pickupStatus: createPickupStatus(app, module),
    getLineWorkStatus: lineId => workingLines.get(lineId)
  };

  async function executeNextAction()
  {
    const next = queue.shift();
    const action = actions[next.input.action];
    const input = {
      data: {},
      ...next.input
    };

    if (typeof input.data.date === 'string'
      && /^[0-9]{4}-[0-9]{2}-[0-9]{2}$/.test(input.data.date))
    {
      input.data.date = new Date(`${input.data.date}T00:00:00Z`);
    }

    const timeout = setTimeout(handleActionTimeout, 5000, input);

    if (action.detached)
    {
      if (queue.length)
      {
        setImmediate(executeNextAction);
      }
      else
      {
        queue = null;
      }
    }

    try
    {
      if (action.session)
      {
        input.session = await WhOrder.startSession();
        input.session.startTransaction();
      }

      const result = {
        response: {},
        messages: [],
        results: [],
        ...(await action.execute(app, module, input))
      };

      result.messages.forEach(({topic, message, meta}) =>
      {
        app.broker.publish(topic, message, meta);
      });

      _.flatten(result.results).forEach(subResult =>
      {
        if (!subResult)
        {
          return;
        }

        if (Array.isArray(subResult.messages))
        {
          subResult.messages.forEach(({topic, message, meta}) =>
          {
            app.broker.publish(topic, message, meta);
          });
        }
      });

      next.done(null, result.response);
    }
    catch (err)
    {
      next.done(err);

      if (input.session)
      {
        await abortTransaction(input.session);
      }
    }
    finally
    {
      clearTimeout(timeout);

      if (input.session)
      {
        input.session.endSession();
        input.session = null;
      }

      if (releases.length)
      {
        releases.forEach(release => release());
        releases = [];
      }
    }

    if (!action.detached)
    {
      if (queue.length)
      {
        setImmediate(executeNextAction);
      }
      else
      {
        queue = null;
      }
    }
  }

  function handleActionTimeout(input)
  {
    module.warn(`Action timeout.`, {
      input,
      queueLength: queue ? queue.length : 0
    });
  }

  async function onAppStarted()
  {
    try
    {
      module.state.settings = await settings.findValues('wh.planning.');
    }
    catch (err)
    {
      module.error(err, 'Failed to load settings.');
    }

    module.state.pickupStatus.updateNow();
    scheduleNextPlanInfoUpdate();
    snapshotLines();
  }

  function onSettingUpdated({_id, value})
  {
    module.state.settings[_id.replace('wh.planning.', '')] = value;
  }

  function onSnChecked({serialNumber})
  {
    handleSerialNumber(app, module, serialNumber);
  }

  async function abortTransaction(session)
  {
    if (session)
    {
      try
      {
        await session.abortTransaction();
      }
      catch (err)
      {
        if (err.message !== 'Cannot call abortTransaction twice')
        {
          module.error(err, `Failed to abort a transaction.`);
        }
      }
    }
  }

  function toggleLineWorkStatus(lineId, working)
  {
    if (typeof working !== 'boolean')
    {
      const production = app[module.config.productionId];
      const lineState = production && production.getProdLineState(lineId);

      working = !!lineState && !!lineState.getCurrentOrder();
    }

    if (workingLines.get(lineId) === working)
    {
      return;
    }

    workingLines.set(lineId, working);

    app.broker.publish(`${WhLine.TOPIC_PREFIX}.updated`, {
      updated: [{
        _id: lineId,
        working
      }]
    });
  }

  function onProductionSynced(message)
  {
    const {types, prodLine, finishedPso} = message;

    if (Array.isArray(finishedPso) && finishedPso.length)
    {
      handleFinishOrder(app, module, prodLine, finishedPso);
    }

    if (types.includes('changeOrder') || types.includes('endWork'))
    {
      doUpdateStartedPlan(prodLine);
    }
  }

  function onProductionStateChanged(changes)
  {
    if (changes.online || changes.prodShiftOrders !== undefined)
    {
      toggleLineWorkStatus(changes._id);
    }
  }

  function onOrderUpdated(message)
  {
    const psStatus = message.change.newValues.psStatus;

    if (!psStatus)
    {
      return;
    }

    const input = {
      action: 'updatePsStatus',
      data: {
        orderNo: message._id,
        psStatus
      }
    };

    module.state.act(input, err =>
    {
      if (err)
      {
        module.error(err, 'Failed to update PS status.', {input});
      }
    });
  }

  async function lock(key)
  {
    return new Promise((resolve) =>
    {
      const release = locks.create(key, () =>
      {
        resolve(release);
      });

      releases.push(release);
    });
  }

  function scheduleLineRecount(lineId)
  {
    if (recountingLines.has(lineId))
    {
      return;
    }

    recountingLines.add(lineId);

    setTimeout(recountLine, 60000, lineId);
  }

  async function recountLine(lineId)
  {
    const releaseLock = await locks.createAsync(`${WhLine.TOPIC_PREFIX}.${lineId}`);

    try
    {
      const {messages} = await WhLine.recountAvailable(lineId, null);

      messages.forEach(({topic, message, meta}) =>
      {
        app.broker.publish(topic, message, meta);
      });

      recountingLines.delete(lineId);
    }
    catch (err)
    {
      module.error(err, `Failed to recount line.`, {lineId});

      recountingLines.delete(lineId);

      scheduleLineRecount(lineId);
    }

    releaseLock();
  }

  function onPlanChanged()
  {
    scheduleNextPlanInfoUpdate();
  }

  function onShiftChanged()
  {
    doUpdatePlanInfo();
  }

  function onLineUpdated({added})
  {
    if (added && added.length)
    {
      doUpdatePlanInfo();

      added.forEach(line => doUpdateStartedPlan(line._id));
    }
  }

  async function snapshotLines()
  {
    try
    {
      const time = new Date();
      const lines = await WhLine.find({}).lean().exec();
      const ops = lines.map(line => ({
        insertOne: {
          document: {
            _id: new ObjectId(),
            line: line._id,
            time,
            data: line
          }
        }
      }));

      ops.push({
        deleteMany: {
          filter: {
            time: {$lt: new Date(time.getTime() - 14 * 24 * 3600 * 1000)}
          }
        }
      });

      await WhLineSnapshot.collection.bulkWrite(ops);

      module.debug(`Snapshotted lines.`);

      setTimeout(snapshotLines, Math.floor(5 * 60 * 1000 - (time.getTime() % (5 * 60 * 1000))));
    }
    catch (err)
    {
      module.error(err, `Failed to snapshot lines.`);
    }
  }
};
