// wh/state/pickup.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, user, session}) =>
{
  const {
    OldWhEvent: WhEvent,
    OldWhOrder: WhOrder,
    OldWhSetCart: WhSetCart
  } = module;

  const whOrder = await WhOrder.findById(data.whOrderId).session(session).exec();

  if (!whOrder)
  {
    throw app.createError(`Order not found.`, 'INPUT', 400);
  }

  const func = whOrder.funcs.find(f => f._id === data.funcId);

  if (!func)
  {
    throw app.createError(`Function not found.`, 'INPUT', 400);
  }

  const newValue = data.newValue;
  const oldValue = func.pickup;
  const pickupEvent = {
    type: 'pickup',
    user,
    time: new Date(),
    order: whOrder._id,
    data: {
      oldValue: oldValue,
      newValue: newValue,
      func: func._id,
      user: func.user
    }
  };

  switch (newValue)
  {
    case 'pending':
    {
      func.status = 'pickup';
      func.pickup = newValue;
      func.carts = [];
      func.problemArea = '';
      func.comment = '';
      func.finishedAt = null;

      break;
    }

    case 'success':
    {
      const carts = !Array.isArray(data.carts) ? [] : data.carts.filter(v => typeof v === 'number');

      if (!carts.length)
      {
        throw app.createError('Invalid carts.', 'INPUT', 400);
      }

      func.status = 'finished';
      func.pickup = newValue;
      func.carts = carts;
      func.problemArea = '';
      func.comment = '';
      func.finishedAt = new Date();

      pickupEvent.carts = carts;

      break;
    }

    case 'failure':
    {
      if (typeof data.problemArea !== 'string' || !data.problemArea.length)
      {
        throw app.createError('Invalid problem area.', 'INPUT', 400);
      }

      if (typeof data.comment !== 'string')
      {
        throw app.createError('Invalid comment.', 'INPUT', 400);
      }

      func.status = 'problem';
      func.pickup = newValue;
      func.carts = [];
      func.problemArea = data.problemArea;
      func.comment = data.comment;
      func.finishedAt = new Date();

      pickupEvent.problemArea = data.problemArea;
      pickupEvent.comment = data.comment;

      break;
    }

    default:
      throw app.createError(`Invalid picklist status.`, 'INPUT', 400);
  }

  whOrder.finalize();

  const [setCartsResults, eventsResults] = await Promise.all([
    WhSetCart.updateByFunc(whOrder, func._id),
    WhEvent.record([pickupEvent], session),
    whOrder.save()
  ]);

  await session.commitTransaction();

  const leanWhOrder = whOrder.toJSON();

  return {
    response: {
      order: leanWhOrder
    },
    messages: [...setCartsResults.messages, ...eventsResults.messages, {
      topic: `${WhOrder.TOPIC_PREFIX}.updated`,
      message: {updated: [leanWhOrder]}
    }]
  };
};
