// wh/state/picklist.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, user, session}) =>
{
  const {
    OldWhEvent: WhEvent,
    OldWhOrder: WhOrder,
    OldWhUser: WhUser,
    OldWhSetCart: WhSetCart
  } = module;

  const whOrder = await WhOrder.findById(data.whOrderId).session(session).exec();

  if (!whOrder)
  {
    throw app.createError(`Order not found.`, 'INPUT', 400);
  }

  const func = whOrder.funcs.find(f => f._id === data.funcId);

  if (!func)
  {
    throw app.createError(`Function not found.`, 'INPUT', 400);
  }

  const newValue = data.newValue;
  const oldValue = func.picklist;

  func.picklist = newValue;
  func.carts = [];
  func.problemArea = '';
  func.comment = '';
  func.finishedAt = null;

  const distStatus = func._id === 'packer' ? 'packStatus' : 'fifoStatus';

  switch (newValue)
  {
    case 'pending':
      whOrder[distStatus] = 'pending';
      func.status = 'picklist';
      func.pickup = 'pending';
      break;

    case 'require':
      whOrder[distStatus] = 'pending';
      func.status = 'pickup';
      func.pickup = 'pending';
      break;

    case 'ignore':
      func.status = 'finished';
      func.pickup = 'ignore';
      func.finishedAt = new Date();

      if (distStatus === 'packStatus')
      {
        whOrder.packStatus = 'finished';
      }
      else
      {
        const fmx = whOrder.funcs[0].status;
        const kitter = whOrder.funcs[1].status;

        whOrder.fifoStatus = fmx === 'finished' && kitter === 'finished' ? 'finished' : 'pending';
      }
      break;

    default:
      throw app.createError(`Invalid picklist status.`, 'INPUT', 400);
  }

  const oldStatus = whOrder.status;

  whOrder.finalizeOrder();

  const now = new Date();
  const events = [{
    type: 'picklist',
    user,
    time: now,
    order: whOrder._id,
    data: {
      oldValue,
      newValue,
      func: func._id,
      user: func.user
    }
  }];

  const [setCartsResults, eventsResults] = await Promise.all([
    WhSetCart.updateByFunc(whOrder, func._id),
    WhEvent.record(events, session),
    func.user && newValue === 'ignore'
      ? WhUser.collection.updateOne({_id: func.user.id}, {$set: {lastPickupAt: now}}, {session})
      : null,
    whOrder.save()
  ]);

  const setResults = await whOrder.finalizeSet(setCartsResults);

  await session.commitTransaction();

  const leanWhOrder = whOrder.toJSON();

  return {
    response: {
      order: leanWhOrder
    },
    results: [
      setCartsResults,
      eventsResults,
      setResults
    ],
    messages: [{
      topic: `${WhOrder.TOPIC_PREFIX}.updated`,
      message: {updated: [leanWhOrder]}
    }]
  };
};
