// wh/state/startNewSet.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = async (app, module, {data, user, session}) =>
{
  const {
    OldWhOrder: WhOrder,
    OldWhEvent: WhEvent,
    OldWhDowntime: WhDowntime,
    OldWhUser: WhUser,
    OldWhLine: WhLine
  } = module;

  const {
    whUser,
    date,
    downtimeReason,
    downtimeComment,
    personnelId
  } = data;

  if (whUser.func === 'packer')
  {
    throw app.createError('No orders for Packer.', 'NO_PENDING_ORDERS', 400);
  }

  const now = new Date();
  const minDowntimeDuration = (module.state.settings.minPickupDowntime || 120) * 1000;
  const downtime = whUser.lastPickupAt && (now.getTime() - whUser.lastPickupAt.getTime() > minDowntimeDuration);

  if (downtime && !downtimeReason)
  {
    return {
      response: {
        result: 'pickDowntimeReason',
        startedAt: whUser.lastPickupAt,
        personnelId,
        user: whUser
      }
    };
  }

  const problemLines = await findLastProblemLines();

  const pendingOrder = await findNextPendingOrder(
    date,
    problemLines,
    module.state.settings.enabledMrps || []
  );

  const [setOrders, nextSetId] = await Promise.all([
    findMorePendingOrders(pendingOrder),
    findNextSetId(date)
  ]);

  const withPicklist = {$in: [], $set: {}};
  const withoutPicklist = {$in: [], $set: {}};
  const events = [];
  const mrps = new Set();

  setOrders.forEach(order =>
  {
    const {$in, $set} = order.picklistDone ? withPicklist : withoutPicklist;

    mrps.add(order.mrp);
    $in.push(order._id);

    $set.status = order.status = 'started';
    $set.set = order.set = nextSetId;
    $set.startedAt = order.startedAt = now;
    $set.picklistFunc = order.picklistFunc = whUser.func;
    $set.users = order.users = [whUser._id];

    if (order.picklistDone !== true)
    {
      $set.picklistDone = order.picklistDone = null;
    }

    const func = order.funcs.find(f => f._id === whUser.func);
    const funcI = order.funcs.indexOf(func);

    $set[`funcs.${funcI}.user`] = func.user = {id: whUser._id, label: whUser.label};
    $set[`funcs.${funcI}.startedAt`] = func.startedAt = now;
    $set[`funcs.${funcI}.status`] = func.status = 'picklist';

    events.push({
      type: 'newSetStarted',
      time: now,
      user,
      order: order._id,
      data: {
        set: order.set,
        user: whUser
      }
    });
  });

  const writeOperations = [];

  [withPicklist, withoutPicklist].forEach(({$in, $set}) =>
  {
    if ($in.length)
    {
      writeOperations.push({
        updateMany: {
          filter: {_id: {$in}},
          update: {$set}
        }
      });
    }
  });

  await WhLine.lockLines(setOrders[0].lines);

  const [eventsResults, downtimeResults, lineResults] = await Promise.all([
    WhEvent.record(events, session),
    WhDowntime.createDowntime({
      reason: downtimeReason,
      startedAt: whUser.lastPickupAt,
      user: {id: whUser._id, label: whUser.label},
      comment: downtimeComment,
      session
    }),
    WhLine.startNewSet(setOrders, session),
    WhUser.collection.updateOne({_id: whUser._id}, {$set: {lastPickupAt: now}}, {session}),
    WhOrder.collection.bulkWrite(writeOperations, {session})
  ]);

  await session.commitTransaction();

  setImmediate(async () =>
  {
    try
    {
      await module.state.printLabels(setOrders, 1, whUser.func);
    }
    catch (err)
    {
      module.error(err, `Failed to print labels after starting a new set.`);
    }
  });

  return {
    response: {
      result: 'newSetStarted',
      orders: setOrders,
      user: whUser
    },
    results: [
      eventsResults,
      downtimeResults,
      lineResults
    ],
    messages: [
      {
        topic: `${WhOrder.TOPIC_PREFIX}.updated`,
        message: {updated: setOrders}
      },
      {
        topic: 'old.wh.newSetStarted',
        message: {
          date,
          set: nextSetId,
          mrps: Array.from(mrps)
        }
      }
    ]
  };

  async function findLastProblemLines()
  {
    const results = await WhOrder
      .aggregate([
        {$match: {date}},
        {$group: {_id: null, set: {$max: '$set'}}}
      ])
      .session(session)
      .exec();

    if (results.length)
    {
      const problemOrder = await WhOrder
        .findOne({
          date,
          set: results[0].set,
          status: 'problem'
        })
        .select({lines: 1})
        .lean()
        .session(session)
        .exec();

      if (problemOrder)
      {
        return problemOrder.lines;
      }
    }

    return [];
  }

  async function tryFindNextPendingOrderForNextLine(date, lines, mrps, ignore)
  {
    try
    {
      return await tryFindNextPendingOrder(date, lines.shift()._id, mrps, ignore);
    }
    catch (err)
    {
      if (err.code === 'NO_PENDING_ORDERS')
      {
        if (lines.length)
        {
          return await tryFindNextPendingOrderForNextLine(date, lines, mrps, ignore);
        }

        return await tryFindNextPendingOrder(date, null, mrps, ignore);
      }

      throw err;
    }
  }

  async function tryFindNextPendingOrder(date, line, mrps, ignore)
  {
    const conditions = {
      status: 'pending',
      date
    };

    if (!module.state.settings.ignorePsStatus)
    {
      conditions.psStatus = {$in: ['unknown', 'finished']};
    }

    if (line)
    {
      conditions['lines._id'] = line;
    }

    if (ignore.length)
    {
      conditions._id = {$nin: ignore};
    }

    if (mrps.length)
    {
      conditions.mrp = {$in: mrps};
    }

    const whOrder = await WhOrder
      .findOne(conditions)
      .sort({startTime: 1, line: 1})
      .lean()
      .session(session)
      .exec();

    if (!whOrder)
    {
      throw app.createError('No valid, pending orders.', 'NO_PENDING_ORDERS', 400);
    }

    return whOrder;
  }

  async function findNextPendingOrder(date, problemLines, enabledMrps)
  {
    if (problemLines.length)
    {
      return await tryFindNextPendingOrderForNextLine(date, problemLines, enabledMrps, []);
    }

    return await tryFindNextPendingOrder(date, null, enabledMrps, []);
  }

  async function findMorePendingOrders(pendingOrder)
  {
    const whSettings = module.state.settings;

    const groupDuration = (whSettings.groupDuration || 2) * 3600 * 1000;
    const minSetDuration = (whSettings.minSetDuration || 120) * 60 * 1000;
    const maxSetDuration = (whSettings.maxSetDuration || 120) * 60 * 1000;
    const maxSetDifference = (whSettings.maxSetDifference || 120) * 60 * 1000;
    const maxSetSize = whSettings.maxSetSize || 5;

    const conditions = {
      status: 'pending',
      date: pendingOrder.date,
      line: pendingOrder.line
    };

    if (!whSettings.ignorePsStatus)
    {
      conditions.psStatus = {$in: ['unknown', 'finished']};
    }

    const whOrders = await WhOrder
      .find(conditions)
      .sort({startTime: 1})
      .lean()
      .session(session)
      .exec();

    const setOrders = [];
    let setDuration = 0;

    while (whOrders.length && setOrders.length < maxSetSize)
    {
      const whOrder = whOrders.shift();
      const orderDuration = whOrder.finishTime - whOrder.startTime;

      // Stop: the time difference between orders is too big.
      if (setOrders.length
        && (whOrder.startTime - setOrders[setOrders.length - 1].finishTime > maxSetDifference))
      {
        break;
      }

      // Order fits in the group's boundaries or it's the first order.
      if (setOrders.length === 0
        || (setDuration + orderDuration <= groupDuration))
      {
        setDuration += orderDuration;

        setOrders.push(whOrder);

        continue;
      }

      // Min set duration exceeded, don't extend.
      if (setDuration > minSetDuration)
      {
        break;
      }

      // Max set duration exceeded, ignore.
      if (setDuration + orderDuration > maxSetDuration)
      {
        continue;
      }

      // Extend set.
      setDuration += orderDuration;

      setOrders.push(whOrder);
    }

    return setOrders;
  }

  async function findNextSetId(date)
  {
    const result = await WhOrder
      .aggregate([
        {$match: {date}},
        {$group: {_id: null, set: {$max: '$set'}}}
      ])
      .session(session)
      .exec();

    if (result && result.length)
    {
      return (result[0].set || 0) + 1;
    }

    return 1;
  }
};
