// wmes-luma2-backend/models/luma2Event.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'Luma2Event';

exports.setUp = (app, mongoose) =>
{
  const TYPES = {
    10001: 'connected',
    10002: 'disconnected',
    10003: 'order-queued',
    1: 'order-started',
    2: 'order-finished',
    3: 'rearm-started',
    4: 'rearm-finished',
    5: 'pce-started',
    6: 'pce-finished',
    7: 'inspection-started',
    8: 'inspection-finished',
    9: 'receipt-started',
    10: 'receipt-finished',
    11: 'delivery-started',
    12: 'delivery-finished',
    13: 'estop-started',
    14: 'estop-finished',
    15: 'pause-started',
    16: 'pause-finished'
  };

  const luma2EventSchema = new mongoose.Schema({
    line: {
      type: String,
      required: true
    },
    type: {
      type: String,
      required: true
    },
    time: {
      type: Date,
      required: true
    },
    station: {
      type: Number,
      default: -1
    },
    program: {
      type: Number,
      default: -1
    },
    order: {
      type: String,
      default: ''
    },
    pce: {
      type: Number,
      default: -1
    }
  }, {
    id: false,
    versionKey: false
  });

  luma2EventSchema.statics.TOPIC_PREFIX = 'luma2.events';
  luma2EventSchema.statics.BROWSE_LIMIT = 100;
  luma2EventSchema.statics.TYPES = TYPES;

  luma2EventSchema.index({createdAt: -1});
  luma2EventSchema.index({order: 1});
  luma2EventSchema.index({line: 1, createdAt: -1});
  luma2EventSchema.index({type: 1, createdAt: -1});
  luma2EventSchema.index({station: 1, createdAt: -1});
  luma2EventSchema.index({program: 1, createdAt: -1});

  mongoose.model(exports.name, luma2EventSchema);
};
