// wmes-compRel/routes/acceptEntry.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');
const locks = require('util/locks');
const jsonDeepEqual = require('util/jsonDeepEqual');

module.exports = (app, {user, CompRelEntry}, req, res, next) =>
{
  const sessionUser = req.session.user;

  step(
    function()
    {
      this.releaseLock = locks.create(`compRel/entries/${req.params.id}`, this.next());
    },
    function()
    {
      CompRelEntry.findById(req.params.id).exec(this.next());
    },
    function(err, entry)
    {
      if (err)
      {
        return this.skip(err);
      }

      if (!entry)
      {
        return this.skip(app.createError('Not found.', 'NOT_FOUND', 400));
      }

      const funcI = entry.funcs.findIndex(f => f._id === req.body.func);

      if (funcI === -1)
      {
        return this.skip(app.createError('Invalid func.', 'INPUT', 400));
      }

      const oldFunc = entry.funcs[funcI].toObject();

      if (!user.isAllowedTo(sessionUser, [['PROD_DATA:MANAGE', 'COMP_REL:MANAGE']])
        && !oldFunc.users.some(u => u.id === sessionUser._id))
      {
        return this.skip(app.createError('Not allowed.', 'AUTH', 400));
      }

      const pending = req.body.status === 'pending';
      const now = new Date();
      const userInfo = user.createUserInfo(sessionUser, req);
      const newFunc = {
        _id: oldFunc._id,
        acceptedAt: pending ? null : now,
        acceptedBy: pending ? null : userInfo,
        status: req.body.status,
        comment: req.body.comment || '',
        users: oldFunc.users
      };

      if (jsonDeepEqual(newFunc, oldFunc))
      {
        return this.skip();
      }

      entry.updatedAt = now;
      entry.updater = userInfo;
      entry.funcs[funcI] = newFunc;

      const changes = {
        funcs: [{}, {}]
      };

      changes.funcs[0][newFunc._id] = oldFunc;
      changes.funcs[1][newFunc._id] = newFunc;

      entry.updateStatus(changes);

      entry.changes.push({
        date: now,
        user: userInfo,
        data: changes,
        comment: newFunc.comment
      });

      entry.save(this.next());
    },
    function(err, entry)
    {
      setImmediate(this.releaseLock);

      if (err)
      {
        return next(err);
      }

      if (!entry)
      {
        return res.sendStatus(204);
      }

      app.broker.publish(`${CompRelEntry.TOPIC_PREFIX}.updated.${entry._id}`, {
        _id: entry._id,
        change: _.last(entry.changes)
      });

      res.json(entry);
    }
  );
};
