// wmes-compRel/routes/releaseOrder.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');
const locks = require('util/locks');

module.exports = (app, {user, CompRelEntry}, req, res, next) =>
{
  const sessionUser = req.session.user;

  step(
    function()
    {
      if (!user.isAllowedTo(sessionUser, [['PROD_DATA:MANAGE', 'COMP_REL:MANAGE', 'FN:production-planner']]))
      {
        return this.skip(app.createError('Not allowed.', 'AUTH', 400));
      }

      this.releaseLock = locks.create(`compRel/entries/${req.params.id}`, this.next());
    },
    function()
    {
      CompRelEntry.findById(req.params.id).exec(this.next());
    },
    function(err, entry)
    {
      if (err)
      {
        return this.skip(err);
      }

      if (!entry)
      {
        return this.skip(app.createError('Not found.', 'NOT_FOUND', 400));
      }

      if (entry.status !== 'accepted')
      {
        return this.skip(app.createError('Invalid status.', 'STATUS', 400));
      }

      const remove = req.body.remove === true;
      const oldOrders = new Map();
      const newOrders = new Set();

      entry.orders.forEach(order =>
      {
        oldOrders.set(order._id, order);
      });

      if (Array.isArray(req.body.orders))
      {
        req.body.orders.forEach(orderNo =>
        {
          if (/^[0-9]{9}$/.test(orderNo) && (remove || !oldOrders.has(orderNo)))
          {
            newOrders.add(orderNo);
          }
        });
      }

      if (!newOrders.size)
      {
        return this.skip();
      }

      const now = new Date();
      const userInfo = user.createUserInfo(sessionUser, req);
      const changes = {
        orders: [null, null]
      };

      if (remove)
      {
        changes.orders[0] = [];

        newOrders.forEach(orderNo =>
        {
          const order = oldOrders.get(orderNo);

          if (!order)
          {
            return;
          }

          changes.orders[0].push(order);
          oldOrders.delete(orderNo);
        });

        if (!changes.orders[0].length)
        {
          return this.skip();
        }

        entry.orders = Array.from(oldOrders.values());
      }
      else
      {
        changes.orders[1] = [];

        newOrders.forEach(orderNo =>
        {
          const order = {
            _id: orderNo,
            releasedAt: now,
            releasedBy: userInfo
          };

          changes.orders[1].push(order);
          entry.orders.push(order);
        });
      }

      entry.updatedAt = now;
      entry.updater = userInfo;

      entry.changes.push({
        date: now,
        user: userInfo,
        data: changes,
        comment: req.body.comment || ''
      });

      entry.save(this.next());
    },
    function(err, entry)
    {
      setImmediate(this.releaseLock);

      if (err)
      {
        return next(err);
      }

      if (!entry)
      {
        return res.sendStatus(204);
      }

      app.broker.publish(`${CompRelEntry.TOPIC_PREFIX}.updated.${entry._id}`, {
        _id: entry._id,
        change: _.last(entry.changes)
      });

      res.json(entry);
    }
  );
};
