// suggestions/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  sioId: 'sio',
  userId: 'user',
  mailSenderId: 'mail/sender',
  reportsId: 'reports',
  kaizenId: 'kaizen',
  attachmentsDest: null,
  multiType: true,
  remind: 14
};

exports.models = [
  require('./models/suggestion'),
  require('user/models/user'),
  require('kaizen/models/kaizenOrder'),
  require('kaizen/models/kaizenSection'),
  require('kaizen/models/kaizenCategory'),
  require('kaizen/models/kaizenProductFamily'),
  require('behaviorObsCards/models/behaviorObsCard'),
  require('minutesForSafetyCards/models/minutesForSafetyCard')
];

exports.recordTopics = {
  blacklist: ['suggestions.added', 'suggestions.edited']
};

exports.republishTopics = ['suggestions.seen.*'];

exports.userPrivileges = ['SUGGESTIONS:MANAGE'];

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'mongoose': [
    require('./durationRecalculator'),
    require('./stats'),
    require('./observerCleaner'),
  ],
  'express user reports kaizen': require('./routes'),
  'sio user': require('./commands'),
  'mailSender': [
    require('./notifier'),
    require('./reminder')
  ]
};
