// suggestions/reminder.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const fs = require('fs');
const _ = require('lodash');
const step = require('h5.step');
const ejs = require('ejs');
const moment = require('moment');

module.exports = (app, module) =>
{
  if (!module.config.remind)
  {
    return;
  }

  const logger = module.logger.create({submodule: 'reminder'});
  const mailSender = app[module.config.mailSenderId];
  const mongoose = app[module.config.mongooseId];
  const User = mongoose.model('User');
  const Suggestion = mongoose.model('Suggestion');

  const DAYS_AGO = module.config.remind;

  const emailTemplateFile = `${__dirname}/templates/reminder.email.pl.ejs`;
  const renderEmail = ejs.compile(fs.readFileSync(emailTemplateFile, 'utf8'), {
    cache: true,
    filename: emailTemplateFile,
    compileDebug: false,
    rmWhitespace: true
  });

  app.broker.subscribe('app.started', scheduleNextReminder).setLimit(1);

  function scheduleNextReminder()
  {
    const delay = moment()
      .hours(12)
      .minutes(0)
      .seconds(0)
      .milliseconds(0)
      .add(1, 'days')
      .diff(Date.now());

    setTimeout(remind, delay);
  }

  function remind()
  {
    step(
      function aggregateStep()
      {
        const weekAgo = moment().subtract(DAYS_AGO, 'days').valueOf();

        this.conditions = {
          status: {$in: ['new', 'accepted', 'inProgress', 'verification']},
          remindedAt: {$lt: weekAgo}
        };

        Suggestion.aggregate([
          {$match: this.conditions},
          {$group: {_id: '$confirmer.id', totalCount: {$sum: 1}}}
        ], this.parallel());

        Suggestion.aggregate([
          {$match: {
            status: {$in: ['new', 'accepted', 'inProgress', 'verification']},
            updatedAt: {$lt: new Date(weekAgo)}
          }},
          {$group: {_id: '$confirmer.id', totalCount: {$sum: 1}}}
        ], this.parallel());
      },
      function getUserDataStep(err, remindedAtResults, updatedAtResults)
      {
        if (err)
        {
          return this.skip(err);
        }

        const updatedAtStats = {};

        _.forEach(updatedAtResults, function(result)
        {
          updatedAtStats[result._id] = result;
        });

        const userIds = [];
        const userIdToStats = {};

        _.forEach(remindedAtResults, function(result)
        {
          userIds.push(new mongoose.Types.ObjectId(result._id));
          userIdToStats[result._id] = updatedAtStats[result._id];
        });

        this.userIdToStats = userIdToStats;

        const conditions = {_id: {$in: userIds}, email: {$ne: null}};
        const fields = {firstName: 1, lastName: 1, email: 1, gender: 1};

        User.find(conditions, fields).exec(this.next());
      },
      function sendEmailsStep(err, users)
      {
        if (err)
        {
          return this.skip(err);
        }

        for (let i = 0; i < users.length; ++i)
        {
          const user = users[i];

          sendReminderEmail(user, this.userIdToStats[user._id]);
        }

        logger.info('Reminded users about incomplete suggestions.', {count: users.length});
      },
      function updateUpdatedAtStep()
      {
        Suggestion.updateMany(this.conditions, {$set: {remindedAt: Date.now()}}, this.next());
      },
      function(err)
      {
        if (err)
        {
          logger.error(err);
        }

        this.conditions = null;
        this.userIdToStats = null;

        scheduleNextReminder();
      }
    );
  }

  function sendReminderEmail(user, stats)
  {
    const mailOptions = {
      to: user.email,
      replyTo: user.email,
      subject: '[WMES] [Usprawnienia] Niezakończone zgłoszenia',
      html: renderEmail({
        urlPrefix: app.options.emailUrlPrefix,
        user,
        plural: plural(stats.totalCount),
        totalCount: stats.totalCount
      })
    };

    mailSender.send(mailOptions, function(err)
    {
      if (err)
      {
        logger.error(err, 'Failed to remind about incomplete suggestions.', {recipients: mailOptions.to});
      }
    });
  }

  function plural(n)
  {
    return (n % 10 < 5) && (n % 10 > 1) && ((~~(n / 10) % 10) !== 1);
  }
};
