// wmes-compRel/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  morId: 'mor',
  mailSenderId: 'mail/sender',
  uploadsDest: './data/uploads/compRel'
};

exports.models = [
  require('./models/compRelEntry'),
  require('./models/compRelFunc'),
  require('orders/models/component'),
  require('mrpControllers/models/mrpController'),
  require('user/models/user')
];

exports.userPrivileges = [
  'COMP_REL:VIEW',
  'COMP_REL:ADD',
  'COMP_REL:MANAGE'
];

exports.republishTopics = [
  'compRel.entries.added', 'compRel.entries.updated.*', 'compRel.entries.deleted',
  'compRel.funcs.added', 'compRel.funcs.edited', 'compRel.funcs.deleted'
];

exports.recordTopics = {
  debug: [
    'compRel.funcs.added', 'compRel.funcs.edited'
  ],
  warning: [
    'compRel.entries.deleted',
    'compRel.funcs.deleted'
  ]
};

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'user express settings': require('./routes'),
  'mailSender': require('./notifier')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    funcs: {model: module.CompRelFunc}
  };

  module.tmpAttachments = {};
};
