// orders/importer/emptyOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const createParser = require('./createParser');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  filterRe: /^T_WMES_EMPTY_(ORDERS|OPERS)_([0-9]+)\.txt$/,
  parsedOutputDir: null
};

exports.models = [
  require('../models/emptyOrder')
];

exports.recordTopics = {
  info: ['emptyOrders.synced']
};

exports.start = function startEmptyOrdersImporterModule(app, module)
{
  const mongoose = app[module.config.mongooseId];

  if (!mongoose)
  {
    throw new Error('orders/importer/emptyOrders module requires the mongoose module!');
  }

  const EmptyOrder = mongoose.model('EmptyOrder');

  createParser(app, module, handleParseResult);

  function handleParseResult(err, orders)
  {
    if (err)
    {
      return;
    }

    const emptyOrders = [];

    _.forEach(orders, function(order)
    {
      if (order.operations === null)
      {
        emptyOrders.push(new EmptyOrder(order).toObject({depopulate: true}));
      }
    });

    if (emptyOrders.length === 0)
    {
      return publishMessage(0);
    }

    EmptyOrder.collection.insertMany(emptyOrders, {ordered: false}, function(err, docs)
    {
      if (err)
      {
        module.error(err, 'Error during insertion of empty orders.');
      }

      publishMessage(Array.isArray(docs) ? docs.length : 0);
    });
  }

  function publishMessage(count)
  {
    module.info(`Synced ${count} empty orders.`);

    app.broker.publish('emptyOrders.synced', {count: count});
  }
};
