// orders/importer/intake.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const path = require('path');
const _ = require('lodash');
const moment = require('moment');
const step = require('h5.step');
const fs = require('fs-extra');
const parseSapTextTable = require('sap/util/parseSapTextTable');
const parseSapNumber = require('sap/util/parseSapNumber');
const parseSapString = require('sap/util/parseSapString');
const parseSapDate = require('sap/util/parseSapDate');
const parseSapTime = require('sap/util/parseSapTime');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  filterRe: /^T_ZOIN(?:_(.*?))?\.txt$/,
  parsedOutputDir: null
};

exports.models = [
  require('../models/order'),
  require('../models/orderIntake')
];

exports.recordTopics = {
  info: ['orders.intake.synced'],
  error: ['orders.intake.syncFailed']
};

exports.start = (app, module) =>
{
  const {Order, OrderIntake} = module;

  const filePathCache = {};
  let locked = false;
  const queue = [];

  app.broker.subscribe('directoryWatcher.changed', queueFile).setFilter(filterFile);

  function filterFile(fileInfo)
  {
    if (filePathCache[fileInfo.filePath])
    {
      return false;
    }

    const matches = fileInfo.fileName.match(module.config.filterRe);

    if (matches === null)
    {
      return false;
    }

    fileInfo.plant = matches[1] || '?';
    fileInfo.timeKey = fileInfo.plant + ':' + createTimeKey(fileInfo.timestamp);

    return true;
  }

  function createTimeKey(timestamp)
  {
    return moment(timestamp).subtract(1, 'days').format('YYMMDDHH');
  }

  function queueFile(fileInfo)
  {
    filePathCache[fileInfo.filePath] = true;

    queue.push(fileInfo);

    module.info('Queued...', {fileInfo});

    setImmediate(importNext);
  }

  function importNext()
  {
    if (locked)
    {
      return;
    }

    const fileInfo = queue.shift();

    if (!fileInfo)
    {
      return;
    }

    locked = true;

    const startTime = Date.now();

    module.info('Importing...', {fileInfo});

    importFile(fileInfo, function(err, count)
    {
      cleanUpFileInfoFile(fileInfo);

      if (err)
      {
        module.error(err, 'Failed to import.', {fileInfo});

        app.broker.publish('orders.intake.syncFailed', {
          timestamp: fileInfo.timestamp,
          plant: fileInfo.plant,
          error: err.message
        });
      }
      else
      {
        module.info('Imported!', {fileInfo, count, duration: Date.now() - startTime});

        app.broker.publish('orders.intake.synced', {
          timestamp: fileInfo.timestamp,
          plant: fileInfo.plant,
          count: count
        });
      }

      locked = false;

      setImmediate(importNext);
    });
  }

  function importFile(fileInfo, done)
  {
    step(
      function readFileStep()
      {
        fs.readFile(fileInfo.filePath, {encoding: 'utf8'}, this.next());
      },
      function parseFileStep(err, fileContents)
      {
        if (err)
        {
          return this.skip(err);
        }

        module.info('Parsing...', {fileInfo, fileSize: fileContents.length});

        const t = Date.now();

        this.orderIntakes = parseOrderIntakeTable(fileContents);

        module.info('Parsed!', {fileInfo, count: this.orderIntakes.length, duration: Date.now() - t});

        setImmediate(this.next());
      },
      function upsertOrderIntakesStep()
      {
        module.info('Upserting order intakes...', {fileInfo});

        const t = Date.now();

        upsertNextOrderIntakesBatch(this.orderIntakes, 0, 10, this.next());

        function upsertNextOrderIntakesBatch(orderIntakes, batchNo, batchSize, done)
        {
          const startIndex = batchNo * batchSize;
          const endIndex = Math.min(startIndex + batchSize, orderIntakes.length);

          step(
            function()
            {
              for (let i = startIndex; i < endIndex; ++i)
              {
                const orderIntake = orderIntakes[i];

                OrderIntake.replaceOne(
                  {_id: orderIntake._id},
                  orderIntake,
                  {upsert: true},
                  this.group()
                );
              }
            },
            function(err)
            {
              if (err)
              {
                module.error(err, 'Failed to upsert order intakes.');
              }

              if (endIndex === orderIntakes.length)
              {
                module.info('Upserted!', {fileInfo, count: orderIntakes.length, duration: Date.now() - t});

                return done();
              }

              return setImmediate(upsertNextOrderIntakesBatch, orderIntakes, batchNo + 1, batchSize, done);
            }
          );
        }
      },
      function updateOrdersStep()
      {
        module.info('Updating orders...', {fileInfo});

        upsertNextOrdersBatch(fileInfo, new Date(), this.orderIntakes, 0, 10, this.next());
      },
      function finalizeStep(err)
      {
        return done(err, this.orderIntakes.length);
      }
    );
  }

  function upsertNextOrdersBatch(fileInfo, t, orderIntakes, batchNo, batchSize, done)
  {
    const startIndex = batchNo * batchSize;
    const endIndex = Math.min(startIndex + batchSize, orderIntakes.length);

    step(
      function()
      {
        for (let i = startIndex; i < endIndex; ++i)
        {
          updateOrders(t, orderIntakes[i], this.group());
        }
      },
      function(err)
      {
        if (err)
        {
          module.error(err, 'Failed to update orders.', {fileInfo});
        }

        if (endIndex === orderIntakes.length)
        {
          module.info('Updated orders!', {fileInfo, duration: Date.now() - t.getTime()});

          return done();
        }

        return setImmediate(upsertNextOrdersBatch, fileInfo, t, orderIntakes, batchNo + 1, batchSize, done);
      }
    );
  }

  function updateOrders(t, orderIntake, done)
  {
    step(
      function findOrdersStep()
      {
        const conditions = {
          salesOrder: orderIntake._id.no,
          salesOrderItem: orderIntake._id.item
        };
        const fields = {
          description: 1,
          soldToParty: 1,
          salesOrderDate: 1
        };

        Order.find(conditions, fields).lean().exec(this.next());
      },
      function updateOrdersStep(err, orders)
      {
        if (err)
        {
          return this.skip(err);
        }

        for (let i = 0; i < orders.length; ++i)
        {
          updateOrder(t, orders[i], orderIntake, this.group());
        }
      },
      done
    );
  }

  function updateOrder(t, order, orderIntake, done)
  {
    let changed = false;
    const changes = {
      time: t,
      user: null,
      oldValues: {},
      newValues: {},
      comment: ''
    };
    const $set = {
      updatedAt: t
    };

    ['description', 'soldToParty', 'salesOrderDate'].forEach(p =>
    {
      const oldValue = order[p] || null;
      const newValue = orderIntake[p] || null;

      if (_.isEqual(oldValue, newValue))
      {
        return;
      }

      changed = true;
      changes.oldValues[p] = oldValue;
      changes.newValues[p] = newValue;
      $set[p] = newValue;
    });

    if (!changed)
    {
      return setImmediate(done);
    }

    Order.updateOne({_id: order._id}, {$set: $set, $push: {changes: changes}}, done);
  }

  function cleanUpFileInfoFile(fileInfo)
  {
    setTimeout(removeFilePathFromCache, 15000, fileInfo.filePath);

    if (module.config.parsedOutputDir)
    {
      moveFileInfoFile(fileInfo.filePath);
    }
    else
    {
      deleteFileInfoFile(fileInfo.filePath);
    }
  }

  function moveFileInfoFile(oldFilePath)
  {
    const newFilePath = path.join(module.config.parsedOutputDir, path.basename(oldFilePath));

    fs.move(oldFilePath, newFilePath, {overwrite: true}, function(err)
    {
      if (err)
      {
        module.error(err, 'Failed to rename file.', {oldFilePath, newFilePath});
      }
    });
  }

  function deleteFileInfoFile(filePath)
  {
    fs.unlink(filePath, function(err)
    {
      if (err)
      {
        module.error(err, 'Failed to delete file.', {filePath});
      }
    });
  }

  function removeFilePathFromCache(filePath)
  {
    delete filePathCache[filePath];
  }

  function parseOrderIntakeTable(input)
  {
    return parseSapTextTable(input, {
      columnMatchers: {
        nc12: /^Material$/,
        description: /^Desc/,
        qty: /^Order qty/,
        salesOrder: /^Sales Doc/,
        salesOrderItem: /Item/,
        soldToParty: /^Sold-to/,
        shipTo: /^Ship to/,
        docDate: /^Doc.*?Date/,
        docTime: /^Doc.*?Time/
      },
      valueParsers: {
        nc12: function(input) { return input.replace(/^0+/, ''); },
        description: parseSapString,
        qty: parseSapNumber,
        docDate: parseSapDate,
        docTime: parseSapTime
      },
      itemDecorator: function(obj)
      {
        if (!obj.description.length || !obj.docDate || !obj.docTime)
        {
          return null;
        }

        return {
          _id: {
            no: obj.salesOrder,
            item: obj.salesOrderItem
          },
          nc12: obj.nc12,
          description: obj.description,
          qty: obj.qty,
          soldToParty: obj.soldToParty,
          shipTo: obj.shipTo,
          salesOrderDate: new Date(
            obj.docDate.y, obj.docDate.m - 1, obj.docDate.d,
            obj.docTime.h, obj.docTime.m, obj.docTime.s
          )
        };
      }
    });
  }
};
