// orders/importer/parseOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const parseSapTextTable = require('sap/util/parseSapTextTable');
const parseSapString = require('sap/util/parseSapString');
const parseSapNumber = require('sap/util/parseSapNumber');
const parseSapDate = require('sap/util/parseSapDate');
const parseSapTime = require('sap/util/parseSapTime');

module.exports = function parseOrders(input, orders, importTs, Order)
{
  if (!input.trim().endsWith('----------'))
  {
    return;
  }

  return parseSapTextTable(input, {
    columnMatchers: {
      no: /^Order$/,
      nc12: /^Material$/,
      name: /^Mat.*?desc/,
      mrp: /^MRP/,
      qty: /^Target.*?q/,
      unit: /^Unit$/,
      startDate: /^B.*?sta/,
      finishDate: /^B.*?fin/,
      statuses: /^System Status/,
      salesOrder: /^Sales O/,
      salesOrderItem: /^S.*?It/,
      priority: /^Priority$/,
      scheduledStartDate: /^Sch.*?Sta/,
      scheduledFinishDate: /^Sch.*?Fin/,
      leadingOrder: /^Lead.*?Ord/,
      enteredBy: /^Ent.*?by/i,
      changedBy: /^Cha.*?by/i,
      createdDate: /^Created on$/,
      createdTime: /^Time$/
    },
    valueParsers: {
      nc12: input => input.replace(/^0+/, ''),
      name: parseSapString,
      qty: parseSapNumber,
      startDate: parseSapDate,
      finishDate: parseSapDate,
      scheduledStartDate: parseSapDate,
      scheduledFinishDate: parseSapDate,
      statuses: input => input
        .replace(/\s+/g, ' ')
        .split(' ')
        .map(status => status.replace(/\*/g, ''))
        .filter(status => status.length > 0),
      enteredBy: parseSapString,
      changedBy: parseSapString,
      createdDate: parseSapDate,
      createdTime: parseSapTime
    },
    itemDecorator: obj =>
    {
      obj.sapCreatedAt = new Date(
        obj.createdDate.y, obj.createdDate.m - 1, obj.createdDate.d,
        obj.createdTime.h, obj.createdTime.m, obj.createdTime.s
      );

      orders[obj.no] = Order.createForInsert(obj, importTs);

      return null;
    }
  });
};
