// wh/state/actions/recountLines.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, user, session}) =>
{
  const {
    OldWhLine: WhLine
  } = module;

  const linesResult = {
    updated: [],
    messages: []
  };

  const conditions = {};

  if (Array.isArray(data.lines) && data.lines.length)
  {
    conditions._id = {$in: data.lines};
  }

  const allLines = await WhLine.find(conditions).select({_id: 1}).lean().exec();

  if (!allLines.length)
  {
    return;
  }

  const lineIds = allLines.map(l => l._id);

  await WhLine.lockLines(lineIds);

  for (const lineId of lineIds)
  {
    const result = await WhLine.recount(lineId, true, true, session);

    linesResult.updated = linesResult.updated.concat(result.updated);
  }

  if (!linesResult.updated.length)
  {
    return;
  }

  await session.commitTransaction();

  linesResult.messages.push({
    topic: `${WhLine.TOPIC_PREFIX}.updated`,
    message: {updated: linesResult.updated}
  });

  return {
    results: [
      linesResult
    ]
  };
};
