// wh/state/actions/resetOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, user, session}) =>
{
  const {
    OldWhEvent: WhEvent,
    OldWhDeliveredOrder: WhDeliveredOrder,
    OldWhOrder: WhOrder,
    OldWhSetCart: WhSetCart,
    OldWhLine: WhLine
  } = module;

  const conditions = {};

  if (data.cancel)
  {
    conditions.status = {$ne: 'cancelled'};
  }

  if (data.orders)
  {
    if (!Array.isArray(data.orders) || data.orders.length !== 1)
    {
      throw app.createError('Invalid orders.', 'INPUT', 400);
    }

    conditions._id = {$in: data.orders};
  }
  else
  {
    conditions.set = data.set;
    conditions.date = data.date;
  }

  const whOrders = await WhOrder
    .find(conditions)
    .session(session)
    .exec();

  if (!whOrders.length)
  {
    return {response: {orders: whOrders}};
  }

  const time = new Date();
  const event = {
    type: data.cancel ? 'ordersCancelled' : 'ordersReset',
    time,
    user,
    data: {
      orders: WhEvent.orders(whOrders)
    }
  };

  const date = whOrders[0].date;
  const set = whOrders[0].set;

  const $set = {
    status: data.cancel ? 'cancelled' : 'pending',
    distStatus: 'pending',
    fifoStatus: 'pending',
    packStatus: 'pending',
    psDistStatus: 'pending',
    problem: '',
    set: null,
    startedAt: null,
    finishedAt: null,
    picklistFunc: null,
    picklistDone: 'pending',
    users: [],
    funcs: whOrders[0].funcs.map(func =>
    {
      return {
        _id: func._id,
        user: null,
        startedAt: null,
        finishedAt: null,
        status: 'pending',
        picklist: 'pending',
        pickup: 'pending',
        carts: [],
        problemArea: '',
        comment: ''
      };
    })
  };

  const whOrderIds = [];

  for (const whOrder of whOrders)
  {
    whOrderIds.push(whOrder._id);
    Object.assign(whOrder, $set);
  }

  const results = [];

  const [deliveredOrders, eventsResults] = await Promise.all([
    WhDeliveredOrder
      .find({whOrder: {$in: whOrderIds}})
      .select({_id: 1})
      .lean()
      .session(session)
      .exec(),
    WhEvent.record(event, session),
    WhOrder.collection.updateMany({_id: {$in: whOrderIds}}, {$set}, {session})
  ]);

  results.push(eventsResults);

  if (deliveredOrders.length)
  {
    results.push({
      messages: [{
        topic: `${WhDeliveredOrder.TOPIC_PREFIX}.updated`,
        message: {
          deleted: deliveredOrders
        }
      }]
    });
  }

  if (data.orders)
  {
    const [setCartsResults] = await Promise.all([
      WhSetCart.deleteByWhOrder(whOrders[0]),
      deliveredOrders.length
        ? WhDeliveredOrder.collection.deleteMany({whOrder: {$in: whOrderIds}}, {session})
        : null
    ]);

    const [lineResults, setResults] = await Promise.all([
      WhLine.recountByWhOrder(whOrders[0], true, deliveredOrders.length > 0),
      WhOrder.finalizeSet(date, set, setCartsResults, session)
    ]);

    results.push(
      setCartsResults,
      lineResults,
      setResults
    );
  }
  else
  {
    const [setCartsResults] = await Promise.all([
      WhSetCart.deleteBySet(data.date, data.set, session),
      deliveredOrders.length
        ? WhDeliveredOrder.collection.deleteMany({whOrder: {$in: whOrderIds}}, {session})
        : null
    ]);

    const lineResults = await WhLine.recountByWhOrder(whOrders[0], true, deliveredOrders.length > 0);

    results.push(
      setCartsResults,
      lineResults
    );
  }

  await session.commitTransaction();

  const leanWhOrders = whOrders.map(whOrder => whOrder.toJSON());

  return {
    response: {
      orders: leanWhOrders
    },
    results,
    messages: [{
      topic: `${WhOrder.TOPIC_PREFIX}.updated`,
      message: {updated: leanWhOrders}
    }]
  };
};
