// wh/state/actions/resetOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.session = true;

exports.execute = async (app, module, {data, user, session}) =>
{
  const {
    OldWhEvent: WhEvent,
    OldWhDeliveredOrder: WhDeliveredOrder,
    OldWhOrder: WhOrder,
    OldWhSetCart: WhSetCart,
    OldWhLine: WhLine,
    OldWhPendingSetAssign: WhPendingSetAssign
  } = module;

  const conditions = {};

  if (data.finish)
  {
    data.cancel = true;
  }

  if (data.cancel)
  {
    conditions.status = {$ne: 'cancelled'};
  }

  if (data.orders)
  {
    if (!Array.isArray(data.orders) || data.orders.length !== 1)
    {
      throw app.createError('Invalid orders.', 'INPUT', 400);
    }

    conditions._id = {$in: data.orders};
  }
  else
  {
    conditions.set = data.set;
    conditions.date = data.date;
  }

  const whOrders = await WhOrder
    .find(conditions)
    .session(session)
    .exec();

  if (!whOrders.length)
  {
    return {response: {orders: whOrders}};
  }

  const time = new Date();
  const finish = data.finish;
  const event = {
    type: finish ? 'ordersFinished' : data.cancel ? 'ordersCancelled' : 'ordersReset',
    time,
    user,
    data: {
      orders: WhEvent.orders(whOrders)
    }
  };

  const date = whOrders[0].date;
  const set = whOrders[0].set;

  const $set = {
    status: data.cancel ? 'cancelled' : 'pending',
    distStatus: 'pending',
    fifoStatus: 'pending',
    packStatus: 'pending',
    psDistStatus: 'pending',
    problem: '',
    set: null,
    startedAt: null,
    finishedAt: null,
    picklistFunc: null,
    picklistDone: 'pending',
    users: [],
    funcs: whOrders[0].funcs.map(func =>
    {
      return {
        _id: func._id,
        user: null,
        startedAt: null,
        finishedAt: null,
        status: 'pending',
        picklist: 'pending',
        pickup: 'pending',
        carts: [],
        problemArea: '',
        comment: ''
      };
    })
  };

  const whOrderIds = [];

  for (const whOrder of whOrders)
  {
    whOrderIds.push(whOrder._id);

    if (!finish)
    {
      Object.assign(whOrder, $set);
    }
  }

  const results = [];

  const [deliveredOrders, eventsResults] = await Promise.all([
    WhDeliveredOrder
      .find({whOrder: {$in: whOrderIds}})
      .select({
        quantityTodo: 1,
        quantityDone: 1
      })
      .lean()
      .session(session)
      .exec(),
    WhEvent.record(event, session),
    finish ? null : WhOrder.collection.updateMany({_id: {$in: whOrderIds}}, {$set}, {session})
  ]);

  results.push(eventsResults);

  const deliveredOrdersResults = {
    deleted: [],
    updated: []
  };
  const deliveredOrdersOps = [];

  if (finish)
  {
    deliveredOrders.forEach(deliveredOrder =>
    {
      if (deliveredOrder.qtyDone >= deliveredOrder.quantityTodo)
      {
        return;
      }

      deliveredOrder.quantityDone = deliveredOrder.quantityTodo;
      deliveredOrder.qtyRemaining = 0;
      deliveredOrder.status = 'done';

      deliveredOrdersResults.updated.push(deliveredOrder);
    });
  }
  else
  {
    deliveredOrders.forEach(deliveredOrder =>
    {
      deliveredOrdersResults.deleted.push({_id: deliveredOrder._id});
    });
  }

  if (deliveredOrdersResults.deleted.length)
  {
    deliveredOrdersOps.push({
      deleteMany: {
        filter: {_id: {$in: deliveredOrdersResults.deleted.map(o => o._id)}}
      }
    });
  }

  deliveredOrdersResults.updated.forEach(update =>
  {
    deliveredOrdersOps.push({
      updateOne: {
        filter: {_id: update._id},
        update: {$set: update}
      }
    });
  });

  if (finish && !deliveredOrdersOps.length)
  {
    return {response: {orders: whOrders}};
  }

  if (data.orders)
  {
    const [setCartsResults] = await Promise.all([
      finish ? null : WhSetCart.deleteByWhOrder(whOrders[0]),
      deliveredOrdersOps.length
        ? WhDeliveredOrder.collection.bulkWrite(deliveredOrdersOps, {session})
        : null
    ]);

    const [lineResults, setResults] = await Promise.all([
      WhLine.recountByWhOrder(whOrders[0], !finish, deliveredOrdersOps.length > 0),
      finish ? null : WhOrder.finalizeSet(date, set, setCartsResults, session)
    ]);

    results.push(
      setCartsResults,
      lineResults,
      setResults
    );
  }
  else
  {
    const [setCartsResults] = await Promise.all([
      finish ? null : WhSetCart.deleteBySet(data.date, data.set, session),
      deliveredOrdersOps.length
        ? WhDeliveredOrder.collection.bulkWrite(deliveredOrdersOps, {session})
        : null,
      WhPendingSetAssign.collection.deleteMany({date: data.date, set: data.set}, {session})
    ]);

    const lineResults = await WhLine.recountByWhOrder(whOrders[0], !finish, deliveredOrdersOps.length > 0);

    results.push(
      setCartsResults,
      lineResults
    );
  }

  await session.commitTransaction();

  const orders = whOrders.map(whOrder => whOrder.toJSON());

  if (!finish)
  {
    results.push({
      updated: orders,
      messages: [{
        topic: `${WhOrder.TOPIC_PREFIX}.updated`,
        message: {updated: orders}
      }]
    });
  }

  return {
    response: {orders},
    results
  };
};
