// wh/state/actions/resolveDeliveryAction.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const startDelivery = require('./startDelivery');

module.exports = async (app, module, {data, user, whUser, session}) =>
{
  const {
    OldWhSetCart: WhSetCart
  } = module;

  const deliveryFuncs = module.state.settings.deliveryFuncs;

  if (!Array.isArray(deliveryFuncs)
    || !deliveryFuncs.includes(whUser.func)
    || (data.kind === 'ps' && whUser.func !== 'painter')
    || (data.kind !== 'ps' && whUser.func === 'painter'))
  {
    throw app.createError('Invalid delivery function.', 'INVALID_FUNC', 400);
  }

  const deliveringSetCarts = await WhSetCart
    .find({
      status: 'delivering',
      'deliveringBy.id': whUser._id,
      kind: data.kind
    })
    .select({_id: 1})
    .hint({status: 1, date: -1})
    .lean()
    .session(session)
    .exec();

  if (deliveringSetCarts.length)
  {
    return {
      response: {
        result: 'continueDelivery',
        setCarts: deliveringSetCarts.map(setCart => setCart._id.toString()),
        user: whUser,
        personnelId: data.personnelId
      }
    };
  }

  return await startDelivery.execute(app, module, {
    data, user, whUser, session
  });
};
