// wh/state/actions/resolveDeliveryAction.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const startComponentsDelivery = require('./startComponentsDelivery');
const startPackagingDelivery = require('./startPackagingDelivery');

module.exports = async (app, module, {data, user, whUser, session}) =>
{
  const {
    OldWhSetCart: WhSetCart
  } = module;

  if (whUser.func !== `dist-${data.kind}`)
  {
    throw app.createError('Invalid function.', 'INVALID_FUNC', 400);
  }

  const deliveringSetCarts = await WhSetCart
    .find({
      status: 'delivering',
      'deliveringBy.id': whUser._id,
      kind: data.kind
    })
    .select({_id: 1})
    .lean()
    .session(session)
    .exec();

  if (deliveringSetCarts.length)
  {
    return {
      response: {
        result: 'continueDelivery',
        setCarts: deliveringSetCarts.map(setCart => setCart._id.toString()),
        user: whUser,
        personnelId: data.personnelId
      }
    };
  }

  if (data.kind === 'components')
  {
    return await startComponentsDelivery(app, module, {
      data, user, whUser, session
    });
  }

  return await startPackagingDelivery(app, module, {
    data, user, whUser, session
  });
};
