// wmes-osh/routes/permalink.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module) =>
{
  const {
    express
  } = module;

  const PREFIX_TO_MODULE = {
    Z: 'nearMisses',
    K: 'kaizens',
    A: 'actions',
    O: 'observations'
  };

  express.get(`/r/osh/entry/:rid`, detailsRoute);
  express.get(`/r/osh/:entryModule/:filter`, listRoute);

  function detailsRoute(req, res, next)
  {
    const {rid} = req.params;
    const entryModule = PREFIX_TO_MODULE[rid.split('-')[0]];

    if (!entryModule)
    {
      return next(app.createError('Invalid ID.', 'INPUT', 400));
    }

    res.redirect(`/#osh/${entryModule}/${rid}`);
  }

  function listRoute(req, res, next)
  {
    const {entryModule, filter} = req.params;

    if (!Object.values(PREFIX_TO_MODULE).includes(entryModule))
    {
      return next(app.createError('Invalid entry module.', 'INPUT', 400));
    }

    let query = '';

    if (filter === 'unseen' || filter === 'mine')
    {
      query = `?exclude(changes)&sort(-createdAt)&limit(-1337)&users.user.id=${filter}`;
    }

    res.redirect(`/#osh/${entryModule}${query}`);
  }
};
