// sapGui/importer/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const fs = require('fs');
const path = require('path');

exports.DEFAULT_CONFIG = {
  expressId: 'express',
  secretKey: '',
  importPath: './'
};

exports.start = (app, module) =>
{
  app.onModuleReady(module.config.expressId, () =>
  {
    app[module.config.expressId].post('/sapGui;import', importRoute);
  });

  function importRoute(req, res, next)
  {
    res.type('text/plain');

    if (req.query.secretKey !== module.config.secretKey)
    {
      return res.status(403).send('INVALID_SECRET_KEY');
    }

    if (!req.is('text/plain'))
    {
      return res.status(400).send('INVALID_CONTENT_TYPE');
    }

    const fileName = req.query.fileName;

    if (!/^[a-zA-Z0-9_.-]+$/.test(fileName))
    {
      return res.status(400).send('INVALID_FILE_NAME');
    }

    const timestamp = parseInt(req.query.timestamp, 10);
    const importFileName = (timestamp > 0 ? timestamp : Date.now()) + `@${fileName}`;
    const importFilePath = path.join(module.config.importPath, importFileName);

    fs.writeFile(importFilePath, req.body, err =>
    {
      if (err)
      {
        return next(err);
      }

      module.info('Imported.', {importFileName});

      res.sendStatus(204);
    });
  }
};
