// sapGui/importer/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const fs = require('fs');
const path = require('path');
const {pipeline} = require('stream');
const {createGunzip} = require('zlib');
const multer = require('multer');

exports.DEFAULT_CONFIG = {
  expressId: 'express',
  userId: 'user',
  secretKey: '',
  importPath: './'
};

exports.optionalModules = {
  'user express': (app, module) =>
  {
    const {
      user,
      express
    } = module;

    express.post(
      '/sapGui;import',
      user.auth('ORDERS:MANAGE'),
      multer({dest: module.config.importPath}).single('file'),
      importRoute.bind(null, app, module)
    );
  }
};

function importRoute(app, module, req, res, next)
{
  const {file} = req;

  if (!file)
  {
    return next(app.createError('No file.', 'INPUT', 400));
  }

  if (file.mimetype !== 'application/gzip')
  {
    clean(file);

    return next(app.createError('Invalid type.', 'INPUT', 400));
  }

  pipeline(
    fs.createReadStream(file.path),
    createGunzip(),
    fs.createWriteStream(path.join(module.config.importPath, file.originalname)),
    err =>
    {
      if (err)
      {
        return next(err);
      }

      clean(file);

      res.sendStatus(204);
    }
  );
}

function clean(file)
{
  fs.unlink(file.path, () => {});
}
