// wmes-snf/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const path = require('path');
const fs = require('fs-extra');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  html2pdfId: 'html2pdf',
  snfImagesPath: './'
};

exports.models = [
  require('./models/snfProgram'),
  require('./models/snfTest'),
  require('orders/models/order'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('printing/models/printer'),
  require('counters/models/counter')
];

exports.republishTopics = [
  'snf.programs.**',
  'snf.tests.**'
];

exports.userPrivileges = [
  'SNF:VIEW',
  'SNF:MANAGE'
];

exports.optionalModules = {
  'mongoose user express html2pdf': require('./routes')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    programs: 'SnfProgram'
  };

  app.broker.subscribe('snf.programs.deleted', cleanUpSnfImages);

  function cleanUpSnfImages(message)
  {
    message.model.images.forEach(image =>
    {
      fs.unlink(path.join(module.config.snfImagesPath, `${image._id}.${image.type}`), () => {});
    });
  }
};
