// wmes-snf/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const path = require('path');
const fs = require('fs-extra');
const step = require('h5.step');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  sioId: 'sio',
  expressId: 'express',
  directoryWatcherId: 'directoryWatcher',
  productionId: 'production',
  userId: 'user',
  licensesId: 'licenses',
  mailSenderId: 'mail/sender',
  settingsId: 'settings',
  updaterId: 'updater',
  orgUnitsId: 'orgUnits',
  html2pdfId: 'html2pdf',
  featureDbPath: './',
  zipStoragePath: './',
  vncTemplatePath: `${__dirname}/template.vnc`,
  updatesPath: './',
  snfImagesPath: './'
};

exports.models = [
  require('./models/snfProgram')
];

exports.republishTopics = [
  'snf.programs.**'
];

exports.optionalModules = {
  'mongoose user express': require('./routes')
};

exports.start = (app, module) =>
{
  app.broker.subscribe('snf.programs.deleted', cleanUpSnfImages);

  function cleanUpSnfImages(message)
  {
    message.model.images.forEach(image =>
    {
      fs.unlink(path.join(module.config.snfImagesPath, `${image._id}.${image.type}`), () => {});
    });
  }
};
