// orderDocuments/settings.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

module.exports = (app, module) =>
{
  module.cachedSettings = {
    useCatalog: false,
    path: '',
    extra: [],
    remoteServer: ''
  };

  app.broker.subscribe('settings.updated.orders.documents.**', updateSettings);

  module.settings.findValues('orders.documents.', (err, values) =>
  {
    if (err)
    {
      return module.error(err, 'Failed to find settings.');
    }

    module.cachedSettings.useCatalog = !!values.useCatalog;

    if (_.isString(values.path))
    {
      module.cachedSettings.path = values.path;
    }

    if (_.isString(values.extra))
    {
      module.cachedSettings.extra = parseExtraDocumentsSetting(values.extra);
    }

    if (_.isString(values.remoteServer))
    {
      module.cachedSettings.remoteServer = values.remoteServer;
    }
  });

  function updateSettings(message)
  {
    const {cachedSettings} = module;
    const {_id, value} = message;

    if (_id === 'orders.documents.useCatalog')
    {
      cachedSettings.useCatalog = !!value;
    }
    else if (_id === 'orders.documents.path')
    {
      cachedSettings.path = value;
    }
    else if (_id === 'orders.documents.extra')
    {
      cachedSettings.extra = parseExtraDocumentsSetting(value);

      app.broker.publish('orderDocuments.extraUpdated', cachedSettings.extra);
    }
    else if (_id === 'orders.documents.remoteServer')
    {
      cachedSettings.remoteServer = value;
    }
  }

  function parseExtraDocumentsSetting(rawValue)
  {
    const extra = [];

    if (!_.isString(rawValue) || _.isEmpty(rawValue))
    {
      return extra;
    }

    const namesToDocuments = {};
    let lastName = '';

    _.forEach(rawValue.split('\n'), (line) =>
    {
      const matches = line.match(/([0-9]{15})\s+(.*?)$/);

      if (matches)
      {
        if (lastName === '')
        {
          return;
        }

        namesToDocuments[lastName][matches[1]] = matches[2];
      }
      else
      {
        lastName = line;

        namesToDocuments[lastName] = {};
      }
    });

    _.forEach(namesToDocuments, (documents, names) =>
    {
      names = names
        .split(';')
        .map((name) => name.trim())
        .filter((name) => !!name.length);

      if (!names.length)
      {
        return;
      }

      _.forEach(names, (name) =>
      {
        extra.push({
          pattern: prepareExtraDocumentPattern(name),
          documents
        });
      });
    });

    return extra;
  }

  function prepareExtraDocumentPattern(name)
  {
    try
    {
      return new RegExp(name);
    }
    catch (err)
    {
      return name;
    }
  }
};
