// minutesForSafetyCards/models/minutesForSafetyCard.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const autoIncrement = require('mongoose-plugin-autoinc-fix');
const lastModifiedPaths = require('h5-mongoose/plugins/lastModifiedPaths');

exports.name = 'MinutesForSafetyCard';

exports.setUp = (app, mongoose) =>
{
  const causeSchema = new mongoose.Schema({
    category: String,
    why: [String]
  }, {
    _id: false,
    minimize: false
  });

  const observationSchema = new mongoose.Schema({
    what: String,
    why: String
  }, {
    _id: false,
    minimize: false
  });

  const propositionSchema = new mongoose.Schema({
    what: String,
    who: {},
    when: Date
  }, {
    _id: false,
    minimize: false
  });

  const cardSchema = new mongoose.Schema({
    creator: {},
    createdAt: Date,
    updater: {},
    updatedAt: Date,
    owner: {},
    confirmer: {},
    section: String,
    date: Date,
    subject: String,
    risks: String,
    causes: [causeSchema],
    observations: [observationSchema],
    safeBehavior: String,
    orgPropositions: [propositionSchema],
    techPropositions: [propositionSchema],
    nearMiss: {
      type: Number,
      default: null
    },
    suggestion: {
      type: Number,
      default: null
    },
    participants: {},
    users: [String]
  }, {
    id: false,
    minimize: false
  });

  cardSchema.index({date: -1});
  cardSchema.index({section: 1});
  cardSchema.index({'participants.id': 1});
  cardSchema.index({'owner.id': 1});
  cardSchema.index({'confirmer.id': 1});

  cardSchema.statics.TOPIC_PREFIX = 'minutesForSafetyCards';

  cardSchema.pre('save', function(next)
  {
    if (this.isNew)
    {
      this.createdAt = new Date();
      this.updatedAt = this.createdAt;
    }
    else
    {
      this.updatedAt = new Date();
    }

    const users = {};

    [this.creator, this.updater, this.owner, this.confirmer].concat(this.participants).forEach(user =>
    {
      if (user)
      {
        users[user.id] = 1;
      }
    });

    this.orgPropositions.concat(this.techPropositions).forEach(proposition =>
    {
      (Array.isArray(proposition.who) ? proposition.who : []).forEach(user =>
      {
        if (user)
        {
          users[user.id] = 1;
        }
      });
    });

    this.users = Object.keys(users);

    next();
  });

  cardSchema.plugin(autoIncrement.plugin, {
    model: 'MinutesForSafetyCard',
    field: 'rid',
    startAt: 1,
    incrementBy: 1
  });

  cardSchema.plugin(lastModifiedPaths);

  mongoose.model(exports.name, cardSchema);
};
