// kaizen/durationRecalculator.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const later = require('later');
const moment = require('moment');

module.exports = (app, module) =>
{
  const {KaizenOrder} = module;
  const logger = module.logger.create({submodule: 'durationRecalculator'});

  let inProgress = false;

  app.broker.subscribe('app.started', recalcDurations).setLimit(1);

  later.setInterval(recalcDurations, later.parse.text('at 00:02 am'));

  module.recalcDurations = recalcDurations;

  function recalcDurations(all)
  {
    if (inProgress)
    {
      return;
    }

    const startedAt = Date.now();
    const currentDate = moment().startOf('day').toDate();

    logger.info('Started...');

    inProgress = true;

    const conditions = {
      status: {
        $in: ['new', 'accepted', 'todo', 'inProgress', 'paused']
      }
    };
    const fields = {
      rid: 1,
      types: 1,
      status: 1,
      eventDate: 1,
      kaizenStartDate: 1,
      kaizenFinishDate: 1,
      finishedAt: 1,
      finishDuration: 1,
      kaizenDuration: 1
    };

    if (all === true)
    {
      delete conditions.status;
    }

    const stream = KaizenOrder.find(conditions, fields).lean().cursor();

    stream.on('error', function(err)
    {
      inProgress = false;

      logger.error(err, 'Failed to recalc orders.');
    });

    stream.on('data', function(doc)
    {
      recalcNext(doc, currentDate);
    });

    stream.on('end', function()
    {
      inProgress = false;

      logger.info(`Done in ${Date.now() - startedAt} ms.`);
    });
  }

  function recalcNext(doc, currentDate)
  {
    const newKaizenDuration = KaizenOrder.recalcKaizenDuration(doc, currentDate);
    const newFinishDuration = KaizenOrder.recalcFinishDuration(doc, currentDate);
    let changed = false;
    const changes = {};

    if (newKaizenDuration !== doc.kaizenDuration)
    {
      changed = true;
      changes.kaizenDuration = newKaizenDuration;
    }

    if (newFinishDuration !== doc.finishDuration)
    {
      changed = true;
      changes.finishDuration = newFinishDuration;
    }

    if (!changed)
    {
      return;
    }

    KaizenOrder.updateOne({_id: doc._id}, {$set: changes}, (err) =>
    {
      if (err)
      {
        module.error(err, `Failed to recalc [${doc.rid}].`);
      }
    });
  }
};
