// kaizen/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  sioId: 'sio',
  userId: 'user',
  mailSenderId: 'mail/sender',
  reportsId: 'reports',
  settingsId: 'settings',
  attachmentsDest: null,
  multiType: true,
  remind: 14
};

exports.models = [
  require('./models/kaizenArea'),
  require('./models/kaizenBehaviour'),
  require('./models/kaizenCategory'),
  require('./models/kaizenCause'),
  require('./models/kaizenOrder'),
  require('./models/kaizenProductFamily'),
  require('./models/kaizenRisk'),
  require('./models/kaizenSection'),
  require('user/models/user')
];

exports.recordTopics = {
  debug: ['kaizen.*.added', 'kaizen.*.edited'],
  warning: ['kaizen.*.deleted'],
  blacklist: ['kaizen.orders.added', 'kaizen.orders.edited']
};

exports.republishTopics = ['kaizen.*.added', 'kaizen.*.edited', 'kaizen.*.deleted', 'kaizen.orders.seen.*'];

exports.userPrivileges = [
  'KAIZEN:MANAGE',
  'KAIZEN:DICTIONARIES:VIEW', 'KAIZEN:DICTIONARIES:MANAGE'
];

exports.optionalModules = {
  'mongoose': [
    require('./durationRecalculator'),
    require('./stats'),
    require('./observerCleaner')
  ],
  'mongoose mailSender settings': [
    require('./notifier'),
    require('./reminder')
  ],
  'mongoose sio user': require('./commands'),
  'mongoose user reports settings express': require('./routes')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    sections: 'KaizenSection',
    areas: 'KaizenArea',
    categories: 'KaizenCategory',
    causes: 'KaizenCause',
    risks: 'KaizenRisk',
    productFamilies: 'KaizenProductFamily',
    behaviours: 'KaizenBehaviour'
  };
};
