// kaizen/reminder.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const fs = require('fs');
const _ = require('lodash');
const step = require('h5.step');
const ejs = require('ejs');
const moment = require('moment');

module.exports = (app, module) =>
{
  const DAYS_AGO = module.config.remind;

  if (!DAYS_AGO)
  {
    return;
  }

  const {mongoose, mailSender, User, KaizenOrder} = module;
  const logger = module.logger.create({
    submodule: 'reminder'
  });

  const emailTemplateFile = `${__dirname}/templates/reminder.email.pl.ejs`;
  const renderEmail = ejs.compile(fs.readFileSync(emailTemplateFile, 'utf8'), {
    cache: true,
    filename: emailTemplateFile,
    compileDebug: false,
    rmWhitespace: true
  });

  app.broker.subscribe('app.started', scheduleNextReminder).setLimit(1);

  function scheduleNextReminder()
  {
    const delay = moment().hours(12).startOf('hour').add(1, 'days').diff(Date.now());

    setTimeout(remind, delay);
  }

  function remind()
  {
    step(
      function aggregateStep()
      {
        const weekAgo = moment().subtract(DAYS_AGO, 'days').valueOf();

        this.conditions = {
          status: {$in: ['new', 'accepted', 'todo', 'inProgress', 'paused']},
          remindedAt: {$lt: weekAgo}
        };

        KaizenOrder.aggregate([
          {$match: this.conditions},
          {$group: {_id: '$confirmer.id', totalCount: {$sum: 1}}}
        ], this.parallel());

        KaizenOrder.aggregate([
          {$match: {
            status: {$in: ['new', 'accepted', 'todo', 'inProgress', 'paused']},
            updatedAt: {$lt: new Date(weekAgo)}
          }},
          {$group: {_id: '$confirmer.id', totalCount: {$sum: 1}}}
        ], this.parallel());
      },
      function getUserDataStep(err, remindedAtResults, updatedAtResults)
      {
        if (err)
        {
          return this.skip(err);
        }

        const updatedAtStats = {};

        _.forEach(updatedAtResults, (result) =>
        {
          updatedAtStats[result._id] = result;
        });

        const userIds = [];
        const userIdToStats = {};

        _.forEach(remindedAtResults, (result) =>
        {
          userIds.push(new mongoose.Types.ObjectId(result._id));
          userIdToStats[result._id] = updatedAtStats[result._id];
        });

        this.userIdToStats = userIdToStats;

        const conditions = {_id: {$in: userIds}, email: {$ne: null}};
        const fields = {firstName: 1, lastName: 1, email: 1, gender: 1};

        User.find(conditions, fields).exec(this.next());
      },
      function sendEmailsStep(err, users)
      {
        if (err)
        {
          return this.skip(err);
        }

        for (let i = 0; i < users.length; ++i)
        {
          const user = users[i];

          sendReminderEmail(users[i], this.userIdToStats[user._id]);
        }

        logger.info(`Reminded [${users.length}] users about incomplete orders.`);
      },
      function updateUpdatedAtStep()
      {
        KaizenOrder.updateMany(this.conditions, {$set: {remindedAt: Date.now()}}, this.next());
      },
      function(err)
      {
        if (err)
        {
          logger.error(err, `Failed to remind.`);
        }

        this.conditions = null;
        this.userIdToStats = null;

        scheduleNextReminder();
      }
    );
  }

  function sendReminderEmail(user, stats)
  {
    const mailOptions = {
      to: user.email,
      replyTo: user.email,
      subject: `[WMES] [ZPW] Niezakończone zgłoszenia`,
      html: renderEmail({
        urlPrefix: app.options.emailUrlPrefix,
        user: user,
        plural: plural(stats.totalCount),
        totalCount: stats.totalCount
      })
    };

    mailSender.send(mailOptions, (err) =>
    {
      if (err)
      {
        module.error(err, `Failed to remind about incomplete near misses.`, {recipients: mailOptions.to});
      }
    });
  }

  function plural(n)
  {
    return (n % 10 < 5) && (n % 10 > 1) && ((~~(n / 10) % 10) !== 1);
  }
};
