// prodDowntimes/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const validateOverlappingDowntimes = require('./validateOverlappingDowntimes');
const addProdDowntime = require('./addProdDowntime');
const editProdDowntime = require('./editProdDowntime');
const deleteProdDowntime = require('./deleteProdDowntime');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  sioId: 'sio',
  productionId: 'production',
  aorsId: 'aors',
  downtimeReasonsId: 'downtimeReasons',
  orgUnitsId: 'orgUnits',
  settingsId: 'settings'
};

exports.recordTopics = {
  warning: ['prodDowntimes.confirmedEdited'],
  blacklist: []
};

exports.republishTopics = ['prodDowntimes.**'];

exports.userPrivileges = ['PROD_DOWNTIMES:VIEW', 'PROD_DOWNTIMES:MANAGE', 'PROD_DOWNTIMES:ALL'];

exports.models = [
  require('./models/prodDowntime'),
  require('prodChangeRequests/models/prodChangeRequest'),
  require('prodLogEntries/models/prodLogEntry')
];

exports.optionalModules = {
  'mongoose user express aors downtimeReasons orgUnits production settings': require('./routes'),
  'sio production': require('./commands'),
  'mongoose settings orgUnits production': require('./autoConfirmation')
};

exports.start = (app, module) =>
{
  module.validateOverlappingDowntimes = validateOverlappingDowntimes.bind(null, app, module);
  module.addProdDowntime = addProdDowntime.bind(null, app, module);
  module.editProdDowntime = editProdDowntime.bind(null, app, module);
  module.deleteProdDowntime = deleteProdDowntime.bind(null, app, module);

  app.onModuleReady(
    [
      module.config.productionId
    ],
    setUpConfirmedDowntimeEditWarning
  );

  function setUpConfirmedDowntimeEditWarning()
  {
    app.broker.subscribe('prodDowntimes.updated.*', checkConfirmedDowntimeEdit);
  }

  function checkConfirmedDowntimeEdit(partialProdDowntime)
  {
    if (!Array.isArray(partialProdDowntime.changes))
    {
      return;
    }

    const lastChange = partialProdDowntime.changes[partialProdDowntime.changes.length - 1];

    if (!lastChange || !lastChange.data || !lastChange.user || !lastChange.user.id)
    {
      return;
    }

    const data = lastChange.data;

    if (data.status && data.status[1] === 'confirmed')
    {
      return;
    }

    app[module.config.productionId].getProdData('downtime', partialProdDowntime._id, (err, prodDowntime) =>
    {
      if (err || !prodDowntime || prodDowntime.status !== 'confirmed')
      {
        return;
      }

      app.broker.publish('prodDowntimes.confirmedEdited', {
        _id: prodDowntime._id,
        rid: prodDowntime.rid,
        user: lastChange.user
      });
    });
  }
};
