// wmes-fap/models/fapSubCategory.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'FapSubCategory';

exports.setUp = (app, mongoose) =>
{
  const notificationSchema = new mongoose.Schema({
    subdivisions: [{
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Subdivision'
    }],
    prodFunctions: [{
      type: String,
      ref: 'ProdFunction'
    }]
  }, {
    _id: false,
    minimize: false
  });

  const subCategorySchema = new mongoose.Schema({
    active: {
      type: Boolean,
      required: true
    },
    parent: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'FapCategory',
      required: true
    },
    name: {
      type: String,
      trim: true,
      required: true
    },
    users: [{}],
    notifications: [notificationSchema],
    etoCategory: {
      type: String,
      default: null
    },
    planners: {
      type: Boolean,
      default: false
    },
    reqFields: {}
  }, {
    id: false
  });

  subCategorySchema.statics.TOPIC_PREFIX = 'fap.subCategories';
  subCategorySchema.statics.BROWSE_LIMIT = 1000;

  mongoose.model(exports.name, subCategorySchema);
};
