// subdivisions/models/subdivision.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'Subdivision';

exports.setUp = (app, mongoose) =>
{
  const autoDowntimeTimeSchema = new mongoose.Schema({
    h: {
      type: Number,
      min: 0,
      max: 23
    },
    m: {
      type: Number,
      min: 0,
      max: 59
    },
    d: {
      type: Number,
      min: 0,
      max: 8 * 60,
      default: 0
    }
  }, {
    _id: false
  });

  const autoDowntimeSchema = new mongoose.Schema({
    reason: {
      type: String,
      ref: 'DowntimeReason',
      required: true
    },
    when: {
      type: String,
      enum: ['initial', 'always', 'after', 'time'],
      default: 'always'
    },
    after: {
      type: String,
      ref: 'DowntimeReason',
      default: ''
    },
    time: [autoDowntimeTimeSchema]
  }, {
    _id: false
  });

  const subdivisionSchema = new mongoose.Schema({
    division: {
      type: String,
      ref: 'Division',
      required: true
    },
    type: {
      type: String,
      enum: [
        'assembly',
        'press',
        'storage',
        'paintShop',
        'other'
      ],
      default: 'assembly'
    },
    name: {
      type: String,
      trim: true,
      required: true
    },
    prodTaskTags: [String],
    aor: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Aor',
      default: null
    },
    autoDowntimes: [autoDowntimeSchema],
    deactivatedAt: {
      type: Date,
      default: null
    }
  }, {
    id: false
  });

  subdivisionSchema.statics.TOPIC_PREFIX = 'subdivisions';
  subdivisionSchema.statics.BROWSE_LIMIT = 1000;

  subdivisionSchema.pre('remove', function(next)
  {
    mongoose.model('MrpController').findOne({subdivision: this._id}).select({_id: 1}).lean().exec((err, child) =>
    {
      if (err)
      {
        return next(err);
      }

      if (child)
      {
        return next(app.createError(`Org unit not empty: ${child._id}`, 'NOT_EMPTY', 400));
      }

      next();
    });
  });

  mongoose.model(exports.name, subdivisionSchema);
};
