// settings/models/setting.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

exports.name = 'Setting';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    _id: {
      type: String,
      required: true,
      trim: true
    },
    value: {},
    updatedAt: Date,
    updater: {}
  }, {
    id: false
  });

  schema.statics.TOPIC_PREFIX = 'settings';
  schema.statics.BROWSE_LIMIT = 1000;

  schema.statics.findValues = function(conditions, ns, done)
  {
    if (typeof conditions === 'string' && (!ns || typeof ns === 'function'))
    {
      done = ns;
      ns = conditions;
      conditions = {_id: new RegExp(`^${_.escapeRegExp(ns)}`)};
    }

    const find = this.find(conditions).select({value: 1}).lean();

    if (!done)
    {
      return find.then(prepareValues);
    }

    find.then(settings => done(null, prepareValues(settings)), done);

    function prepareValues(settings)
    {
      const values = {};

      settings.forEach(setting =>
      {
        values[setting._id.replace(ns, '')] = setting.value;
      });

      return values;
    }
  };

  return schema;
};
