// sio/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const {exec} = require('child_process');
const _ = require('lodash');
const engineIo = require('engine.io');
const socketIo = require('socket.io');
const SocketIoMultiServer = require('./SocketIoMultiServer');
const setUpRoutes = require('./routes');
let pmx = null;

try { pmx = require('pmx'); }
catch (err) {} // eslint-disable-line no-empty

engineIo.Server.errors.SERVER_UNAVAILABLE = 503;
engineIo.Server.errorMessages[503] = 'Server unavailable';

exports.DEFAULT_CONFIG = {
  httpServerIds: ['httpServer'],
  expressId: 'express',
  userId: 'user',
  pingsId: 'pings',
  path: '/sio',
  socketIo: {
    pingInterval: 30000,
    pingTimeout: 10000
  },
  netstatCmd: ''
};

exports.start = (app, module, done) =>
{
  module.config.socketIo = Object.assign({}, module.config.socketIo, {
    path: module.config.path,
    transports: ['websocket', 'xhr-polling'],
    serveClient: false,
    allowRequest: checkRequest
  });

  let socketCount = 0;
  const probes = {
    currentUsersCounter: null,
    totalConnectionTime: null,
    totalConnectionCount: null
  };

  if (pmx)
  {
    const pmxProbe = pmx.probe();

    probes.currentUsersCounter = pmxProbe.counter({name: 'sio:currentUsers'});
    probes.totalConnectionTime = pmxProbe.histogram({name: 'sio:totalConnectionTime', measurement: 'sum'});
    probes.totalConnectionCount = pmxProbe.histogram({name: 'sio:totalConnectionCount', measurement: 'sum'});
  }

  const multiServer = new SocketIoMultiServer();

  app.onModuleReady(module.config.httpServerIds, () =>
  {
    _.forEach(module.config.httpServerIds, (httpServerId) =>
    {
      multiServer.addServer(app[httpServerId].server);
    });

    startSocketIo();

    done();
  });

  function checkRequest(req, done)
  {
    const allServersAvailable = module.config.httpServerIds.every(httpServerId => app[httpServerId].isAvailable());

    if (!allServersAvailable)
    {
      return done(engineIo.Server.errors.SERVER_UNAVAILABLE, false);
    }

    return module.checkRequest(req, done);
  }

  function startSocketIo()
  {
    const sio = socketIo(multiServer, module.config.socketIo);

    module = app[module.name] = Object.assign(sio, module);

    app.onModuleReady(
      [
        module.config.userId,
        module.config.expressId
      ],
      setUpRoutes.bind(null, app, module)
    );

    sio.sockets.setMaxListeners(25);

    module.on('connection', (socket) =>
    {
      ++socketCount;

      socket.handshake.connectedAt = Date.now();

      if (pmx)
      {
        probes.currentUsersCounter.inc();
      }

      if (app[module.config.pingsId])
      {
        app[module.config.pingsId].recordHttpRequest(socket.conn.request, socket.handshake.user);
      }

      app.broker.publish('sockets.connected', {
        socket: {
          _id: socket.id,
          headers: socket.handshake.headers || {},
          user: socket.handshake.user || {}
        },
        socketCount: socketCount
      });

      socket.on('disconnect', function()
      {
        --socketCount;

        if (pmx)
        {
          probes.totalConnectionCount.update(1);
          probes.totalConnectionTime.update((Date.now() - socket.handshake.connectedAt) / 1000);
          probes.currentUsersCounter.dec();
        }

        app.broker.publish('sockets.disconnected', {
          socketId: socket.id,
          socketCount: socketCount
        });
      });

      socket.on('echo', function()
      {
        socket.emit.apply(socket, ['echo'].concat(Array.prototype.slice.call(arguments)));
      });

      socket.on('time', function(reply)
      {
        if (_.isFunction(reply))
        {
          reply(Date.now(), 'Europe/Warsaw');
        }
      });

      socket.on('killZombies', function()
      {
        if (socket.handshake && socket.handshake.user && socket.handshake.user.super)
        {
          killZombies();
        }
      });
    });
  }

  function killZombies()
  {
    netstat((err, established) => // eslint-disable-line handle-callback-err
    {
      const oldSocketIds = Object.keys(module.sockets.connected);

      oldSocketIds.forEach(socketId =>
      {
        const ioSocket = module.sockets.connected[socketId];
        const nodeSocket = ioSocket.request.socket;
        const foreignAddress = `${nodeSocket.remoteAddress}:${nodeSocket.remotePort}`;

        if (!established.has(foreignAddress)
          || !nodeSocket.readable
          || !nodeSocket.writable
          || nodeSocket.destroyed)
        {
          ioSocket.disconnect(true);
        }
      });

      const newSocketIds = Object.keys(module.sockets.connected);

      if (newSocketIds.length < oldSocketIds.length)
      {
        module.info(`Killed ${oldSocketIds.length - newSocketIds.length} zombies!`);
      }
    });
  }

  function netstat(done)
  {
    const win32 = process.platform === 'win32';
    const cmd = module.config.netstatCmd || (win32 ? 'netstat -p tcp -n' : 'netstat -ant | grep ESTABLISHED');

    exec(cmd, (err, stdout) =>
    {
      const established = new Set();
      const partIndex = win32 ? 2 : 4;

      (stdout || '').trim().split('\n').forEach(line =>
      {
        const parts = line.replace(/\s+/g, ' ').trim().split(' ');
        const foreignAddress = parts[partIndex];
        const state = parts[partIndex + 1];

        if (state === 'ESTABLISHED')
        {
          established.add(foreignAddress);
        }
      });

      done(err, established);
    });
  }
};
