// cags/importer/nc12.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const path = require('path');
const moment = require('moment');
const step = require('h5.step');
const fs = require('fs-extra');
const parseSapTextTable = require('sap/util/parseSapTextTable');

exports.DEFAULT_CONFIG = {
  filterRe: /^T_ZSE16D_MARA\.txt$/,
  resultFile: 'nc12_to_cags.json',
  parsedOutputDir: null
};

exports.recordTopics = {
  info: ['cags.nc12.synced'],
  error: ['cags.nc12.syncFailed']
};

exports.start = (app, module) =>
{
  const filePathCache = {};
  const queue = [];
  let locked = false;

  app.broker.subscribe('directoryWatcher.changed', queueFile).setFilter(filterFile);

  function filterFile(fileInfo)
  {
    if (filePathCache[fileInfo.filePath] || !module.config.filterRe.test(fileInfo.fileName))
    {
      return false;
    }

    fileInfo.timeKey = createTimeKey(fileInfo.timestamp);

    return true;
  }

  function createTimeKey(timestamp)
  {
    return moment(timestamp).subtract(1, 'days').format('YYMMDDHH');
  }

  function queueFile(fileInfo)
  {
    filePathCache[fileInfo.filePath] = true;

    queue.push(fileInfo);

    module.info(`Queued...`, {key: fileInfo.timeKey});

    setImmediate(importNext);
  }

  function importNext()
  {
    if (locked)
    {
      return;
    }

    const fileInfo = queue.shift();

    if (!fileInfo)
    {
      return;
    }

    locked = true;

    const startTime = Date.now();

    module.info(`Importing...`, {key: fileInfo.timeKey});

    importFile(fileInfo, function(err, count)
    {
      cleanUpFileInfoFile(fileInfo);

      if (err)
      {
        module.error(err, 'Import failed.', {key: fileInfo.timeKey});

        app.broker.publish('cags.nc12.syncFailed', {
          timestamp: fileInfo.timestamp,
          error: err.message
        });
      }
      else
      {
        module.info(`Imported.`, {
          key: fileInfo.timeKey,
          count,
          duration: Date.now() - startTime
        });

        app.broker.publish('cags.nc12.synced', {
          timestamp: fileInfo.timestamp,
          count: count
        });
      }

      locked = false;

      setImmediate(importNext);
    });
  }

  function importFile(fileInfo, done)
  {
    step(
      function readFileStep()
      {
        fs.readFile(fileInfo.filePath, {encoding: 'utf8'}, this.next());
      },
      function parseFileStep(err, fileContents)
      {
        if (err)
        {
          return this.skip(err);
        }

        module.info(`Parsing...`, {key: fileInfo.timeKey, fileSize: fileContents.length});

        const nc12ToCag = {};
        const t = Date.now();

        this.count = parseNc12ToCagTable(fileContents, nc12ToCag);

        module.info(`Parsed!`, {
          key: fileInfo.timeKey,
          count: this.count,
          duration: Date.now() - t
        });

        setImmediate(this.next(), nc12ToCag);
      },
      function writeResultFileStep(nc12ToCag)
      {
        fs.writeFile(module.config.resultFile, JSON.stringify(nc12ToCag), this.next());
      },
      function finalizeStep(err)
      {
        return done(err, this.count);
      }
    );
  }

  function cleanUpFileInfoFile(fileInfo)
  {
    setTimeout(removeFilePathFromCache, 15000, fileInfo.filePath);

    if (module.config.parsedOutputDir)
    {
      moveFileInfoFile(fileInfo.filePath);
    }
    else
    {
      deleteFileInfoFile(fileInfo.filePath);
    }
  }

  function moveFileInfoFile(oldFilePath)
  {
    const newFilePath = path.join(module.config.parsedOutputDir, path.basename(oldFilePath));

    fs.move(oldFilePath, newFilePath, {overwrite: true}, function(err)
    {
      if (err)
      {
        module.error(err, `Failed to rename file.`, {oldFilePath, newFilePath});
      }
    });
  }

  function deleteFileInfoFile(filePath)
  {
    fs.unlink(filePath, function(err)
    {
      if (err)
      {
        module.error(err, `Failed to delete file.`, {filePath});
      }
    });
  }

  function removeFilePathFromCache(filePath)
  {
    delete filePathCache[filePath];
  }

  function parseNc12ToCagTable(input, nc12ToCag)
  {
    let count = 0;

    parseSapTextTable(input, {
      columnMatchers: {
        nc12: /^Material$/,
        cag: /^Prod.*?Hier/i
      },
      valueParsers: {
        nc12: function(input) { return input.replace(/^0+/, ''); },
        cag: function(input) { return input.substr(-6); }
      },
      itemDecorator: function(obj)
      {
        ++count;

        if (obj.nc12.length === 12 && obj.cag.length === 6)
        {
          nc12ToCag[obj.nc12] = obj.cag;
        }

        return null;
      }
    });

    return count;
  }
};
