// wh/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  updaterId: 'updater',
  html2pdfId: 'html2pdf',
  productionId: 'production',
  ordersId: 'orders',
  generator: false
};

exports.models = [
  require('./models/oldWhEvent'),
  require('./models/oldWhOrder'),
  require('./models/oldWhUser'),
  require('./models/oldWhLine'),
  require('./models/oldWhLineSnapshot'),
  require('./models/oldWhSetCart'),
  require('./models/oldWhDowntime'),
  require('./models/oldWhDeliveredOrder'),
  require('orders/models/order'),
  require('planning/models/plan'),
  require('planning/models/planSettings'),
  require('user/models/user'),
  require('printing/models/printer'),
  require('prodShiftOrders/models/prodShiftOrder'),
  require('prodLines/models/prodLine'),
  require('paintShop/models/paintShopOrder'),
  require('paintShop/models/paintShopEvent'),
  require('paintShop/models/paintShopIgnoredComponent')
];

exports.userPrivileges = [
  'WH:VIEW',
  'WH:MANAGE', 'WH:MANAGE:USERS', 'WH:MANAGE:CARTS',
  'WH:SOLVER'
];

exports.republishTopics = ['old.wh.**'];

exports.optionalModules = {
  'mongoose settings': require('./generator'),
  'mongoose user settings html2pdf': require('./state'),
  'mongoose user updater settings express': require('./routes')
};
