// wh/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  updaterId: 'updater',
  html2pdfId: 'html2pdf',
  generator: false
};

exports.models = [
  require('./models/whEvent'),
  require('./models/whOrder'),
  require('./models/whUser'),
  require('orders/models/order'),
  require('planning/models/plan'),
  require('planning/models/planSettings'),
  require('user/models/user'),
  require('printing/models/printer')
];

exports.userPrivileges = ['WH:VIEW', 'WH:MANAGE', 'WH:SOLVER', 'WH:MANAGE:USERS'];

exports.republishTopics = ['wh.**'];

exports.optionalModules = {
  'mongoose settings': require('./generator'),
  'mongoose user settings html2pdf': require('./state'),
  'mongoose user updater settings express': require('./routes')
};

exports.start = (app, module) =>
{
  app.broker.subscribe('orders.updated.*', onOrderUpdated);

  function onOrderUpdated(message)
  {
    if (!module.state || !message.change.newValues.psStatus)
    {
      return;
    }

    const input = {
      action: 'updatePsStatus',
      data: {
        orderNo: message._id,
        psStatus: message.change.newValues.psStatus
      }
    };

    module.state.act(input, err =>
    {
      if (err)
      {
        module.error(err, 'Failed to update PS status.', {input});
      }
    });
  }
};
