// orderDocuments/models/orderDocumentFile.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'OrderDocumentFile';

exports.setUp = (app, mongoose) =>
{
  const fileSchema = new mongoose.Schema({
    hash: String,
    type: String,
    date: Date,
    updater: {},
    updatedAt: Date
  }, {
    _id: false
  });

  const componentSchema = new mongoose.Schema({
    nc12: String,
    name: String,
    searchName: String
  }, {
    _id: false
  });

  const orderDocumentFileSchema = new mongoose.Schema({
    _id: String,
    name: String,
    folders: [String],
    files: [fileSchema],
    oldFolders: {
      type: [String],
      default: null
    },
    components: [componentSchema],
    updater: {},
    updatedAt: Date
  }, {
    id: false
  });

  orderDocumentFileSchema.statics.BROWSE_LIMIT = 2000;

  orderDocumentFileSchema.index({folders: 1});
  orderDocumentFileSchema.index({'components.nc12': 1});
  orderDocumentFileSchema.index({'components.searchName': 1});

  mongoose.model(exports.name, orderDocumentFileSchema);
};
