// wmes-fap/routes/countReport.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const countReport = require('../reports/count');

module.exports = (app, {reports}, req, res, next) =>
{
  const query = req.query;
  const options = {
    fromTime: reports.helpers.getTime(query.from) || null,
    toTime: reports.helpers.getTime(query.to) || null,
    interval: reports.helpers.getInterval(query.interval, 'month'),
    categories: _.isEmpty(query.categories) ? [] : query.categories.split(','),
    subCategories: _.isEmpty(query.subCategories) ? [] : query.subCategories.split(','),
    subdivisionTypes: _.isEmpty(query.subdivisionTypes) ? [] : query.subdivisionTypes.split(','),
    divisions: _.isEmpty(query.divisions) ? [] : query.divisions.split(','),
    mrps: _.isEmpty(query.mrps) ? [] : query.mrps.split(','),
    levels: parseInt(query.levels, 10) || 0
  };

  reports.helpers.generateReport(
    app,
    reports,
    countReport,
    'wmes-fap/count',
    req.reportHash,
    options,
    (err, reportJson) =>
    {
      if (err)
      {
        return next(err);
      }

      res.type('json');
      res.send(reportJson);
    }
  );
};
