// wmes-fap/routes/rebuildSearch.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const _ = require('lodash');

module.exports = (app, {lockEntry, FapEntry}, req, res, next) =>
{
  let totalCount = 0;
  const complete = _.once(err =>
  {
    if (err)
    {
      next(err);
    }
    else
    {
      res.json({totalCount});
    }
  });

  const cursor = FapEntry
    .find({})
    .select({
      productName: 1,
      problem: 1,
      solution: 1,
      why5: 1,
      solutionSteps: 1,
      'changes.comment': 1
    })
    .cursor({batchSize: 10});

  cursor.on('error', complete);
  cursor.on('end', complete);

  cursor.eachAsync(entry =>
  {
    totalCount += 1;

    return new Promise((resolve, reject) =>
    {
      step(
        function()
        {
          this.unlock = lockEntry(entry._id, this.next());
        },
        function()
        {
          FapEntry.prepareSearchText(entry, this.next());
        },
        function(err, search)
        {
          if (err)
          {
            return this.skip(err);
          }

          FapEntry.collection.updateOne({_id: entry._id}, {$set: {search}}, this.next());
        },
        function(err)
        {
          this.unlock();

          if (err)
          {
            reject(err);
          }
          else
          {
            resolve();
          }
        }
      );
    });
  }, complete);
};
