// wh/routes/exportSnapshots.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module, req, res, next) =>
{
  const {
    express,
    OldWhLineSnapshot: WhLineSnapshot
  } = module;

  req.rql.sort = {time: 1};

  express.crud.exportRoute(app, {
    filename: 'WMES_WH_LINE_SNAPSHOTS',
    freezeRows: 1,
    freezeColumns: 2,
    columns: {
      time: 'datetime',
      line: 10,
      'pickup.started.sets': 'integer',
      'pickup.started.qty': 'integer',
      'pickup.started.time': 'decimal',
      'pickup.finished.sets': 'integer',
      'pickup.finished.qty': 'integer',
      'pickup.finished.time': 'decimal',
      'available.qty': 'integer',
      'available.time': 'decimal',
      redirLine: 10,
      working: 'boolean',
      planned: 'boolean',
      nextShiftAt: 'datetime+utc',
      startedPlan: 'date+utc',
      mrps: 40
    },
    model: WhLineSnapshot,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    const {time, data} = doc;

    return {
      time,
      line: data._id,
      'pickup.started.sets': data.pickup.started.sets,
      'pickup.started.qty': data.pickup.started.qty,
      'pickup.started.time': Math.round(data.pickup.started.time / 3600000 * 100) / 100,
      'pickup.finished.sets': data.pickup.finished.sets,
      'pickup.finished.qty': data.pickup.finished.qty,
      'pickup.finished.time': Math.round(data.pickup.finished.time / 3600000 * 100) / 100,
      'available.qty': data.available.qty,
      'available.time': Math.round(data.available.time / 3600000 * 100) / 100,
      redirLine: data.redirLine,
      working: data.working,
      planned: data.planned,
      nextShiftAt: data.nextShiftAt,
      startedPlan: data.startedPlan,
      mrps: data.mrps.join(',')
    };
  }
};
