// wh/routes/exportDowntimes.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = async (app, module, req, res, next) =>
{
  const {
    express,
    DowntimeReason,
    OldWhDowntime: WhDowntime
  } = module;

  const reasons = {};

  try
  {
    const downtimeReasons = await DowntimeReason
      .find({
        subdivisionTypes: {$in: ['wh-pickup', 'wh-dist']}
      })
      .select({
        label: 1
      })
      .lean()
      .exec();

    downtimeReasons.forEach(reason =>
    {
      reasons[reason._id] = reason.label;
    });
  }
  catch (err)
  {
    return next(err);
  }

  express.crud.exportRoute(app, {
    filename: 'WMES_WH_DOWNTIMES',
    freezeRows: 1,
    freezeColumns: 2,
    columns: {
      reason: 25,
      startedAt: 'datetime',
      finishedAt: 'datetime',
      duration: 'decimal',
      user: 20,
      comment: 40
    },
    model: WhDowntime,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    return {
      reason: reasons[doc.reason] || doc.reason,
      startedAt: doc.startedAt,
      finishedAt: doc.finishedAt,
      duration: Math.round(doc.duration / 60000 * 100) / 100,
      user: doc.user.label,
      comment: doc.comment
    };
  }
};
