// wh/routes/exportOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module, req, res, next) =>
{
  const {
    express,
    OldWhOrder: WhOrder
  } = module;

  express.crud.exportRoute(app, {
    filename: 'WMES_WH_ORDERS',
    freezeRows: 1,
    freezeColumns: 0,
    columns: {
      date: 'date+utc',
      set: 'integer',
      orderNo: 10,
      mrp: 5,
      qty: 'integer',
      startTime: 'datetime+utc',
      finishTime: 'datetime+utc',
      status: 10,
      startedAt: 'datetime',
      finishedAt: 'datetime',
      duration: 'decimal',
      line: 10,
      lines: 15,
      'fmx.user': 20,
      'fmx.status': 15,
      'fmx.picklist': 15,
      'fmx.pickup': 15,
      'fmx.startedAt': 'datetime',
      'fmx.finishedAt': 'datetime',
      'fmx.duration': 'decimal',
      'fmx.carts': 15,
      'fmx.problemArea': 10,
      'kitter.user': 20,
      'kitter.status': 15,
      'kitter.picklist': 15,
      'kitter.pickup': 15,
      'kitter.startedAt': 'datetime',
      'kitter.finishedAt': 'datetime',
      'kitter.duration': 'decimal',
      'kitter.carts': 15,
      'kitter.problemArea': 10,
      'platformer.user': 20,
      'platformer.status': 15,
      'platformer.picklist': 15,
      'platformer.pickup': 15,
      'platformer.startedAt': 'datetime',
      'platformer.finishedAt': 'datetime',
      'platformer.duration': 'decimal',
      'platformer.carts': 15,
      'platformer.problemArea': 10,
      'packer.user': 20,
      'packer.status': 15,
      'packer.picklist': 15,
      'packer.pickup': 15,
      'packer.startedAt': 'datetime',
      'packer.finishedAt': 'datetime',
      'packer.duration': 'decimal',
      'packer.carts': 15,
      'packer.problemArea': 10,
      'painter.user': 20,
      'painter.status': 15,
      'painter.picklist': 15,
      'painter.pickup': 15,
      'painter.startedAt': 'datetime',
      'painter.finishedAt': 'datetime',
      'painter.duration': 'decimal',
      'painter.carts': 15,
      'painter.problemArea': 10
    },
    model: WhOrder,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    return {
      date: doc.date,
      set: doc.set,
      orderNo: doc.order,
      mrp: doc.mrp,
      qty: doc.qty,
      startTime: doc.startTime,
      finishTime: doc.finishTime,
      status: doc.status,
      startedAt: doc.startedAt,
      finishedAt: doc.finishedAt,
      duration: doc.finishedAt ? (Math.round((doc.finishedAt - doc.startedAt) / 3600000 * 100) / 100) : 0,
      line: doc.line,
      lines: doc.lines.map(l => l._id).join(','),
      'fmx.user': doc.funcs[0].user ? doc.funcs[0].user.label : '',
      'fmx.status': doc.funcs[0].status,
      'fmx.picklist': doc.funcs[0].picklist,
      'fmx.pickup': doc.funcs[0].pickup,
      'fmx.startedAt': doc.funcs[0].startedAt,
      'fmx.finishedAt': doc.funcs[0].finishedAt,
      'fmx.duration': doc.funcs[0].finishedAt
        ? (Math.round((doc.funcs[0].finishedAt - doc.funcs[0].startedAt) / 3600000 * 100) / 100) : 0,
      'fmx.carts': doc.funcs[0].carts.join(','),
      'fmx.problemArea': doc.funcs[0].problemArea,
      'kitter.user': doc.funcs[1].user ? doc.funcs[1].user.label : '',
      'kitter.status': doc.funcs[1].status,
      'kitter.picklist': doc.funcs[1].picklist,
      'kitter.pickup': doc.funcs[1].pickup,
      'kitter.startedAt': doc.funcs[1].startedAt,
      'kitter.finishedAt': doc.funcs[1].finishedAt,
      'kitter.duration': doc.funcs[1].finishedAt
        ? (Math.round((doc.funcs[1].finishedAt - doc.funcs[1].startedAt) / 3600000 * 100) / 100) : 0,
      'kitter.carts': doc.funcs[1].carts.join(','),
      'kitter.problemArea': doc.funcs[1].problemArea,
      'platformer.user': doc.funcs[2].user ? doc.funcs[2].user.label : '',
      'platformer.status': doc.funcs[2].status,
      'platformer.picklist': doc.funcs[2].picklist,
      'platformer.pickup': doc.funcs[2].pickup,
      'platformer.startedAt': doc.funcs[2].startedAt,
      'platformer.finishedAt': doc.funcs[2].finishedAt,
      'platformer.duration': doc.funcs[2].finishedAt
        ? (Math.round((doc.funcs[2].finishedAt - doc.funcs[2].startedAt) / 3600000 * 100) / 100) : 0,
      'platformer.carts': doc.funcs[2].carts.join(','),
      'platformer.problemArea': doc.funcs[2].problemArea,
      'packer.user': doc.funcs[3].user ? doc.funcs[3].user.label : '',
      'packer.status': doc.funcs[3].status,
      'packer.picklist': doc.funcs[3].picklist,
      'packer.pickup': doc.funcs[3].pickup,
      'packer.startedAt': doc.funcs[3].startedAt,
      'packer.finishedAt': doc.funcs[3].finishedAt,
      'packer.duration': doc.funcs[3].finishedAt
        ? (Math.round((doc.funcs[3].finishedAt - doc.funcs[3].startedAt) / 3600000 * 100) / 100) : 0,
      'packer.carts': doc.funcs[3].carts.join(','),
      'packer.problemArea': doc.funcs[3].problemArea,
      'painter.user': doc.funcs[4].user ? doc.funcs[4].user.label : '',
      'painter.status': doc.funcs[4].status,
      'painter.picklist': doc.funcs[4].picklist,
      'painter.pickup': doc.funcs[4].pickup,
      'painter.startedAt': doc.funcs[4].startedAt,
      'painter.finishedAt': doc.funcs[4].finishedAt,
      'painter.duration': doc.funcs[4].finishedAt
        ? (Math.round((doc.funcs[4].finishedAt - doc.funcs[4].startedAt) / 3600000 * 100) / 100) : 0,
      'painter.carts': doc.funcs[4].carts.join(','),
      'painter.problemArea': doc.funcs[4].problemArea
    };
  }
};
