// wh/routes/exportSetCarts.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module, req, res, next) =>
{
  const {
    express,
    OldWhSetCart: WhSetCart
  } = module;

  express.crud.exportRoute(app, {
    filename: 'WMES_WH_CARTS',
    freezeRows: 1,
    freezeColumns: 2,
    columns: {
      kind: 10,
      cartNo: 5,
      date: 'date+utc',
      set: 'integer',
      line: 10,
      lines: 15,
      status: 10,
      startTime: 'datetime+utc',
      orders: 20,
      completedAt: 'datetime',
      completedBy: 40,
      deliveringAt: 'datetime',
      deliveringBy: 20,
      deliveredAt: 'datetime',
      deliveredBy: 20,
      duration: 'decimal'
    },
    model: WhSetCart,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    return {
      kind: doc.kind,
      cartNo: doc.cart,
      date: doc.date,
      set: doc.set,
      line: doc.line,
      lines: doc.lines.join(','),
      status: doc.status,
      startTime: doc.startTime,
      orders: doc.orders.map(o => o.sapOrder).join(','),
      completedAt: doc.completedAt,
      completedBy: doc.completedBy.map(u => u.label).join(','),
      deliveringAt: doc.deliveringAt,
      deliveringBy: doc.deliveringBy ? doc.deliveringBy.label : '',
      deliveredAt: doc.deliveredAt,
      deliveredBy: doc.deliveredBy ? doc.deliveredBy.label : '',
      duration: doc.deliveredAt
        ? (Math.round((doc.deliveredAt - doc.deliveringAt) / 60000 * 100) / 100)
        : 0
    };
  }
};
