// wh/routes/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');
const step = require('h5.step');
const shifts = require('util/shifts');
const checkMrpLockRoute = require('./checkMrpLock');
const printPsPicklistRoute = require('./printPsPicklist');
const exportEventsRoute = require('./exportEvents');
const exportSnapshotsRoute = require('./exportSnapshots');

module.exports = (app, module) =>
{
  const {
    updater,
    settings,
    express,
    user,
    Plan,
    OldWhEvent: WhEvent,
    OldWhOrder: WhOrder,
    OldWhUser: WhUser,
    OldWhDowntime: WhDowntime,
    OldWhSetCart: WhSetCart,
    OldWhLine: WhLine,
    OldWhLineSnapshot: WhLineSnapshot,
    OldWhDeliveredOrder: WhDeliveredOrder,
    OldWhRedirReason: WhRedirReason
  } = module;

  const canView = user.auth('LOCAL', 'WH:VIEW', 'PLANNING:VIEW', 'PAINT_SHOP:VIEW');
  const canUpdate = user.auth('LOCAL', 'WH:VIEW', 'PLANNING:MANAGE', 'PLANNING:PLANNER');
  const canManage = user.auth('WH:MANAGE', 'WH:MANAGE:USERS', 'PLANNING:MANAGE', 'PLANNING:PLANNER');

  // Apps
  [
    'wh-pickup',
    'wh-problems',
    'wh-delivery-components',
    'wh-delivery-packaging',
    'wh-delivery-ps'
  ].forEach(appId =>
  {
    express.get(`/${appId}`, canView, (req, res) =>
    {
      res.format({
        'text/html': function()
        {
          res.render('index', updater.getAppTemplateData(appId, req));
        }
      });
    });
  });

  // Settings
  express.get('/old/wh/settings', canView, settings.browseRoute('wh'));
  express.put('/old/wh/settings/:id', canManage, settings.updateRoute);

  // Actions
  express.post('/old/wh;generate', canManage, generateRoute);
  express.post('/old/wh;act', canUpdate, actRoute);
  express.post('/old/wh;check-mrp-lock', canView, checkMrpLockRoute.bind(null, app, module));
  express.get('/old/wh;print-ps-picklist', canManage, printPsPicklistRoute.bind(null, app, module));

  // Events
  express.get('/old/wh/events', canView, prepareEventFilter, express.crud.browseRoute.bind(null, app, WhEvent));
  express.get('/old/wh/events;export.:format?', canView, prepareEventFilter, exportEventsRoute.bind(null, app, module));

  // Orders
  express.get(
    '/old/wh/orders',
    canView,
    prepareCurrentDate,
    express.crud.browseRoute.bind(null, app, WhOrder)
  );
  express.post('/old/wh/orders/:id;reload', canManage, reloadRoute);
  express.get('/old/wh/orders/:id', canView, express.crud.readRoute.bind(null, app, WhOrder));

  // Users
  express.get('/old/wh/users', canView, express.crud.browseRoute.bind(null, app, WhUser));
  express.post('/old/wh/users', canManage, express.crud.addRoute.bind(null, app, WhUser));
  express.get('/old/wh/users/:id', canView, express.crud.readRoute.bind(null, app, WhUser));
  express.put('/old/wh/users/:id', canManage, express.crud.editRoute.bind(null, app, WhUser));
  express.delete('/old/wh/users/:id', canManage, express.crud.deleteRoute.bind(null, app, WhUser));

  // Downtimes
  express.get('/old/wh/downtimes', canView, express.crud.browseRoute.bind(null, app, WhDowntime));
  express.get('/old/wh/downtimes/:id', canView, express.crud.readRoute.bind(null, app, WhDowntime));

  // Set carts
  express.get('/old/wh/setCarts', canView, express.crud.browseRoute.bind(null, app, WhSetCart));
  express.get('/old/wh/setCarts/:id', canView, express.crud.readRoute.bind(null, app, WhSetCart));

  // Lines
  express.get('/old/wh/lines', canView, express.crud.browseRoute.bind(null, app, {
    model: WhLine,
    prepareResult: prepareLines
  }));
  express.get('/old/wh/lines/:id', canView, express.crud.readRoute.bind(null, app, {
    model: WhLine,
    prepareResult: prepareLine
  }));

  // Line snapshots
  express.get('/old/wh/lineSnapshots', canView, express.crud.browseRoute.bind(null, app, {
    model: WhLineSnapshot
  }));
  express.get('/old/wh/lineSnapshots;export.:format?', canView, exportSnapshotsRoute.bind(null, app, module));

  // Delivered orders
  express.get('/old/wh/deliveredOrders', canView, express.crud.browseRoute.bind(null, app, WhDeliveredOrder));
  express.post('/old/wh/deliveredOrders;redir', canManage, redirDeliveredOrderRoute);
  express.put('/old/wh/deliveredOrders/:id', canManage, editDeliveredOrderRoute);

  // Pickup status
  express.get('/old/wh/pickupStatus', canView, (req, res) => res.json(module.state.pickupStatus.get()));

  // Redir reasons
  express.get('/old/wh/redirReasons', canView, express.crud.browseRoute.bind(null, app, WhRedirReason));
  express.post('/old/wh/redirReasons', canManage, express.crud.addRoute.bind(null, app, WhRedirReason));
  express.get('/old/wh/redirReasons/:id', canView, express.crud.readRoute.bind(null, app, WhRedirReason));
  express.put('/old/wh/redirReasons/:id', canManage, express.crud.editRoute.bind(null, app, WhRedirReason));
  express.delete('/old/wh/redirReasons/:id', canManage, express.crud.deleteRoute.bind(null, app, WhRedirReason));

  function prepareCurrentDate(req, res, next)
  {
    req.rql.selector.args.forEach(function(term)
    {
      if (term.name !== 'eq' || term.args[0] !== 'date')
      {
        return;
      }

      const date = term.args[1];

      if (date === 'current')
      {
        const m = moment();

        if (m.hours() < 17)
        {
          m.startOf('day').subtract(1, 'days');
        }
        else
        {
          m.startOf('day').add(1, 'days');
        }

        term.args[1] = moment.utc(m.format('YYYY-MM-DD'), 'YYYY-MM-DD').toDate();
      }
      else if (/^[0-9]+-[0-9]+-[0-9]+$/.test(date))
      {
        term.args[1] = moment.utc(date, 'YYYY-MM-DD').toDate();
      }
    });

    next();
  }

  function prepareEventFilter(req, res, next)
  {
    let objects = req.rql.selector.args.find(term => term.name === 'in' && term.args[0] === 'objects');

    if (!objects)
    {
      objects = {
        name: 'in',
        args: ['objects', []]
      };

      req.rql.selector.args.push(objects);
    }

    req.rql.selector.args = req.rql.selector.args.filter(term =>
    {
      const prop = term.args[0];

      if (term.name === 'eq' && (prop === 'order' || prop === 'line' || prop === 'user' || prop === 'set'))
      {
        objects.name = 'all';
        objects.args[1].push(term.args[1]);

        return false;
      }

      return true;
    });

    next();
  }

  async function prepareLines(totalCount, collection, formatResult)
  {
    try
    {
      formatResult(null, {
        totalCount,
        collection: await WhLine.prepareLines(collection)
      });
    }
    catch (err)
    {
      formatResult(err);
    }
  }

  function prepareLine(line, formatResult)
  {
    step(
      function()
      {
        const shiftInfo = shifts.getCurrentShiftInfo(true);
        const planDate = shifts.getPlanDate(shiftInfo.time, true).toDate();
        const pipeline = [
          {$match: {_id: planDate}},
          {$unwind: '$lines'},
          {$match: {
            'lines._id': line._id,
            [`lines.shiftData.${shiftInfo.no - 1}.orderCount`]: {$gt: 0}
          }},
          {$project: {_id: '$lines._id'}}
        ];

        Plan.aggregate(pipeline, this.next());
      },
      function(err, plannedLineList)
      {
        if (err)
        {
          return formatResult(err);
        }

        line.planned = plannedLineList.length === 1;
        line.working = !!module.state.getLineWorkStatus(line._id);

        formatResult(null, line);
      }
    );
  }

  function generateRoute(req, res)
  {
    app.broker.publish('old.wh.generator.requested', {
      date: req.query.date || req.body.date
    });

    res.sendStatus(204);
  }

  function actRoute(req, res, next)
  {
    if (!req.body.data)
    {
      req.body.data = {};
    }

    req.body.user = user.createUserInfo(req.session.user, req);

    if (req.body.date)
    {
      const dateMoment = moment.utc(req.params.date, 'YYYY-MM-DD');

      if (!dateMoment.isValid())
      {
        return next(app.createError('Invalid date.', 'INPUT', 400));
      }

      req.body.data.date = dateMoment.toDate();
    }

    module.state.act(req.body, (err, result) =>
    {
      if (err)
      {
        return next(err);
      }

      if (result)
      {
        res.json(result);
      }
      else
      {
        res.sendStatus(204);
      }
    });
  }

  function reloadRoute(req, res, next)
  {
    WhOrder.findById(req.params.id).lean().exec((err, whOrder) =>
    {
      if (err)
      {
        return next(err);
      }

      if (!whOrder)
      {
        return next(app.createError('Order not found.', 'NOT_FOUND', 404));
      }

      app.broker.publish(`${WhOrder.TOPIC_PREFIX}.updated`, {updated: [whOrder]});

      res.json(whOrder);
    });
  }

  function redirDeliveredOrderRoute(req, res, next)
  {
    const input = {
      action: 'redirDeliveredOrder',
      user: user.createUserInfo(req.session.user, req),
      data: req.body
    };

    module.state.act(input, (err, result) =>
    {
      if (err)
      {
        return next(err);
      }

      res.json(result);
    });
  }

  function editDeliveredOrderRoute(req, res, next)
  {
    const input = {
      action: 'editDeliveredOrder',
      user: user.createUserInfo(req.session.user, req),
      data: req.body
    };

    if (typeof input.data.qtyDone !== 'number' || typeof input.data.qtyTodo !== 'number')
    {
      return next(app.createError('Invalid quantities.', 'INPUT', 400));
    }

    module.state.act(input, (err, result) =>
    {
      if (err)
      {
        return next(err);
      }

      res.json(result);
    });
  }
};
