// wh/routes/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');
const checkMrpLockRoute = require('./checkMrpLock');

module.exports = (app, module) =>
{
  const {
    updater,
    settings,
    express,
    user,
    OldWhEvent: WhEvent,
    OldWhOrder: WhOrder,
    OldWhUser: WhUser,
    OldWhDowntime: WhDowntime,
    OldWhSetCart: WhSetCart,
    OldWhLine: WhLine,
    OldWhDeliveredOrder: WhDeliveredOrder,
    OldWhPendingComponents: WhPendingComponents,
    OldWhPendingPackaging: WhPendingPackaging
  } = module;

  const canView = user.auth('LOCAL', 'WH:VIEW', 'PLANNING:VIEW', 'PAINT_SHOP:VIEW');
  const canUpdate = user.auth('LOCAL', 'WH:VIEW', 'PLANNING:MANAGE', 'PLANNING:PLANNER');
  const canManage = user.auth('WH:MANAGE', 'WH:MANAGE:USERS', 'PLANNING:MANAGE', 'PLANNING:PLANNER');

  // Apps
  ['wh-pickup', 'wh-problems', 'wh-delivery-components', 'wh-delivery-packaging'].forEach(appId =>
  {
    express.get(`/${appId}`, canView, (req, res) =>
    {
      res.format({
        'text/html': function()
        {
          res.render('index', updater.getAppTemplateData(appId, req));
        }
      });
    });
  });

  // Settings
  express.get('/old/wh/settings', canView, settings.browseRoute('wh'));
  express.put('/old/wh/settings/:id', canManage, settings.updateRoute);

  // Actions
  express.post('/old/wh;generate', canManage, generateRoute);
  express.post('/old/wh;act', canUpdate, actRoute);
  express.post('/old/wh;check-mrp-lock', canView, checkMrpLockRoute.bind(null, app, module));

  // Events
  express.get(
    '/old/wh/events',
    canView,
    express.crud.browseRoute.bind(null, app, WhEvent)
  );

  // Orders
  express.get(
    '/old/wh/orders',
    canView,
    prepareCurrentDate,
    express.crud.browseRoute.bind(null, app, WhOrder)
  );
  express.post('/old/wh/orders/:id;reload', canManage, reloadRoute);
  express.get('/old/wh/orders/:id', canView, express.crud.readRoute.bind(null, app, WhOrder));

  // Users
  express.get('/old/wh/users', canView, express.crud.browseRoute.bind(null, app, WhUser));
  express.post('/old/wh/users', canManage, express.crud.addRoute.bind(null, app, WhUser));
  express.get('/old/wh/users/:id', canView, express.crud.readRoute.bind(null, app, WhUser));
  express.put('/old/wh/users/:id', canManage, express.crud.editRoute.bind(null, app, WhUser));
  express.delete('/old/wh/users/:id', canManage, express.crud.deleteRoute.bind(null, app, WhUser));

  // Downtimes
  express.get('/old/wh/downtimes', canView, express.crud.browseRoute.bind(null, app, WhDowntime));
  express.get('/old/wh/downtimes/:id', canView, express.crud.readRoute.bind(null, app, WhDowntime));

  // Downtimes
  express.get('/old/wh/setCarts', canView, express.crud.browseRoute.bind(null, app, WhSetCart));
  express.get('/old/wh/setCarts/:id', canView, express.crud.readRoute.bind(null, app, WhSetCart));

  // Lines
  express.get('/old/wh/lines', canView, express.crud.browseRoute.bind(null, app, WhLine));
  express.get('/old/wh/lines/:id', canView, express.crud.readRoute.bind(null, app, WhLine));

  // Delivered orders
  express.get('/old/wh/deliveredOrders', canView, express.crud.browseRoute.bind(null, app, WhDeliveredOrder));
  express.put('/old/wh/deliveredOrders/:id', canManage, editDeliveredOrderRoute);

  // Pending deliveries
  express.get('/old/wh/pending/components', canView, express.crud.browseRoute.bind(null, app, WhPendingComponents));
  express.get('/old/wh/pending/packaging', canView, express.crud.browseRoute.bind(null, app, WhPendingPackaging));

  function prepareCurrentDate(req, res, next)
  {
    req.rql.selector.args.forEach(function(term)
    {
      if (term.name !== 'eq' || term.args[0] !== 'date')
      {
        return;
      }

      const date = term.args[1];

      if (date === 'current')
      {
        const m = moment();

        if (m.hours() < 17)
        {
          m.startOf('day').subtract(1, 'days');
        }
        else
        {
          m.startOf('day').add(1, 'days');
        }

        term.args[1] = moment.utc(m.format('YYYY-MM-DD'), 'YYYY-MM-DD').toDate();
      }
      else if (/^[0-9]+-[0-9]+-[0-9]+$/.test(date))
      {
        term.args[1] = moment.utc(date, 'YYYY-MM-DD').toDate();
      }
    });

    next();
  }

  function generateRoute(req, res)
  {
    app.broker.publish('old.wh.generator.requested', {
      date: req.query.date || req.body.date
    });

    res.sendStatus(204);
  }

  function actRoute(req, res, next)
  {
    if (!req.body.data)
    {
      req.body.data = {};
    }

    req.body.user = user.createUserInfo(req.session.user, req);

    if (req.body.date)
    {
      const dateMoment = moment.utc(req.params.date, 'YYYY-MM-DD');

      if (!dateMoment.isValid())
      {
        return next(app.createError('Invalid date.', 'INPUT', 400));
      }

      req.body.data.date = dateMoment.toDate();
    }

    module.state.act(req.body, (err, result) =>
    {
      if (err)
      {
        return next(err);
      }

      if (result)
      {
        res.json(result);
      }
      else
      {
        res.sendStatus(204);
      }
    });
  }

  function reloadRoute(req, res, next)
  {
    WhOrder.findById(req.params.id).lean().exec((err, whOrder) =>
    {
      if (err)
      {
        return next(err);
      }

      if (!whOrder)
      {
        return next(app.createError('Order not found.', 'NOT_FOUND', 404));
      }

      app.broker.publish(`${WhOrder.TOPIC_PREFIX}.updated`, {updated: [whOrder]});

      res.json(whOrder);
    });
  }

  function editDeliveredOrderRoute(req, res, next)
  {
    const input = {
      action: 'editDeliveredOrder',
      user: user.createUserInfo(req.session.user, req),
      data: req.body
    };

    if (typeof input.data.qtyDone !== 'number' || typeof input.data.qtyTodo !== 'number')
    {
      return next(app.createError('Invalid quantities.', 'INPUT', 400));
    }

    module.state.act(input, (err, result) =>
    {
      if (err)
      {
        return next(err);
      }

      res.json(result);
    });
  }
};
