// wmes-osh/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  uploadsPath: './data/uploads/osh',
  anonymousUserInfo: {
    id: '5f7a32c969226110e835d75b',
    label: 'Anonim'
  }
};

exports.models = [
  require('./models/oshWorkplace'),
  require('./models/oshDivision'),
  require('./models/oshBuilding'),
  require('./models/oshLocation'),
  require('./models/oshStation'),
  require('./models/oshKind'),
  require('./models/oshEventCategory'),
  require('./models/oshReasonCategory'),
  require('./models/oshActivityKind'),
  require('./models/oshNearMiss'),
  require('./models/oshKaizen')
];

exports.modelPlugins = {
  User: require('./plugins/user')
};

exports.republishTopics = [
  'osh.workplaces.added', 'osh.workplaces.edited', 'osh.workplaces.deleted',
  'osh.divisions.added', 'osh.divisions.edited', 'osh.divisions.deleted',
  'osh.buildings.added', 'osh.buildings.edited', 'osh.buildings.deleted',
  'osh.locations.added', 'osh.locations.edited', 'osh.locations.deleted',
  'osh.stations.added', 'osh.stations.edited', 'osh.stations.deleted',
  'osh.kinds.added', 'osh.kinds.edited', 'osh.kinds.deleted',
  'osh.eventCategories.added', 'osh.eventCategories.edited', 'osh.eventCategories.deleted',
  'osh.reasonCategories.added', 'osh.reasonCategories.edited', 'osh.reasonCategories.deleted',
  'osh.activityKinds.added', 'osh.activityKinds.edited', 'osh.activityKinds.deleted',
  'osh.nearMisses.added', 'osh.nearMisses.edited', 'osh.nearMisses.deleted',
  'osh.nearMisses.updated.*', 'osh.nearMisses.seen.*',
  'osh.kaizens.added', 'osh.kaizens.edited', 'osh.kaizens.deleted',
  'osh.kaizens.updated.*', 'osh.kaizens.seen.*'
];

exports.recordTopics = {
  debug: [
    'osh.workplaces.added', 'osh.workplaces.edited',
    'osh.divisions.added', 'osh.divisions.edited',
    'osh.buildings.added', 'osh.buildings.edited',
    'osh.locations.added', 'osh.locations.edited',
    'osh.stations.added', 'osh.stations.edited',
    'osh.kinds.added', 'osh.kinds.edited',
    'osh.eventCategories.added', 'osh.eventCategories.edited',
    'osh.reasonCategories.added', 'osh.reasonCategories.edited',
    'osh.activityKinds.added', 'osh.activityKinds.edited'
  ],
  warning: [
    'osh.workplaces.deleted',
    'osh.divisions.deleted',
    'osh.buildings.deleted',
    'osh.locations.deleted',
    'osh.stations.deleted',
    'osh.kinds.deleted',
    'osh.eventCategories.deleted',
    'osh.reasonCategories.deleted',
    'osh.activityKinds.deleted',
    'osh.nearMisses.deleted',
    'osh.kaizens.deleted'
  ]
};

exports.userPrivileges = [
  'OSH:DICTIONARIES:VIEW', 'OSH:DICTIONARIES:MANAGE',
  'OSH:NEAR_MISSES:MANAGE',
  'OSH:KAIZENS:MANAGE'
];

exports.optionalModules = {
  'express': require('./attachments'),
  'mongoose': require('./relations'),
  'mongoose user express settings': require('./routes')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    workplaces: {model: module.OshWorkplace},
    divisions: {model: module.OshDivision},
    buildings: {model: module.OshBuilding},
    locations: {model: module.OshLocation},
    stations: {model: module.OshStation},
    kinds: {model: module.OshKind},
    eventCategories: {model: module.OshEventCategory},
    reasonCategories: {model: module.OshReasonCategory},
    activityKinds: {model: module.OshActivityKind}
  };

  module.TYPE_TO_MODEL = {
    nearMiss: module.OshNearMiss,
    kaizen: module.OshKaizen
  };
};
