// wmes-osh/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  uploadsPath: './data/uploads/osh',
  anonymousUserInfo: {
    id: '5f7a32c969226110e835d75b',
    label: 'Anonim'
  }
};

exports.models = [
  require('./models/oshCompany'),
  require('./models/oshWorkplace'),
  require('./models/oshDivision'),
  require('./models/oshBuilding'),
  require('./models/oshLocation'),
  require('./models/oshStation'),
  require('./models/oshKind'),
  require('./models/oshActivityKind'),
  require('./models/oshObservationKind'),
  require('./models/oshObservationCategory'),
  require('./models/oshEventCategory'),
  require('./models/oshReasonCategory'),
  require('./models/oshRootCauseCategory'),
  require('./models/oshNearMiss'),
  require('./models/oshKaizen'),
  require('./models/oshAction'),
  require('./models/oshObservation')
];

exports.modelPlugins = {
  User: require('./plugins/user')
};

exports.republishTopics = [
  'osh.companies.added', 'osh.companies.edited', 'osh.companies.deleted',
  'osh.workplaces.added', 'osh.workplaces.edited', 'osh.workplaces.deleted',
  'osh.divisions.added', 'osh.divisions.edited', 'osh.divisions.deleted',
  'osh.buildings.added', 'osh.buildings.edited', 'osh.buildings.deleted',
  'osh.locations.added', 'osh.locations.edited', 'osh.locations.deleted',
  'osh.stations.added', 'osh.stations.edited', 'osh.stations.deleted',
  'osh.kinds.added', 'osh.kinds.edited', 'osh.kinds.deleted',
  'osh.activityKinds.added', 'osh.activityKinds.edited', 'osh.activityKinds.deleted',
  'osh.observationKinds.added', 'osh.observationKinds.edited', 'osh.observationKinds.deleted',
  'osh.observationCategories.added', 'osh.observationCategories.edited', 'osh.observationCategories.deleted',
  'osh.eventCategories.added', 'osh.eventCategories.edited', 'osh.eventCategories.deleted',
  'osh.reasonCategories.added', 'osh.reasonCategories.edited', 'osh.reasonCategories.deleted',
  'osh.rootCauseCategories.added', 'osh.rootCauseCategories.edited', 'osh.rootCauseCategories.deleted',
  'osh.nearMisses.added', 'osh.nearMisses.edited', 'osh.nearMisses.deleted',
  'osh.nearMisses.updated.*', 'osh.nearMisses.seen.*',
  'osh.kaizens.added', 'osh.kaizens.edited', 'osh.kaizens.deleted',
  'osh.kaizens.updated.*', 'osh.kaizens.seen.*',
  'osh.actions.added', 'osh.actions.edited', 'osh.actions.deleted',
  'osh.actions.updated.*', 'osh.actions.seen.*',
  'osh.observations.added', 'osh.observations.edited', 'osh.observations.deleted',
  'osh.observations.updated.*', 'osh.observations.seen.*'
];

exports.recordTopics = {
  debug: [
    'osh.companies.added', 'osh.companies.edited',
    'osh.workplaces.added', 'osh.workplaces.edited',
    'osh.divisions.added', 'osh.divisions.edited',
    'osh.buildings.added', 'osh.buildings.edited',
    'osh.locations.added', 'osh.locations.edited',
    'osh.stations.added', 'osh.stations.edited',
    'osh.kinds.added', 'osh.kinds.edited',
    'osh.activityKinds.added', 'osh.activityKinds.edited',
    'osh.observationKinds.added', 'osh.observationKinds.edited',
    'osh.observationCategories.added', 'osh.observationCategories.edited',
    'osh.eventCategories.added', 'osh.eventCategories.edited',
    'osh.reasonCategories.added', 'osh.reasonCategories.edited',
    'osh.rootCauseCategories.added', 'osh.rootCauseCategories.edited'
  ],
  warning: [
    'osh.companies.deleted',
    'osh.workplaces.deleted',
    'osh.divisions.deleted',
    'osh.buildings.deleted',
    'osh.locations.deleted',
    'osh.stations.deleted',
    'osh.kinds.deleted',
    'osh.activityKinds.deleted',
    'osh.observationKinds.deleted',
    'osh.observationCategories.deleted',
    'osh.eventCategories.deleted',
    'osh.reasonCategories.deleted',
    'osh.rootCauseCategories.deleted',
    'osh.nearMisses.deleted',
    'osh.kaizens.deleted',
    'osh.actions.deleted'
  ]
};

exports.userPrivileges = [
  'OSH:DICTIONARIES:VIEW', 'OSH:DICTIONARIES:MANAGE',
  'OSH:NEAR_MISSES:MANAGE',
  'OSH:KAIZENS:MANAGE',
  'OSH:ACTIONS:MANAGE',
  'OSH:OBSERVATIONS:MANAGE'
];

exports.optionalModules = {
  'express': require('./attachments'),
  'mongoose': require('./relations'),
  'mongoose user express settings': require('./routes')
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    companies: {model: module.OshCompany},
    workplaces: {model: module.OshWorkplace},
    divisions: {model: module.OshDivision},
    buildings: {model: module.OshBuilding},
    locations: {model: module.OshLocation},
    stations: {model: module.OshStation},
    kinds: {model: module.OshKind},
    activityKinds: {model: module.OshActivityKind},
    observationKinds: {model: module.OshObservationKind},
    observationCategories: {model: module.OshObservationCategory},
    eventCategories: {model: module.OshEventCategory},
    reasonCategories: {model: module.OshReasonCategory},
    rootCauseCategories: {model: module.OshRootCauseCategory}
  };

  module.TYPE_TO_MODEL = {
    [module.OshNearMiss.RELATION_TYPE]: module.OshNearMiss,
    [module.OshKaizen.RELATION_TYPE]: module.OshKaizen,
    [module.OshAction.RELATION_TYPE]: module.OshAction,
    [module.OshObservation.RELATION_TYPE]: module.OshObservation
  };
};
