// orders/models/productTag.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'ProductTag';

exports.setUp = (app, mongoose) =>
{
  const productTagSchema = new mongoose.Schema({
    _id: String,
    label: String,
    conditions: [{
      mrp: [String],
      nc12: [String],
      bom: [String]
    }]
  }, {
    id: false
  });

  productTagSchema.statics.TOPIC_PREFIX = 'orders.productTags';

  productTagSchema.statics.createMatcher = function(conditions)
  {
    if (!conditions || !conditions.length)
    {
      return () => false;
    }

    const matchers = conditions.map(condition =>
    {
      return (order, bom) =>
      {
        if (condition.mrp.length && !condition.mrp.includes(order.mrp))
        {
          return false;
        }

        if (condition.nc12.length && !condition.nc12.includes(order.nc12))
        {
          return false;
        }

        if (condition.bom.length && !condition.bom.some(code => bom.has(code)))
        {
          return false;
        }

        return true;
      };
    });

    return (order, bom) => matchers.some(match => match(order, bom));
  };

  mongoose.model(exports.name, productTagSchema);
};
