// users/routes/loginIn.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const request = require('request');
const uuid = require('uuid');

module.exports = (app, module, req, res, next) =>
{
  const loginIn = module.config.loginIn[req.params.app];

  if (!loginIn)
  {
    return next(app.createError('Unknown app.', 'INVALID_APP', 400));
  }

  const email = req.session.user.email || '';

  if (!email.includes('@'))
  {
    return res.redirect(loginIn.url);
  }

  const loginAsUrl = Object.assign(new URL(loginIn.url), {pathname: '/loginAs'});
  const loginSecretKey = uuid.v4();

  step(
    function()
    {
      const req = {
        method: 'POST',
        url: loginAsUrl.toString(),
        json: true,
        body: {
          app: loginIn.app,
          appSecretKey: loginIn.secretKey,
          loginSecretKey,
          email
        }
      };

      request(req, this.next());
    },
    function(err, res)
    {
      if (err)
      {
        return this.skip(err);
      }

      if (res.statusCode !== 204)
      {
        return this.skip(app.createError('Unexpected response code.', 'LOGIN_AS_FAILURE', 500));
      }
    },
    function(err)
    {
      if (err)
      {
        module.error(err, `Failed to login to app.`, {
          app: req.params.app,
          email
        });

        return res.redirect(loginIn.url);
      }

      loginAsUrl.searchParams.set('secretKey', loginSecretKey);

      res.redirect(loginAsUrl.toString());
    }
  );
};
