// production/routes/getActiveLines.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module, req, res) =>
{
  const {
    orgUnits
  } = module;

  const includeDeactivated = req.query.active === '0';
  const subdivisionType = req.query.subdivisionType || null;

  const activeProdLines = [];

  orgUnits.getAllByType('prodLine').forEach(prodLine =>
  {
    const deactivated = !!prodLine.deactivatedAt;

    if (deactivated && !includeDeactivated)
    {
      return;
    }

    const lineOrgUnits = orgUnits.getAllForProdLine(prodLine);

    if (!lineOrgUnits.division)
    {
      return;
    }

    const subdivision = orgUnits.getByTypeAndId('subdivision', lineOrgUnits.subdivision);

    if (subdivisionType && subdivisionType !== subdivision.type)
    {
      return;
    }

    activeProdLines.push({
      _id: prodLine._id,
      description: prodLine.description,
      deactivated,
      orgUnits: lineOrgUnits,
      subdivisionType: subdivision.type
    });
  });

  activeProdLines.sort((a, b) =>
  {
    return a._id.localeCompare(b._id, undefined, {
      numeric: true,
      ignorePunctuation: true
    });
  });

  res.json({
    totalCount: activeProdLines.length,
    collection: activeProdLines
  });
};
