// production/routes/getRecentPersonnelRoute.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');

const PERSONNEL_TYPES = {
  master: 1,
  leader: 1,
  operator: 1,
  operators: 1
};

module.exports = function getRecentPersonnelRoute(app, module, req, res, next)
{
  const {type, prodLine, shift} = req.query;

  if (!PERSONNEL_TYPES[type])
  {
    return next(app.createError('UNKNOWN_TYPE'));
  }

  const {orgUnits, ProdShift} = module;

  const pipeline = [
    {$match: {
      prodLine: orgUnits.fix.prodLine(prodLine),
      date: {$gte: moment().subtract(14, 'days').toDate()},
      shift: parseInt(shift, 10)
    }}
  ];

  if (type === 'operators')
  {
    pipeline.push({
      $unwind: '$operators'
    });
  }

  pipeline.push(
    {$group: {_id: `$${type}.id`, label: {$max: `$${type}.label`}, count: {$sum: 1}}},
    {$match: {_id: {$ne: null}}},
    {$sort: {count: -1}},
    {$limit: 7}
  );

  ProdShift.aggregate(pipeline, function(err, results)
  {
    if (err)
    {
      return next(err);
    }

    res.json(results);
  });
};
