// wmes-osh/routes/kaizens/add.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

module.exports = async (app, module, req, res, next) =>
{
  const {
    OshKaizen
  } = module;

  const input = _.pick(req.body, [
    'subject',
    'kind',
    'kaizenCategory',
    'division',
    'workplace',
    'department',
    'building',
    'location',
    'station',
    'problem',
    'reason',
    'suggestion',
    'solution',
    'implementers',
    'plannedAt',
    'attachments'
  ]);

  if (input.attachments && Array.isArray(input.attachments.added))
  {
    input.attachments = module.attachments.claimPendingUploads(input.attachments.added);
  }
  else
  {
    input.attachments = [];
  }

  try
  {
    const creator = module.createUserInfo(req.session.user, req);
    const createdAt = new Date();

    const kaizen = new OshKaizen({
      status: 'new',
      ...input,
      createdAt,
      updatedAt: createdAt,
      startedAt: null,
      implementedAt: null,
      finishedAt: null,
      duration: 0,
      creator,
      updater: creator,
      acceptor: null,
      solver: null,
      finisher: null,
      kom: 0,
      reward: 0
    });

    if (kaizen.status === 'inProgress')
    {
      kaizen.startedAt = kaizen.createdAt;
      kaizen.acceptor = creator;
    }
    else if (kaizen.status === 'verification')
    {
      kaizen.startedAt = kaizen.createdAt;
      kaizen.implementedAt = kaizen.createdAt;
      kaizen.acceptor = creator;
      kaizen.solver = creator;
    }

    const [coordinators] = await Promise.all([
      OshKaizen.resolveCoordinators(kaizen, {exceptUsers: [kaizen.creator]}),
      OshKaizen.assignNextRid(kaizen)
    ]);

    kaizen.coordinators = coordinators;

    kaizen.updateUsers([], kaizen.creator);

    await kaizen.save();

    res.json(kaizen);

    app.broker.publish(`${OshKaizen.TOPIC_PREFIX}.added`, {
      user: kaizen.creator,
      model: kaizen
    });
  }
  catch (err)
  {
    next(err);
  }
};
