// wmes-osh/routes/kaizens/edit.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const locks = require('util/locks');

module.exports = async (app, module, req, res, next) =>
{
  const {
    user,
    OshKaizen,
    OshReward
  } = module;

  const releaseLocks = await Promise.all([
    locks.createAsync(`${OshKaizen.TOPIC_PREFIX}.${req.params.id}`),
    locks.createAsync(OshReward.TOPIC_PREFIX)
  ]);

  const sessionUser = req.session.user;
  const updater = module.createUserInfo(sessionUser, req);
  const updatedAt = new Date();
  let session = null;

  try
  {
    session = await OshKaizen.startSession();
    session.startTransaction();

    const kaizen = await OshKaizen
      .findById(req.params.id)
      .session(session)
      .exec();

    if (!kaizen)
    {
      throw app.createError('Entry not found.', 'NOT_FOUND', 404);
    }

    const canManage = user.isAllowedTo(sessionUser, [[`${OshKaizen.PRIVILEGE_PREFIX}:MANAGE`]]);
    const isCreator = kaizen.isCreator(updater.id);
    const isImplementer = kaizen.isImplementer(updater.id);
    const isCoordinator = kaizen.isCoordinator(updater.id);
    const allowedInput = ['comment'];

    if (canManage || isCoordinator)
    {
      allowedInput.push(
        'division',
        'workplace',
        'department',
        'building',
        'location',
        'station',
        'kind',
        'implementers',
        'kom',
        'reward'
      );
    }

    if (isCreator || isImplementer || canManage || isCoordinator)
    {
      allowedInput.push(
        'status',
        'subject',
        'kaizenCategory',
        'problem',
        'reason',
        'suggestion',
        'solution',
        'plannedAt',
        'attachments'
      );
    }

    const input = _.pick(req.body, allowedInput);

    module.attachments.prepareInput(input, kaizen);

    if (input.status && input.status !== kaizen.status)
    {
      if (input.status === 'inProgress' && !kaizen.startedAt)
      {
        input.startedAt = updatedAt;
      }
      else if (input.status === 'verification')
      {
        input.implementedAt = updatedAt;
      }
      else if (input.status === 'finished')
      {
        input.finishedAt = updatedAt;
      }

      input.statusComment = input.comment || '';
      input.statusUpdater = updater;

      if (input.status !== 'verification' && input.status !== 'finished')
      {
        input.kom = 0;
        input.reward = 0;
      }
    }

    await OshKaizen.prepareCoordinatorsInput(
      kaizen,
      input,
      ['kind', 'department', 'location'],
      {session, exceptUsers: [kaizen.creator].concat(input.implementers || [])}
    );

    const usersToNotify = kaizen.applyChanges(input, updater, updatedAt);

    if (!usersToNotify)
    {
      return res.sendStatus(204);
    }

    const change = _.last(kaizen.changes);
    const messages = [{
      topic: `${OshKaizen.TOPIC_PREFIX}.edited`,
      message: {
        user: updater,
        model: kaizen
      }
    }, {
      topic: `${OshKaizen.TOPIC_PREFIX}.updated.${kaizen._id}`,
      message: {
        _id: kaizen._id,
        change,
        notify: usersToNotify
      }
    }];

    if (kaizen.status === 'finished' && (change.data.implementers || change.data.reward))
    {
      await OshReward.updateEntry({
        entry: kaizen,
        createdAt: updatedAt,
        creator: updater,
        recipients: kaizen.implementers,
        totalAmount: kaizen.reward,
        session,
        messages
      });
    }

    await kaizen.save();

    await session.commitTransaction();

    res.json(kaizen);

    messages.forEach(({topic, message}) => app.broker.publish(topic, message));
  }
  catch (err)
  {
    next(err);

    if (session)
    {
      try
      {
        await session.abortTransaction();
      }
      catch (err)
      {
        if (err.message !== 'Cannot call abortTransaction twice')
        {
          module.error(err, `Failed to abort a transaction.`);
        }
      }
    }
  }
  finally
  {
    if (session)
    {
      session.endSession();
    }

    releaseLocks.forEach(release => release());
  }
};
