// wmes-osh/plugins/user.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const userInfoSchema = require('user/models/userInfoSchema');

module.exports = (app, module, userSchema) =>
{
  userSchema.path('oshWorkplace', {
    type: Number,
    ref: 'OshWorkplace',
    default: 0
  });

  userSchema.path('oshDepartment', {
    type: Number,
    ref: 'OshDepartment',
    default: 0
  });

  userSchema.index({oshWorkplace: 1});
  userSchema.index({oshDepartment: 1});

  userSchema.statics.dataAnonymizers.push(user =>
  {
    user.oshWorkplace = 0;
    user.oshDepartment = 0;
  });

  userSchema.statics.EDITABLE_PROPERTIES.oshWorkplace = true;
  userSchema.statics.EDITABLE_PROPERTIES.oshDepartment = true;

  userInfoSchema.path('oshWorkplace', {
    type: Number,
    ref: 'OshWorkplace'
  });

  userInfoSchema.path('oshDepartment', {
    type: Number,
    ref: 'OshDepartment'
  });

  app.broker.subscribe(`osh.workplaces.deleted`, onWorkplaceDeleted);
  app.broker.subscribe(`osh.departments.edited`, onDepartmentEdited);
  app.broker.subscribe(`osh.departments.deleted`, onDepartmentDeleted);

  async function onWorkplaceDeleted({model})
  {
    const {User} = module;

    try
    {
      const conditions = {
        oshWorkplace: model._id
      };

      const users = await User
        .find(conditions)
        .select({_id: 1})
        .lean()
        .exec();

      if (!users.length)
      {
        return;
      }

      await User.collection.updateMany(conditions, {$set: {oshWorkplace: 0, oshDepartment: 0}});

      users.forEach(user =>
      {
        app.broker.publish(`users.updated.${user._id}`, {
          _id: user._id,
          oshWorkplace: 0,
          oshDepartment: 0
        });
      });
    }
    catch (err)
    {
      module.error(err, `Failed to update users after workplace deletion.`);
    }
  }

  async function onDepartmentEdited({model})
  {
    if (!model.lastModifiedPaths().includes('workplace'))
    {
      return;
    }

    const {User} = module;

    try
    {
      const conditions = {
        oshDepartment: model._id,
        oshWorkplace: {$ne: model.workplace}
      };

      const users = await User
        .find(conditions)
        .select({_id: 1})
        .lean()
        .exec();

      if (!users.length)
      {
        return;
      }

      await User.collection.updateMany(conditions, {$set: {oshWorkplace: model.workplace}});

      users.forEach(user =>
      {
        app.broker.publish(`users.updated.${user._id}`, {
          _id: user._id,
          oshWorkplace: model.workplace
        });
      });
    }
    catch (err)
    {
      module.error(err, `Failed to update users after department edit.`);
    }
  }

  async function onDepartmentDeleted({model})
  {
    const {User} = module;

    try
    {
      const conditions = {
        oshDepartment: model._id
      };

      const users = await User
        .find(conditions)
        .select({_id: 1})
        .lean()
        .exec();

      if (!users.length)
      {
        return;
      }

      await User.collection.updateMany(conditions, {$set: {oshDepartment: 0}});

      users.forEach(user =>
      {
        app.broker.publish(`users.updated.${user._id}`, {
          _id: user._id,
          oshDepartment: 0
        });
      });
    }
    catch (err)
    {
      module.error(err, `Failed to update users after department deletion.`);
    }
  }
};
