// prodSerialNumbers/models/prodSerialNumber.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'ProdSerialNumber';

exports.setUp = (app, mongoose) =>
{
  const psnSchema = new mongoose.Schema({
    _id: String,
    scannedAt: {
      type: Date,
      required: true
    },
    prodLine: {
      type: String,
      required: true
    },
    mrp: String,
    station: {
      type: Number,
      default: 0
    },
    prodShiftOrder: {
      type: String,
      required: true
    },
    orderNo: {
      type: String,
      required: true,
      pattern: /^[0-9]+$/
    },
    serialNo: {
      type: Number,
      required: true,
      min: 0,
      max: 9999
    },
    taktTime: {
      type: Number,
      required: true
    },
    sapTaktTime: {
      type: Number,
      required: true
    },
    iptTaktTime: {
      type: Number,
      default: 0
    },
    iptAt: {
      type: Date,
      default: null
    },
    bom: [String]
  }, {
    id: false,
    versionKey: false,
    minimize: false
  });

  psnSchema.statics.TOPIC_PREFIX = 'prodSerialNumbers';

  psnSchema.index({mrp: 1, scannedAt: -1});
  psnSchema.index({orderNo: 1});
  psnSchema.index({prodShiftOrder: 1, scannedAt: -1});
  psnSchema.index({prodLine: 1, scannedAt: -1});
  psnSchema.index({scannedAt: -1});
  psnSchema.index({bom: 1}, {partialFilterExpression: {bom: {$type: 'array'}}});

  psnSchema.post('save', function(doc)
  {
    if (app.production.recreating)
    {
      return;
    }

    app.broker.publish(`prodSerialNumbers.created.${doc.prodLine}`, doc.toJSON());
  });

  mongoose.model(exports.name, psnSchema);
};
