// wmes-osh/routes/nearMisses/export.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');

module.exports = async (app, module, req, res, next) =>
{
  const {
    express,
    nameCache,
    OshNearMiss
  } = module;

  const names = await nameCache.load([
    'divisions',
    'workplaces',
    'departments',
    'buildings',
    'locations',
    'stations',
    'kinds',
    'eventCategories',
    'reasonCategories'
  ]);

  const columns = {
    rid: 13,
    status: 10,
    createdAt: 'datetime',
    startedAt: 'datetime',
    plannedAt: 'date+utc',
    finishedAt: 'datetime',
    duration: 'decimal',
    creator: 20,
    acceptor: 20,
    finisher: 20,
    implementer: 20,
    coordinators: 20,
    division: 20,
    workplace: 20,
    department: 20,
    building: 20,
    location: 20,
    station: 20,
    kind: 20,
    eventDate: 'datetime+utc',
    eventCategory: 20,
    reasonCategory: 20,
    priority: 'integer',
    materialLoss: 'boolean',
    subject: 40,
    problem: 40,
    reason: 40,
    suggestion: 40,
    resolution: 13,
    relations: 13,
    comments: 60
  };

  req.rql.fields = {
    changes: 1
  };

  Object.keys(columns).forEach(field => req.rql.fields[field] = 1);

  express.crud.exportRoute(app, {
    filename: 'WMES_OSH_NEAR_MISSES',
    freezeRows: 1,
    freezeColumns: 1,
    columns,
    model: OshNearMiss,
    serializeRow
  }, req, res, next);

  function serializeRow(doc)
  {
    const comments = [];

    doc.changes.forEach(change =>
    {
      if (!change.comment)
      {
        return;
      }

      const time = moment(change.date).format('YYYY-MM-DD HH:mm:ss');
      const user = change.user.label;
      const text = change.comment.replace(/\r\n/g, '\n').replace('\n', '\r\n');

      comments.push(`[${time}] ${user}: ${text}`);
    });

    return {
      rid: doc.rid,
      status: doc.status,
      createdAt: doc.createdAt,
      startedAt: doc.startedAt,
      plannedAt: doc.plannedAt,
      finishedAt: doc.finishedAt,
      duration: Math.round(doc.duration / 3600 * 1000) / 1000,
      creator: doc.creator.label,
      acceptor: doc.acceptor ? doc.acceptor.label : '',
      finisher: doc.finisher ? doc.finisher.label : '',
      implementer: doc.implementer ? doc.implementer.label : '',
      coordinators: doc.coordinators.map(u => u.label).join('; '),
      division: names.single('divisions', doc.division),
      workplace: names.single('workplaces', doc.division),
      department: names.single('departments', doc.department),
      building: names.single('buildings', doc.building),
      location: names.single('locations', doc.location),
      station: names.single('stations', doc.station),
      kind: names.multi('kinds', doc.kind).join('; '),
      eventDate: doc.eventDate,
      eventCategory: names.single('eventCategories', doc.eventCategory),
      reasonCategory: names.single('reasonCategories', doc.reasonCategory),
      priority: doc.priority,
      materialLoss: doc.materialLoss,
      subject: doc.subject,
      problem: doc.problem,
      reason: doc.reason,
      suggestion: doc.suggestion,
      resolution: doc.resolution.rid,
      relations: doc.relations.map(r => r.rid).join('; '),
      comments: comments.join('\r\n')
    };
  }
};
