// paintShop/routes/exportOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');

module.exports = (app, module, req, res, next) =>
{
  const {
    express,
    PaintShopOrder,
    DrillingOrder
  } = module;

  const drillingStatuses = new Map();

  const options = {
    filename: 'WMES-PAINT_SHOP-ORDERS',
    freezeRows: 1,
    freezeColumns: 1,
    columns: {
      no: 7,
      type: 11,
      orderNo: 10,
      status: 9,
      nc12: 13,
      name: 45,
      quantity: {type: 'integer', width: 8},
      unit: 4,
      mrp: 4,
      manHours: 'decimal',
      assemblyTime: {type: 'datetime+utc', width: 18},
      startedAt: {type: 'datetime+utc', width: 18},
      finishedAt: {type: 'datetime+utc', width: 18},
      duration: 'decimal'
    },
    serializeRow: (leadingOrder) =>
    {
      const rows = [{
        no: `${leadingOrder.no}.`,
        type: 'leading',
        orderNo: leadingOrder.order,
        status: leadingOrder.status,
        nc12: leadingOrder.nc12,
        name: leadingOrder.name,
        quantity: leadingOrder.qty,
        unit: 'PCE',
        mrp: leadingOrder.mrp,
        manHours: 0,
        assemblyTime: new Date(leadingOrder.startTime),
        startedAt: leadingOrder.startedAt,
        finishedAt: leadingOrder.finishedAt,
        duration: leadingOrder.finishedAt
          ? (leadingOrder.finishedAt - leadingOrder.startedAt) / 3600 / 1000
          : 0,
        drilling: drillingStatuses.get(leadingOrder.order) || ''
      }];

      leadingOrder.childOrders.forEach((childOrder, childOrderI) =>
      {
        rows[0].manHours += childOrder.manHours;

        rows.push({
          no: `${leadingOrder.no}.${childOrderI + 1}.`,
          type: 'child',
          orderNo: childOrder.order,
          status: leadingOrder.status,
          nc12: childOrder.nc12,
          name: childOrder.name,
          quantity: childOrder.qty,
          unit: 'PCE',
          mrp: leadingOrder.mrp,
          manHours: childOrder.manHours,
          assemblyTime: new Date(leadingOrder.startTime),
          startedAt: leadingOrder.startedAt,
          finishedAt: leadingOrder.finishedAt,
          duration: rows[0].duration,
          drilling: drillingStatuses.get(childOrder.order) || ''
        });

        childOrder.components.forEach((component, componentI) =>
        {
          rows.push({
            no: `${leadingOrder.no}.${childOrderI + 1}.${componentI + 1}.`,
            type: 'component',
            orderNo: childOrder.order,
            status: leadingOrder.status,
            nc12: component.nc12,
            name: component.name,
            quantity: Math.ceil(component.qty),
            unit: component.unit,
            mrp: leadingOrder.mrp,
            manHours: childOrder.manHours,
            assemblyTime: new Date(leadingOrder.startTime),
            startedAt: leadingOrder.startedAt,
            finishedAt: leadingOrder.finishedAt,
            duration: rows[0].duration,
            drilling: ''
          });
        });
      });

      return rows;
    },
    model: PaintShopOrder
  };

  step(
    function()
    {
      DrillingOrder
        .find({date: new Date(`${req.query.date}T00:00:00Z`)})
        .select({
          _id: 0,
          status: 1,
          order: 1,
          'childOrders.order': 1
        })
        .lean()
        .exec(this.next());
    },
    function(err, drillingOrders)
    {
      if (err)
      {
        return next(err);
      }

      drillingOrders.forEach(drillingOrder =>
      {
        drillingStatuses.set(drillingOrder.order, drillingOrder.status);

        drillingOrder.childOrders.forEach(childOrder =>
        {
          drillingStatuses.set(childOrder.order, drillingOrder.status);
        });
      });

      express.crud.exportRoute(app, options, req, res, next);
    }
  );
};
