// paintShop/routes/exportLoad.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const COUNTERS = {
  1: 'in',
  2: '1',
  3: '2'
};

module.exports = async (app, module, req, res, next) =>
{
  const {
    express,
    PaintShopLoad,
    PaintShopLoadReason
  } = module;

  const reasonList = await PaintShopLoadReason
    .find({})
    .select({description: 1})
    .lean()
    .exec();
  const reasonMap = new Map();

  reasonMap.set(null, '');

  reasonList.forEach(reason =>
  {
    reasonMap.set(reason._id.toString(), reason.description);
  });

  const options = {
    filename: 'WMES-PAINT_SHOP-LOAD',
    freezeRows: 1,
    freezeColumns: 1,
    columns: {
      time: {type: 'datetime', width: 18},
      counter: 5,
      duration: {type: 'decimal'},
      reason: 40
    },
    serializeRow: doc =>
    {
      return {
        time: doc._id.ts,
        counter: COUNTERS[doc._id.c] || `?${doc._id.c}?`,
        duration: doc.d / 1000,
        reason: doc.r ? reasonMap.get(doc.r.toString()) : ''
      };
    },
    batchSize: 100,
    model: PaintShopLoad
  };

  express.crud.exportRoute(app, options, req, res, next);
};
