// warehouse/importer/transferOrders.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const path = require('path');
const _ = require('lodash');
const moment = require('moment');
const step = require('h5.step');
const fs = require('fs-extra');
const parseTransferOrders = require('./parseTransferOrders');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  filterRe: /^T_LT23\.txt$/,
  parsedOutputDir: null
};

exports.models = [
  require('../models/whTransferOrder'),
  require('../models/whControlCycle')
];

exports.recordTopics = {
  info: ['warehouse.transferOrders.synced'],
  error: ['warehouse.transferOrders.syncFailed']
};

exports.start = function startTransferOrdersImporterModule(app, module)
{
  const mongoose = app[module.config.mongooseId];

  if (!mongoose)
  {
    throw new Error('mongoose module is required!');
  }

  const WhTransferOrder = mongoose.model('WhTransferOrder');
  const WhControlCycle = mongoose.model('WhControlCycle');

  const filePathCache = {};

  app.broker.subscribe('directoryWatcher.changed', queueFile).setFilter(filterFile);

  app.broker.subscribe('warehouse.importQueue.transferOrders', importNext);

  function filterFile(fileInfo)
  {
    if (filePathCache[fileInfo.filePath] || !module.config.filterRe.test(fileInfo.fileName))
    {
      return false;
    }

    fileInfo.timeKey = createTimeKey(fileInfo.timestamp);

    return true;
  }

  function createTimeKey(timestamp)
  {
    return moment(timestamp).subtract(1, 'days').format('YYYYMMDD');
  }

  function queueFile(fileInfo)
  {
    filePathCache[fileInfo.filePath] = true;

    app.broker.publish('warehouse.importQueue.push', {
      timestamp: fileInfo.timestamp,
      type: 'transferOrders',
      data: fileInfo
    });

    module.info('Queued...', {fileInfo});
  }

  function importNext(fileInfo)
  {
    const startTime = Date.now();

    module.info('Importing...', {fileInfo});

    importFile(fileInfo, function(err, transferOrders)
    {
      cleanUpFileInfoFile(fileInfo);

      if (err)
      {
        module.error(err, 'Failed to import.', {fileInfo});

        app.broker.publish('warehouse.transferOrders.syncFailed', {
          timestamp: fileInfo.timestamp,
          error: err.message
        });
      }
      else
      {
        module.info('Imported!', {fileInfo, count: transferOrders.length, duration: Date.now() - startTime});

        app.broker.publish('warehouse.transferOrders.synced', {
          timestamp: fileInfo.timestamp,
          count: transferOrders.length
        });
      }
    });
  }

  function importFile(fileInfo, done)
  {
    step(
      function fetchControlCyclesStep()
      {
        const next = _.once(this.next());
        const nc12ToS = this.nc12ToS = {};
        const stream = WhControlCycle.find({}, {s: 1}).lean().cursor();

        stream.on('error', next);
        stream.on('end', next);
        stream.on('data', function(doc)
        {
          nc12ToS[doc._id] = doc.s;
        });
      },
      function readFileStep(err)
      {
        if (err)
        {
          return this.skip(err);
        }

        fs.readFile(fileInfo.filePath, {encoding: 'utf8'}, this.next());
      },
      function parseFileStep(err, fileContents)
      {
        if (err)
        {
          return this.skip(err);
        }

        module.info('Parsing...', {fileInfo, fileSize: fileContents.length});

        const t = Date.now();

        this.transferOrders = parseTransferOrders(fileContents, new Date(fileInfo.timestamp), this.nc12ToS);
        this.nc12ToS = null;

        module.info('Parsed!', {fileInfo, duration: Date.now() - t});

        setImmediate(this.next());
      },
      function createBatchesStep()
      {
        const batchSize = 1000;
        const batchCount = Math.ceil(this.transferOrders.length / batchSize);

        this.batches = [];

        for (let i = 0; i < batchCount; ++i)
        {
          this.batches.push(this.transferOrders.slice(i * batchSize, i * batchSize + batchSize));
        }

        setImmediate(this.next());
      },
      function insertTransferOrdersStep()
      {
        for (let i = 0, l = this.batches.length; i < l; ++i)
        {
          WhTransferOrder.collection.insertMany(this.batches[i], {ordered: false}, this.parallel());
        }
      },
      function handleArchiveError(err)
      {
        if (!err || err.code === 11000)
        {
          return done(null, this.transferOrders);
        }

        if (err.err && !err.message)
        {
          const code = err.code;

          err = new Error(err.err);
          err.name = 'MongoError';
          err.code = code;
        }

        done(err, this.transferOrders);
      }
    );
  }

  function cleanUpFileInfoFile(fileInfo)
  {
    setTimeout(removeFilePathFromCache, 15000, fileInfo.filePath);

    if (module.config.parsedOutputDir)
    {
      moveFileInfoFile(fileInfo.filePath);
    }
    else
    {
      deleteFileInfoFile(fileInfo.filePath);
    }
  }

  function moveFileInfoFile(oldFilePath)
  {
    const newFilePath = path.join(module.config.parsedOutputDir, path.basename(oldFilePath));

    fs.move(oldFilePath, newFilePath, {overwrite: true}, function(err)
    {
      if (err)
      {
        module.error(err, 'Failed to rename file.', {oldFilePath, newFilePath});
      }
    });
  }

  function deleteFileInfoFile(filePath)
  {
    fs.unlink(filePath, function(err)
    {
      if (err)
      {
        module.error(err, 'Failed to delete file.', {filePath});
      }
    });
  }

  function removeFilePathFromCache(filePath)
  {
    delete filePathCache[filePath];
  }
};
