// wmes-wiring/models/wiringEvent.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'WiringEvent';

exports.setUp = (app, mongoose) =>
{
  const wiringEventSchema = new mongoose.Schema({
    type: {
      type: String,
      required: true
    },
    time: {
      type: Date,
      required: true
    },
    user: {
      type: {},
      required: true
    },
    order: {
      type: mongoose.Schema.Types.ObjectId,
      default: null,
      ref: 'WiringOrder'
    },
    data: {}
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  wiringEventSchema.statics.TOPIC_PREFIX = 'wiring.events';
  wiringEventSchema.statics.BROWSE_LIMIT = 100;

  wiringEventSchema.index({time: -1});
  wiringEventSchema.index({type: 1, time: -1});
  wiringEventSchema.index({order: 1, time: -1});

  wiringEventSchema.post('save', function(doc)
  {
    app.broker.publish(`${this.TOPIC_PREFIX}.saved`, doc.toJSON());
  });

  wiringEventSchema.statics.record = function(event)
  {
    (new this(event)).save(err =>
    {
      if (err)
      {
        app.error(err, 'Failed to record event.', {module: 'wmes-wiring', event});
      }
    });
  };

  mongoose.model(exports.name, wiringEventSchema);
};
