// watchdog/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const setUpEmptyDirectoryCheck = require('./emptyDirectoryCheck');
const setUpPing = require('./ping');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  mailSenderId: 'mail/sender',
  smsSenderId: 'sms/sender',
  twilioId: 'twilio',
  expressId: 'express',
  recipients: [],
  appStartedRecipients: [],
  appStartedCallRecipient: null,
  noEventRecipients: [],
  events: [],
  emptyDirectories: [],
  ping: {
    secretKey: '?',
    remoteUrl: 'http://127.0.0.1/watchdog/ping',
    localUrl: 'http://127.0.0.1/ping',
    interval: 0,
    window: 0
  },
  restartBroadcastPort: 1337,
  dnsCheck: {
    dnsServer: '0.0.0.0',
    domainToCheck: 'wmes.pl',
    addressToExpect: '0.0.0.0',
    notify: []
  }
};

exports.models = [
  require('events/models/event')
];

exports.optionalModules = {
  'mongoose': require('./dbCheck'),
  'mongoose mailSender': [
    require('./noEventCheck'),
    require('./appStartedEventCheck')
  ],
  'mailSender': require('./dnsCheck')
};

exports.start = (app, module) =>
{
  setUpEmptyDirectoryCheck(app, module);
  setUpPing(app, module);
};
