// watchdog/ping.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const step = require('h5.step');
const request = require('request');

module.exports = (app, module) =>
{
  const logger = module.logger.create({submodule: 'ping'});
  const config = module.config.ping;

  let lastNotifyAt = 0;
  let notifyTimer = null;

  if (config.window > 0)
  {
    app.onModuleReady(module.config.expressId, setUpReceiver);

    app.broker.subscribe('app.started', () => scheduleNotify(1)).setLimit(1);
  }

  if (config.interval > 0)
  {
    app.broker.subscribe('app.started', ping).setLimit(1);
  }

  function ping()
  {
    step(
      function sendPingRequestStep()
      {
        request({url: config.localUrl, timeout: 5000}, this.next());
      },
      function handlePingResponseStep(err, res, body)
      {
        if (err)
        {
          logger.info(err, `Failed local request.`);

          return this.skip();
        }

        if (res.statusCode === 200 && body === 'pong')
        {
          request({
            url: config.remoteUrl,
            timeout: 5000,
            method: 'POST',
            json: {secretKey: config.secretKey}
          }, this.next());
        }
        else
        {
          logger.info(`Invalid local response.`, {
            statusCode: res.statusCode,
            body: typeof body === 'string' ? body.substring(0, 100) : '-'
          });
        }
      },
      function(err, res)
      {
        if (err)
        {
          logger.warn(err, `Failed remote request`);
        }
        else if (res && res.statusCode >= 300)
        {
          logger.warn(`Invalid remote response status code`, {statusCode: res.statusCode});
        }

        setTimeout(ping, config.interval);
      }
    );
  }

  function setUpReceiver()
  {
    app[module.config.expressId].post('/watchdog/ping', (req, res, next) =>
    {
      if (req.body.secretKey !== config.secretKey)
      {
        return next(app.createError('INVALID_SECRET_KEY', 403));
      }

      scheduleNotify(1);

      res.end();
    });
  }

  function scheduleNotify(delayMultiplier)
  {
    clearTimeout(notifyTimer);
    notifyTimer = setTimeout(notify, config.window * delayMultiplier);
  }

  function notify()
  {
    logger.info('[ping] Service unavailable: notifying!');

    const subject = `[${app.options.id}:${module.name}:ping] Service unavailable`;
    const window = Math.round(config.window / 1000);
    const text = [
      `No ping request received in ${window}s: the Internet or the local HTTP server is down!`,
      '',
      'This message was generated automatically.',
      'Sincerely, WMES Bot'
    ];

    mail(subject, text);

    const now = Date.now();

    if (now - lastNotifyAt > 60 * 60 * 1000)
    {
      lastNotifyAt = now;

      call(`Service unavailable - no ping in ${window} seconds!`);
    }

    scheduleNotify(30);
  }

  function mail(subject, text)
  {
    const mailSender = app[module.config.mailSenderId];

    if (!mailSender)
    {
      return;
    }

    const to = _.uniq([].concat(
      module.config.recipients,
      module.config.appStartedRecipients
    ));

    if (!to.length)
    {
      return;
    }

    mailSender.send(to, subject, text.join('\r\n'), (err) =>
    {
      if (err)
      {
        logger.error(err, 'Failed to e-mail.', {recipients: to});
      }
      else
      {
        logger.info(`E-mailed!`, {recipients: to});
      }
    });
  }

  function call(text)
  {
    const twilio = app[module.config.twilioId];

    if (!twilio || !module.config.appStartedCallRecipient)
    {
      return;
    }

    const sayOptions = {
      to: module.config.appStartedCallRecipient,
      message: text,
      voice: 'alice',
      language: 'en-US'
    };

    twilio.say(sayOptions, (err) =>
    {
      if (err)
      {
        logger.error(err, 'Failed to call.', {recipients: [sayOptions.to]});
      }
      else
      {
        logger.info(`Called!`, {recipients: [sayOptions.to]});
      }
    });
  }
};
