// wmes-drilling/models/drillingEvent.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'DrillingEvent';

exports.setUp = (app, mongoose) =>
{
  const drillingEventSchema = new mongoose.Schema({
    type: {
      type: String,
      required: true
    },
    time: {
      type: Date,
      required: true
    },
    user: {
      type: {},
      required: true
    },
    order: {
      type: String,
      default: null,
      ref: 'DrillingOrder'
    },
    data: {}
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  drillingEventSchema.statics.TOPIC_PREFIX = 'drilling.events';
  drillingEventSchema.statics.BROWSE_LIMIT = 100;

  drillingEventSchema.index({time: -1});
  drillingEventSchema.index({type: 1, time: -1});
  drillingEventSchema.index({order: 1, time: -1});

  drillingEventSchema.post('save', function(doc)
  {
    app.broker.publish(`${this.TOPIC_PREFIX}.saved`, doc.toJSON());
  });

  drillingEventSchema.statics.record = function(event)
  {
    (new this(event)).save(err =>
    {
      if (err)
      {
        app.error(err, 'Failed to record event.', {module: 'wmes-drilling', event});
      }
    });
  };

  mongoose.model(exports.name, drillingEventSchema);
};
