// kanban/models/kanbanEntry.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'KanbanEntry';

exports.setUp = (app, mongoose) =>
{
  const kanbanEntrySchema = new mongoose.Schema({
    _id: String,
    createdAt: Date,
    updatedAt: Date,
    updater: {},
    kind: String,
    container: String,
    nc12: String,
    supplyArea: String,
    workCenter: String,
    componentQty: Number,
    kanbanId: [Number],
    storageType: Number,
    storingPosition: String,
    storageBin: String,
    workstations: [Number],
    locations: [String],
    discontinued: Boolean,
    deleted: Boolean,
    comment: String,
    split: Number,
    updates: {},
    changes: {}
  }, {
    id: false,
    minimize: false
  });

  kanbanEntrySchema.index({updatedAt: -1});
  kanbanEntrySchema.index({nc12: 1});
  kanbanEntrySchema.index({supplyArea: 1});
  kanbanEntrySchema.index({discontinued: 1});
  kanbanEntrySchema.index({kanbanId: 1});
  kanbanEntrySchema.index({split: 1});

  kanbanEntrySchema.statics.TOPIC_PREFIX = 'kanban.entries';
  kanbanEntrySchema.statics.BROWSE_LIMIT = 0;

  kanbanEntrySchema.statics.createFromImport = function(data, updatedAt, updater)
  {
    return {
      _id: data[0],
      createdAt: updatedAt,
      updatedAt,
      updater,
      kind: null,
      container: null,
      nc12: data[1],
      supplyArea: data[2],
      workCenter: '',
      componentQty: data[4],
      kanbanId: data[5],
      storageType: data[6],
      storingPosition: data[7],
      storageBin: data[8],
      workstations: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      locations: ['', '', '', '', '', '', '', '', '', ''],
      discontinued: false,
      deleted: false,
      comment: '',
      split: 1,
      updates: {},
      changes: []
    };
  };

  mongoose.model(exports.name, kanbanEntrySchema);
};
