// wmes-ct-frontend/routes/exportBalancingPces.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module, req, res, next) =>
{
  const {express, CtBalancingPce} = module;

  const options = {
    filename: 'WMES-CT-BALANCING',
    freezeRows: 1,
    columns: {
      line: 10,
      station: {type: 'integer', width: 4},
      orderNo: 10,
      nc12: 13,
      name: 45,
      mrp: 4,
      qty: {type: 'integer', width: 5},
      workerCount: {type: 'integer', width: 5},
      sapTaktTime: {type: 'integer', width: 10},
      startedAt: {type: 'datetime', width: 18},
      finishedAt: {type: 'datetime', width: 18},
      duration: 'decimal'
    },
    serializeRow: (pce) =>
    {
      return {
        line: pce.line,
        station: pce.station,
        orderNo: pce.order._id,
        nc12: pce.order.nc12,
        name: pce.order.name,
        mrp: pce.order.mrp,
        qty: pce.order.qty,
        workerCount: pce.order.workerCount,
        sapTaktTime: pce.order.sapTaktTime,
        startedAt: pce.startedAt,
        finishedAt: pce.finishedAt,
        duration: pce.duration / 1000
      };
    },
    model: CtBalancingPce
  };

  express.crud.exportRoute(app, options, req, res, next);
};
