// wmes-ct-frontend/routes/exportPces.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module, req, res, next) =>
{
  const {express, CtPce} = module;

  const options = {
    filename: 'WMES-CT-PCES',
    freezeRows: 1,
    columns: {
      line: 10,
      station: {type: 'integer', width: 4},
      orderNo: 10,
      nc12: 13,
      name: 45,
      pce: {type: 'integer', width: 5},
      qty: {type: 'integer', width: 5},
      workerCount: {type: 'integer', width: 5},
      sapTaktTime: {type: 'integer', width: 10},
      startedAt: {type: 'datetime', width: 18},
      finishedAt: {type: 'datetime', width: 18},
      totalDuration: 'decimal',
      workDuration: 'decimal',
      downtimeDuration: 'decimal',
      scheduledDuration: 'decimal'
    },
    serializeRow: (pce) =>
    {
      return {
        line: pce.line,
        station: pce.station,
        orderNo: pce.order._id,
        nc12: pce.order.nc12,
        name: pce.order.name,
        pce: pce.pce,
        qty: pce.order.qty,
        workerCount: pce.order.workerCount,
        sapTaktTime: pce.order.sapTaktTime,
        startedAt: pce.startedAt,
        finishedAt: pce.finishedAt,
        totalDuration: pce.durations.total / 1000,
        workDuration: pce.durations.work / 1000,
        downtimeDuration: pce.durations.downtime / 1000,
        scheduledDuration: pce.durations.scheduled / 1000
      };
    },
    model: CtPce
  };

  express.crud.exportRoute(app, options, req, res, next);
};
