// wmes-ct-frontend/routes/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const pceReportRoute = require('./pceReport');
const groupsReportRoute = require('./groupsReport');
const resultsReportRoute = require('./resultsReport');
const balancingReportRoute = require('./balancingReport');
const inveoRoute = require('./inveo');
const exportPcesRoute = require('./exportPces');
const addBalancingPceRoute = require('./addBalancingPce');
const editBalancingPceRoute = require('./editBalancingPce');
const deleteBalancingPceRoute = require('./deleteBalancingPce');
const exportBalancingPcesRoute = require('./exportBalancingPces');

module.exports = (app, module) =>
{
  const {
    express,
    user,
    settings,
    CtCart,
    CtLine,
    CtTodo,
    CtPce,
    CtOrderGroup,
    CtMrpConfig,
    CtUpphQuarterlyConfig,
    CtBalancingPce,
    messengerClient
  } = module;

  const canView = user.auth('PROD_DATA:VIEW');
  const canManage = user.auth('PROD_DATA:MANAGE');
  const canManagePEng = user.auth('PROD_DATA:MANAGE', 'FN:*process-engineer*');
  const canTodo = user.auth('LOCAL', 'PROD_DATA:MANAGE', 'FN:*process-engineer*');

  const dictionaries = {
    carts: {model: CtCart},
    lines: {
      model: CtLine,
      prepareResult: fetchCarts
    },
    orderGroups: {
      model: CtOrderGroup,
      canManage: user.auth('CT:MANAGE:ORDER_GROUPS')
    },
    mrpConfigs: {
      model: CtMrpConfig,
      canManage: canManagePEng
    },
    upphQuarterlyConfigs: {
      model: CtUpphQuarterlyConfig,
      upsert: true,
      canManage: canManagePEng
    }
  };

  express.get('/ct/settings', canView, settings.browseRoute('ct'));
  express.put('/ct/settings/:id', canManage, settings.updateRoute);

  express.get('/ct/state', canView, getStateRoute);

  express.get('/ct/todos', canView, express.crud.browseRoute.bind(null, app, CtTodo));
  express.get('/ct/todos/:line/:station', canTodo, inveoRoute.bind(null, app, module));
  express.post('/ct/todos/:line/:station', canTodo, inveoRoute.bind(null, app, module));

  express.get('/ct/pces', canView, express.crud.browseRoute.bind(null, app, CtPce));
  express.get(
    '/ct/pces;export.:format?',
    canView,
    exportPcesRoute.bind(null, app, module)
  );

  express.get('/ct/balancing/pces', canView, express.crud.browseRoute.bind(null, app, CtBalancingPce));
  express.get(
    '/ct/balancing/pces;export.:format?',
    canView,
    exportBalancingPcesRoute.bind(null, app, module)
  );
  express.post('/ct/balancing/pces', canTodo, addBalancingPceRoute.bind(null, app, module));
  express.put('/ct/balancing/pces/:id', canTodo, editBalancingPceRoute.bind(null, app, module));
  express.delete('/ct/balancing/pces/:id', canManagePEng, deleteBalancingPceRoute.bind(null, app, module));

  express.get('/ct/reports/pce', canView, pceReportRoute.bind(null, app, module));
  express.get('/ct/reports/groups', canView, groupsReportRoute.bind(null, app, module));
  express.get('/ct/reports/results', canView, resultsReportRoute.bind(null, app, module));
  express.get('/ct/reports/balancing', canView, balancingReportRoute.bind(null, app, module));

  express.get('/inveo/:line/:station', canTodo, inveoRoute.bind(null, app, module));

  Object.keys(dictionaries).forEach(k =>
  {
    const options = dictionaries[k];
    const canViewDictionary = options.canView || canView;
    const canManageDictionary = options.canManage || canManage;

    express.get(`/ct/${k}`, canViewDictionary, express.crud.browseRoute.bind(null, app, options));
    express.post(`/ct/${k}`, canManageDictionary, express.crud.addRoute.bind(null, app, options));
    express.get(`/ct/${k}/:id`, canViewDictionary, express.crud.readRoute.bind(null, app, options));
    express.put(`/ct/${k}/:id`, canManageDictionary, express.crud.editRoute.bind(null, app, options));
    express.delete(`/ct/${k}/:id`, canManageDictionary, express.crud.deleteRoute.bind(null, app, options));
  });

  function getStateRoute(req, res, next)
  {
    messengerClient.request('ct.getAllLineState', {}, (err, lineStates) =>
    {
      if (err)
      {
        return next(err);
      }

      res.json(lineStates);
    });
  }

  function fetchCarts(ctLine, formatResult, req)
  {
    if (!req.query.carts)
    {
      return formatResult(null, ctLine);
    }

    const conditions = {};

    if (req.query.carts !== '1')
    {
      conditions.type = req.query.carts;
    }

    CtCart.find(conditions).select({_id: 0, cards: 1}).lean().exec((err, ctCarts) =>
    {
      if (err)
      {
        return formatResult(err);
      }

      ctLine.carts = ctCarts.map(c => c.cards);

      formatResult(null, ctLine);
    });
  }
};
