// wmes-ct-frontend/routes/inveo.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const duplicates = new Map();

module.exports = (app, module, req, res, next) =>
{
  const {
    CtCart,
    CtTodo
  } = module;

  const params = {
    ...req.body,
    ...req.query,
    ...req.params
  };

  if (!module.lines.has(params.line))
  {
    return next(app.createError('Invalid line.', 'INPUT', 400));
  }

  const todo = new CtTodo({
    time: new Date(parseInt(params.time, 10) || Date.now()),
    line: params.line,
    station: parseInt(params.station, 10),
    action: params.id ? 'cart' : params.in ? 'input' : null
  });

  if (isNaN(todo.time.getTime()))
  {
    return next(app.createError('Invalid time.', 'INPUT', 400));
  }

  if (!todo.action)
  {
    return next(app.createError('Invalid action type.', 'INPUT', 400));
  }

  if (isNaN(todo.station) || todo.station < 1 || todo.station > 7)
  {
    return next(app.createError('Invalid station.', 'INPUT', 400));
  }

  const device = params.mac || '000000000000';

  if (!/^[A-Za-z0-9]{1,32}$/.test(device))
  {
    return next(app.createError('Invalid device ID.', 'INPUT', 400));
  }

  if (todo.action === 'input')
  {
    todo.data.state = params.in === '1';

    return save(todo);
  }

  if (todo.action === 'cart')
  {
    todo.data.cart = params.id;

    if (!todo.data.cart)
    {
      return next(app.createError('Invalid card ID.', 'INPUT', 400));
    }

    const cachedCart = module.cartCache.get(todo.data.cart);

    if (cachedCart)
    {
      todo.data.cart = cachedCart;

      checkDup(device, todo);
    }
    else
    {
      CtCart.findOne({cards: todo.data.cart}).lean().exec((err, cart) =>
      {
        if (err)
        {
          return next(err);
        }

        if (cart)
        {
          module.cartCache.set(todo.data.cart, cart.cards[0]);

          todo.data.cart = cart.cards[0];
        }
        else
        {
          module.cartCache.set(todo.data.cart, todo.data.cart);
        }

        checkDup(device, todo);
      });
    }
  }

  function save(todo)
  {
    todo.save(err =>
    {
      if (err)
      {
        return next(err);
      }

      res.sendStatus(204);
    });
  }

  function checkDup(device, todo)
  {
    const dupKey = `${device}:${todo.station}`;

    if (duplicates.get(dupKey) === todo.data.cart)
    {
      app.broker.publish('ct.todos.ignored', todo);

      return res.sendStatus(204);
    }

    duplicates.set(dupKey, todo.data.cart);

    save(todo);
  }
};
