// wmes-ct-frontend/routes/resultsReport.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const moment = require('moment');
const report = require('../reports/results');

module.exports = (app, module, req, res, next) =>
{
  const reports = app[module.config.reportsId];

  if (!reports)
  {
    return next(app.createError('Reports disabled.', 'UNAVAILABLE', 500));
  }

  const query = req.query;
  const options = {
    fromTime: reports.helpers.getTime(query.from) || null,
    toTime: reports.helpers.getTime(query.to) || null,
    ignoredMrps: !query.ignoredMrps ? null : query.ignoredMrps.split(',').filter(v => !!v.length),
    includedMrps: !query.includedMrps ? null : query.includedMrps.split(',').filter(v => !!v.length),
    minLineWorkDuration: query.minLineWorkDuration >= 0 ? parseFloat(query.minLineWorkDuration) : null,
    minUpphWorkDuration: query.minUpphWorkDuration >= 0 ? parseFloat(query.minUpphWorkDuration) : null,
    shiftCount: query.shiftCount ? parseInt(query.shiftCount, 10) : null,
    availableWorkDuration: query.availableWorkDuration > 0 ? parseFloat(query.availableWorkDuration) : null,
    minMrpUnbalance: query.minMrpUnbalance >= 0 ? parseInt(query.minMrpUnbalance, 10) : null,
    minMrpEfficiency: query.minMrpEfficiency >= 0 ? parseInt(query.minMrpEfficiency, 10) : null,
    lines: []
  };

  if (!options.fromTime)
  {
    options.fromTime = moment(options.toTime).subtract(12, 'months').valueOf();
  }

  module.orgUnits.getAllByType('subdivision').forEach(subdivision =>
  {
    if (subdivision.type !== 'assembly')
    {
      return;
    }

    module.orgUnits.getProdLinesFor(subdivision).forEach(prodLine =>
    {
      if (prodLine.deactivatedAt && prodLine.deactivatedAt < options.fromTime)
      {
        return;
      }

      if (prodLine._id.startsWith('TEST'))
      {
        return;
      }

      options.lines.push(prodLine._id);
    });
  });

  reports.generateResponse(report, options, req, res, next);
};
