// wmes-ct-frontend/routes/resultsReport.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const report = require('../reports/results');

module.exports = (app, module, req, res, next) =>
{
  const reports = app[module.config.reportsId];

  if (!reports)
  {
    return next(app.createError('Reports disabled.', 'UNAVAILABLE', 500));
  }

  const query = req.query;
  const options = {
    fromTime: reports.helpers.getTime(query.from) || null,
    toTime: reports.helpers.getTime(query.to) || null,
    ignoredMrps: (query.ignoredMrps || '').split(',').filter(v => !!v.length),
    minLineWorkDuration: query.minLineWorkDuration == null ? null : parseFloat(query.minLineWorkDuration),
    minUpphWorkDuration: query.minUpphWorkDuration == null ? null : parseFloat(query.minUpphWorkDuration),
    shiftCount: query.shiftCount == null ? null : parseInt(query.shiftCount, 10),
    availableWorkDuration: query.availableWorkDuration == null ? null : parseFloat(query.availableWorkDuration),
    minMrpUnbalance: query.minMrpUnbalance == null ? null : parseInt(query.minMrpUnbalance, 10),
    minMrpEfficiency: query.minMrpEfficiency == null ? null : parseInt(query.minMrpEfficiency, 10)
  };

  reports.generateResponse(report, options, req, res, next);
};
