// d8/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  sioId: 'sio',
  userId: 'user',
  mailSenderId: 'mail/sender',
  attachmentsDest: null,
  emailUrlPrefix: 'http://127.0.0.1/',
  remind: true
};

exports.models = [
  require('./models/d8Area'),
  require('./models/d8Entry'),
  require('./models/d8EntrySource'),
  require('./models/d8ProblemSource')
];

exports.optionalModules = {
  'mongoose': [
    require('./prepareTemplateData'),
    require('./durationRecalculator'),
    require('./observerCleaner')
  ],
  'mongoose user express': require('./routes'),
  'mongoose sio user': require('./commands'),
  'mongoose mailSender': [
    require('./notifier'),
    require('./reminder')
  ]
};

exports.start = (app, module) =>
{
  if (module.config.emailUrlPrefix.substr(-1) !== '/')
  {
    module.config.emailUrlPrefix += '/';
  }

  module.DICTIONARIES = {
    areas: 'D8Area',
    entrySources: 'D8EntrySource',
    problemSources: 'D8ProblemSource'
  };
};
