// d8/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  sioId: 'sio',
  userId: 'user',
  mailSenderId: 'mail/sender',
  attachmentsDest: null,
  remind: true
};

exports.recordTopics = {
  debug: ['d8.*.added', 'd8.*.edited'],
  warning: ['d8.*.deleted'],
  blacklist: ['d8.entries.added', 'd8.entries.edited']
};

exports.republishTopics = ['d8.**'];

exports.userPrivileges = [
  'D8:VIEW', 'D8:MANAGE',
  'D8:DICTIONARIES:VIEW', 'D8:DICTIONARIES:MANAGE',
  'D8:LEADER'
];

exports.models = [
  require('./models/d8Area'),
  require('./models/d8Entry'),
  require('./models/d8EntrySource'),
  require('./models/d8ProblemSource')
];

exports.optionalModules = {
  'mongoose': [
    require('./prepareTemplateData'),
    require('./durationRecalculator'),
    require('./observerCleaner')
  ],
  'mongoose user express': require('./routes'),
  'mongoose sio user': require('./commands'),
  'mongoose mailSender': [
    require('./notifier'),
    require('./reminder')
  ]
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    areas: 'D8Area',
    entrySources: 'D8EntrySource',
    problemSources: 'D8ProblemSource'
  };
};
