// purchaseOrders/routes/renderLabelRoute.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const {exec} = require('child_process');
const path = require('path');
const fs = require('fs-extra');
const step = require('h5.step');
const renderLabelHtmlRoute = require('./renderLabelHtmlRoute');

module.exports = (app, module, req, res, next) =>
{
  const printId = req.params.printId;
  const outputFormat = req.params.format;

  if (outputFormat === 'pdf+html')
  {
    return res.render('purchaseOrders/iframe', {
      mainCssFile: app.options.mainCssFile || '/assets/main.css',
      orderId: req.params.orderId,
      printId: printId
    });
  }

  if (outputFormat === 'html')
  {
    req.query[printId.length === 32 ? 'key' : 'id'] = printId;

    return renderLabelHtmlRoute(app, module, req, res, next);
  }

  const httpServer = app[module.config.httpServerId];
  const mongoose = app[module.config.mongooseId];
  const PurchaseOrder = mongoose.model('PurchaseOrder');
  const PurchaseOrderPrint = mongoose.model('PurchaseOrderPrint');

  const pdfPath = path.join(module.config.pdfStoragePath, printId + '.pdf');

  step(
    function checkStorageStep()
    {
      fs.pathExists(pdfPath, this.next());
    },
    function findPoPrintStep(err, exists)
    {
      if (!err && exists)
      {
        return this.skip();
      }

      const conditions = {};

      conditions[printId.length === 32 ? 'key' : '_id'] = printId;

      PurchaseOrderPrint.findOne(conditions, {paper: 1}).lean().exec(this.next());
    },
    function renderLabelPdfStep(err, print)
    {
      if (err)
      {
        return this.skip(err);
      }

      if (!print)
      {
        return this.skip(app.createError('PO_PRINT_NOT_FOUND', 404));
      }

      this.pdfOptions = PurchaseOrder.PAPERS[print.paper].puppeteer;

      const url = 'http://'
          + (httpServer.config.host === '0.0.0.0' ? '127.0.0.1' : httpServer.config.host)
          + ':' + httpServer.config.port
          + '/purchaseOrders;renderLabelHtml'
          + '?' + (printId.length === 32 ? 'key' : 'id') + '=' + printId;

      const paperOptions = PurchaseOrder.PAPERS[print.paper];

      const cmd = [
        `"${module.config.wkhtmltopdfExe}"`,
        `-q --dpi 96 --zoom 1 --disable-smart-shrinking`,
        `--no-outline`,
        paperOptions.wkhtmltopdf,
        `--page-width ${paperOptions.width}mm`,
        `--page-height ${paperOptions.height}mm`,
        `"${url}"`,
        `"${pdfPath}"`
      ];

      exec(cmd.join(' '), this.next());
    },
    function sendPdfStep(err)
    {
      if (err)
      {
        return next(err);
      }

      res.sendFile(pdfPath);
    }
  );
};
