// purchaseOrders/routes/renderLabelHtmlRoute.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const {exec} = require('child_process');
const _ = require('lodash');
const step = require('h5.step');

module.exports = (app, module, req, res, next) =>
{
  const {html2pdf, PurchaseOrder, PurchaseOrderPrint} = module;

  step(
    function findPoPrintsStep()
    {
      if (req.query.key)
      {
        PurchaseOrderPrint.find({key: req.query.key}).sort({item: 1}).lean().exec(this.next());
      }
      else if (req.query.id)
      {
        PurchaseOrderPrint.find({_id: req.query.id}).lean().exec(this.next());
      }
    },
    function prepareQueryStep(err, poPrints)
    {
      if (!req.query.key && !req.query.id)
      {
        return;
      }

      if (err)
      {
        return this.skip(err);
      }

      if (!poPrints.length)
      {
        return this.skip(app.createError('NO_PO_PRINTS', 400));
      }

      req.query = {
        barcode: poPrints[0].barcode,
        paper: poPrints[0].paper,
        orderNo: [],
        nc12: [],
        quantity: [],
        itemNo: [],
        vendorNo: [],
        shippingNo: []
      };

      _.forEach(poPrints, function(poPrint)
      {
        const itemNo = String(+poPrint.item);

        for (let i = 0; i < poPrint.packageQty; ++i)
        {
          req.query.orderNo.push(poPrint.purchaseOrder);
          req.query.nc12.push(poPrint.nc12);
          req.query.quantity.push(poPrint.componentQty);
          req.query.itemNo.push(itemNo);
          req.query.vendorNo.push(poPrint.vendor);
          req.query.shippingNo.push(poPrint.shippingNo);
        }

        if (poPrint.remainingQty > 0)
        {
          req.query.orderNo.push(poPrint.purchaseOrder);
          req.query.nc12.push(poPrint.nc12);
          req.query.quantity.push(poPrint.remainingQty);
          req.query.itemNo.push(itemNo);
          req.query.vendorNo.push(poPrint.vendor);
          req.query.shippingNo.push(poPrint.shippingNo);
        }
      });
    },
    function prepareBarcodesStep()
    {
      this.barcode = PurchaseOrder.BARCODES[req.query.barcode] ? req.query.barcode : 'code128';
      this.paper = PurchaseOrder.PAPERS[req.query.paper] ? req.query.paper : 'a4';
      this.barcodeType = PurchaseOrder.BARCODES[this.barcode];
      this.barcodeOptions = PurchaseOrder.PAPERS[this.paper][this.barcode];

      const query = req.query;

      _.forEach(['orderNo', 'nc12', 'quantity', 'itemNo', 'vendorNo', 'shippingNo'], function(param)
      {
        if (!Array.isArray(query[param]))
        {
          query[param] = [query[param]];
        }
      });

      this.barcodeDataToPng = {};
      this.pages = _.map(query.orderNo, (orderNo, i) =>
      {
        const page = {
          orderNo: orderNo || '',
          nc12: query.nc12[i] || '',
          quantity: query.quantity[i] || '',
          itemNo: query.itemNo[i] || '',
          vendorNo: query.vendorNo[i] || '',
          shippingNo: query.shippingNo[i] || '',
          barcodeData: null,
          png: null
        };

        let barcodeData = `O${page.orderNo}P${page.nc12}Q${page.quantity}L${page.itemNo}S`;

        if (barcodeData.length + page.shippingNo.length > this.barcodeOptions.maxLength)
        {
          barcodeData += page.shippingNo.substr(-1 * (this.barcodeOptions.maxLength - barcodeData.length));
        }
        else
        {
          barcodeData += page.shippingNo;
        }

        barcodeData = _.trim(barcodeData, '/\\');

        page.barcodeData = barcodeData;

        this.barcodeDataToPng[barcodeData] = null;

        return page;
      });
      this.uniqueBarcodes = Object.keys(this.barcodeDataToPng);
    },
    function generateBarcodesStep()
    {
      for (let i = 0, l = this.uniqueBarcodes.length; i < l; ++i)
      {
        html2pdf.generateBarCode({
          ...this.barcodeOptions.zint,
          barcode: this.barcodeType,
          data: this.uniqueBarcodes[i],
          notext: true
        }, this.group());
      }
    },
    function renderLabelHtmlStep(err, barcodePngs)
    {
      if (err)
      {
        return this.done(next, err);
      }

      _.forEach(this.uniqueBarcodes, (uniqueBarcode, i) => { this.barcodeDataToPng[uniqueBarcode] = barcodePngs[i]; });

      _.forEach(this.pages, (page) => { page.png = this.barcodeDataToPng[page.barcodeData];});

      return res.render('purchaseOrders/' + this.paper, {
        paper: this.paper,
        barcode: this.barcode,
        pages: this.pages
      });
    }
  );
};
