// kanban/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  sapGuiId: 'sapGui',
  settingsId: 'settings',
  sapImporterMessengerId: 'messenger/client',
  html2pdfId: 'html2pdf',
  containerImagesDest: './data/kanban/containers'
};

exports.models = [
  require('./models/kanbanComponent'),
  require('./models/kanbanContainer'),
  require('./models/kanbanEntry'),
  require('./models/kanbanPrintQueue'),
  require('./models/kanbanSupplyArea'),
  require('./models/kanbanTableView'),
  require('orders/models/order')
];

exports.recordTopics = {
  info: [
    'kanban.supplyAreas.added', 'kanban.supplyAreas.edited',
    'kanban.containers.added', 'kanban.containers.edited'
  ],
  warning: [
    'kanban.supplyAreas.deleted',
    'kanban.containers.deleted'
  ]
};

exports.republishTopics = ['kanban.**'];

exports.userPrivileges = ['KANBAN:VIEW', 'KANBAN:MANAGE', 'KANBAN:PRINT'];

exports.optionalModules = {
  'mongoose': require('./state'),
  'mongoose user express settings': require('./routes')
};

exports.start = (app, module) =>
{
  module.importTimer = null;

  module.printing = null;

  app.broker.subscribe('kanban.import.*', () =>
  {
    clearTimeout(module.importTimer);
    module.importTimer = null;
  });

  app.broker.subscribe('app.started').setLimit(1).on('message', () =>
  {
    failInterruptedPrintJobs();
    removeDeletedEntries();
  });

  function failInterruptedPrintJobs()
  {
    const conditions = {'jobs.status': 'printing'};
    const update = {$set: {'jobs.$.status': 'failure'}};

    module.KanbanPrintQueue.collection.updateMany(conditions, update, err =>
    {
      if (err)
      {
        module.error(`Failed to fail interrupted jobs: ${err.message}`);
      }
    });
  }

  function removeDeletedEntries()
  {
    module.KanbanEntry.deleteMany({deleted: true}, err =>
    {
      if (err)
      {
        module.error(`Failed to remove deleted entries: ${err.message}`);
      }
    });
  }
};
