// logs/routes/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

module.exports = (app, module) =>
{
  const {
    express,
    user,
    BrowserError
  } = module;

  const canView = user.auth('SUPER');
  const canSave = user.auth('LOCAL', 'USER');

  express.get('/logs/browserErrors', canView, express.crud.browseRoute.bind(null, app, BrowserError));
  express.post('/logs/browserErrors', canSave, saveBrowserError);

  function saveBrowserError(req, res, next)
  {
    const browserError = new BrowserError({
      time: new Date(),
      user: user.createUserInfo(req.session.user, req),
      headers: req.headers,
      versions: req.body.versions,
      browser: req.body.browser,
      error: req.body.error
    });

    browserError.save((err) =>
    {
      if (err)
      {
        return next(err);
      }

      res.sendStatus(204);

      app.broker.publish('logs.browserErrors.saved', {
        errors: [browserError.toJSON()]
      });
    });
  }
};
