// wmes-oshTalks/models/oshTalk.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const autoIncrement = require('mongoose-plugin-autoinc-fix');
const userInfoSchema = require('user/models/userInfoSchema');

exports.name = 'OshTalk';

exports.setUp = (app, mongoose) =>
{
  const schema = new mongoose.Schema({
    creator: userInfoSchema,
    createdAt: Date,
    updater: userInfoSchema,
    updatedAt: Date,
    auditor: {
      type: userInfoSchema,
      required: true
    },
    section: {
      type: String,
      required: true
    },
    date: {
      type: Date,
      required: true
    },
    topics: [mongoose.Schema.Types.ObjectId],
    participants: [userInfoSchema],
    notes: {
      type: String,
      trim: true
    },
    users: [String]
  }, {
    id: false,
    minimize: false
  });

  schema.plugin(autoIncrement.plugin, {
    model: exports.name,
    field: 'rid',
    startAt: 1,
    incrementBy: 1
  });

  schema.index({date: -1});
  schema.index({section: 1});
  schema.index({'auditor.id': 1});
  schema.index({'participants.id': 1});
  schema.index({users: 1});

  schema.statics.TOPIC_PREFIX = 'oshTalks';

  schema.pre('save', function(next)
  {
    if (this.isNew)
    {
      this.createdAt = new Date();
      this.updatedAt = this.createdAt;
    }
    else
    {
      this.updatedAt = new Date();
    }

    const users = {};

    [this.creator, this.updater, this.auditor].forEach(user =>
    {
      if (user)
      {
        users[user.id] = 1;
      }
    });

    this.participants.forEach(p =>
    {
      users[p.id] = 1;
    });

    this.users = Object.keys(users);

    next();
  });

  return schema;
};
