// wmes-trw/models/trwProgram.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'TrwProgram';

exports.setUp = (app, mongoose) =>
{
  const stepSchema = new mongoose.Schema({
    _id: String,
    color: [String],
    length: Number,
    source: {
      cluster: String,
      row: Number,
      col: Number,
      endpoint: String
    },
    target: {
      cluster: String,
      row: Number,
      col: Number,
      endpoint: String
    },
    connector: String
  }, {
    _id: false,
    minimize: false
  });

  const programSchema = new mongoose.Schema({
    active: {
      type: Boolean,
      default: true
    },
    name: {
      type: String,
      required: true,
      trim: true
    },
    base: {
      type: String,
      ref: 'TrwBase',
      required: true
    },
    steps: [stepSchema]
  }, {
    id: false,
    minimize: false
  });

  programSchema.statics.TOPIC_PREFIX = 'trw.programs';
  programSchema.statics.DICTIONARY_FIELDS = {name: 1};

  mongoose.model(exports.name, programSchema);
};
