// wmes-trw/models/trwTester.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'TrwTester';

exports.setUp = (app, mongoose) =>
{
  const ioSchema = new mongoose.Schema({
    _id: {
      type: String,
      required: true,
      trim: true,
      pattern: /^[A-Z0-9-_.]{1,40}$/
    },
    name: {
      type: String,
      required: true,
      trim: true
    },
    type: {
      type: String,
      required: true,
      enum: ['input', 'output', 'analog']
    },
    device: {
      type: Number,
      required: true,
      min: 0,
      max: 127
    },
    channel: {
      type: Number,
      required: true,
      min: 0,
      max: 24
    },
    min: {
      type: Number,
      default: 0,
      min: 0,
      max: 1024
    },
    max: {
      type: Number,
      default: 0,
      min: 0,
      max: 1024
    },
    tearDown: {
      type: Boolean,
      default: true
    }
  }, {
    minimize: false
  });

  const testerSchema = new mongoose.Schema({
    name: {
      type: String,
      required: true
    },
    io: [ioSchema]
  }, {
    id: false,
    minimize: false
  });

  testerSchema.statics.TOPIC_PREFIX = 'trw.testers';
  testerSchema.statics.DICTIONARY_FIELDS = {name: 1};

  mongoose.model(exports.name, testerSchema);
};
