// workCenters/models/workCenter.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'WorkCenter';

exports.setUp = (app, mongoose) =>
{
  const workCenterSchema = new mongoose.Schema({
    _id: {
      type: String,
      required: true,
      trim: true
    },
    mrpController: {
      type: String,
      ref: 'MrpController',
      default: null
    },
    prodFlow: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'ProdFlow',
      default: null
    },
    description: {
      type: String,
      trim: true
    },
    deactivatedAt: {
      type: Date,
      default: null
    }
  }, {
    id: false
  });

  workCenterSchema.statics.TOPIC_PREFIX = 'workCenters';
  workCenterSchema.statics.BROWSE_LIMIT = 1000;

  workCenterSchema.pre('remove', function(next)
  {
    mongoose.model('ProdLine').findOne({workCenter: this._id}).select({_id: 1}).lean().exec((err, child) =>
    {
      if (err)
      {
        return next(err);
      }

      if (child)
      {
        return next(app.createError(`Org unit not empty: ${child._id}`, 'NOT_EMPTY', 400));
      }

      next();
    });
  });

  mongoose.model(exports.name, workCenterSchema);
};
