// sapGui/jobs/dpLo.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const os = require('os');
const path = require('path');
const fs = require('fs-extra');
const step = require('h5.step');
const parseSapTextTable = require('sap/util/parseSapTextTable');

module.exports = (app, sapGui, job, done) =>
{
  const logger = sapGui.logger.create({jobId: job.id});

  const outputPath = sapGui.config.outputPath;
  const inputPath = os.tmpdir();
  const usedFiles = [];
  const fileUuid = `${Date.now()}@${Math.random()}`;

  step(
    function()
    {
      logger.info('Looking for orders with dummy paint...');

      if (Array.isArray(job.orders) && job.orders.length)
      {
        this.orderFile = `${fileUuid}@WMES_DP_LO_ORDERS.txt`;

        const orderFile = path.join(inputPath, this.orderFile);

        usedFiles.push(orderFile);

        fs.writeFile(orderFile, job.orders.join('\r\n'), this.parallel());
      }

      if (Array.isArray(job.components) && job.components.length)
      {
        this.componentFile = `${fileUuid}@WMES_DP_LO_COMPONENTS.txt`;

        const componentFile = path.join(inputPath, this.componentFile);

        usedFiles.push(componentFile);

        fs.writeFile(componentFile, job.components.join('\r\n'), this.parallel());
      }
    },
    function(err)
    {
      if (err)
      {
        return this.skip(err);
      }

      this.outputFile = `${fileUuid}@WMES_DP_LO_OUTPUT.txt`;

      usedFiles.push(path.join(outputPath, this.outputFile));

      const args = [
        '--output-path', outputPath,
        '--output-file', this.outputFile,
        '--input-path', inputPath
      ];

      if (job.variant)
      {
        args.push('--variant', job.variant);

        if (job.variantCreator)
        {
          args.push('--variant-creator', job.variantCreator);
        }
      }

      if (this.orderFile)
      {
        args.push('--order-file', this.orderFile);
      }

      if (this.componentFile)
      {
        args.push('--component-file', this.componentFile);
      }

      sapGui.runScript(job, 'T_DP_LO.exe', args, this.next());
    },
    function(err, exitCode, output)
    {
      if (err)
      {
        if (err.message === 'NO_DATA')
        {
          return this.skip(null, 0, '{}');
        }

        return this.skip(err, exitCode, output);
      }

      setImmediate(this.parallel(), null, output);

      fs.readFile(path.join(outputPath, this.outputFile), 'utf8', this.parallel());
    },
    function(err, output, sapTextTable) // eslint-disable-line handle-callback-err
    {
      const data = {};
      let childOrderCount = 0;

      parseSapTextTable(sapTextTable || '', {
        columnMatchers: {
          childOrder: /^Order$/,
          nc12: /^Material$/,
          name: /^Material desc/,
          leadingOrder: /^Lead.*?Ord/,
          statuses: /^Sys.*?St/
        },
        optionalColumns: ['name', 'statuses'],
        valueParsers: {
          nc12: input => input.trim().replace(/^0+/, ''),
          statuses: input => input
            .replace(/\s+/g, ' ')
            .split(' ')
            .map(status => status.replace(/\*/g, ''))
            .filter(status => status.length > 0)
        },
        itemDecorator: obj =>
        {
          if (obj.statuses
            && (obj.statuses.includes('CNF')
            || obj.statuses.includes('DLV')
            || obj.statuses.includes('TECO')
            || obj.statuses.includes('DLT')
            || obj.statuses.includes('DLFL')))
          {
            return;
          }

          if (!obj.leadingOrder)
          {
            obj.leadingOrder = obj.childOrder;
          }

          if (!data[obj.leadingOrder])
          {
            data[obj.leadingOrder] = [];
          }

          data[obj.leadingOrder].push({
            no: obj.childOrder,
            nc12: obj.nc12 || '',
            name: obj.name || ''
          });

          childOrderCount += 1;
        }
      });

      logger.info('Orders found.', {
        leadingOrders: Object.keys(data).length,
        childOrders: childOrderCount
      });

      setImmediate(this.next(), null, 0, JSON.stringify(data));
    },
    function(err, exitCode, output)
    {
      usedFiles.forEach(filePath => fs.unlink(filePath, () => {}));

      done(err, exitCode, output);
    }
  );
};
