// sapGui/jobs/dpPing.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const os = require('os');
const axios = require('axios');

module.exports = async (app, module, job, done) =>
{
  const worker = {
    _id: os.hostname(),
    name: os.hostname(),
    apiKey: '',
    remoteServer: 'wmes.pl',
    ...job.worker
  };

  try
  {
    const res = await axios({
      method: 'POST',
      url: `https://${worker.remoteServer}/dummyPaint/workers;ping`,
      headers: {
        'X-Api-Key': worker.apiKey
      },
      data: {
        _id: worker._id.replace(/[^A-Za-z0-9_-]+/g, '-').replace(/-{2,}/g, '-'),
        name: worker.name,
        jobs: Array.from(module.workerJobIds)
      },
      timeout: 10000,
      validateStatus: null
    });

    if (res.status === 200)
    {
      module.debug('Pinged the remote server.', {
        worker: worker._id,
        remoteServer: worker.remoteServer,
        jobs: Array.from(module.workerJobIds)
      });
    }
    else
    {
      module.warn('Failed to ping the remote server.', {
        res: {
          status: res.status,
          statusText: res.statusText,
          data: res.data
        }
      });
    }
  }
  catch (err)
  {
    module.error(err, 'Failed to ping the remote server.');
  }

  done(null, 0, '');
};
