// gdrive/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const {google} = require('googleapis');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  key: {
    client_email: 'wmes@gserviceaccount.com', // eslint-disable-line camelcase
    private_key: '' // eslint-disable-line camelcase
  }
};

exports.start = function startGdriveModule(app, module)
{
  let authQueue = null;

  module.auth = new google.auth.JWT(
    module.config.key.client_email,
    null,
    module.config.key.private_key,
    ['https://www.googleapis.com/auth/drive']
  );

  module.drive = google.drive('v3');

  google.options({
    auth: module.auth
  });

  module.authorize = function(done)
  {
    if (authQueue !== null)
    {
      authQueue.push(done);
    }
    else
    {
      authQueue = [done];

      module.auth.authorize(function(err)
      {
        authQueue.forEach(callback => callback(err));
        authQueue = null;
      });
    }
  };
};
