// heff/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

exports.DEFAULT_CONFIG = {
  expressId: 'express',
  updaterId: 'updater',
  mongooseId: 'mongoose',
  fteId: 'fte',
  userId: 'user',
  orgUnitsId: 'orgUnits'
};

exports.models = [
  require('./models/heffLineState'),
  require('prodShifts/models/prodShift')
];

exports.optionalModules = {
  'express updater mongoose fte user orgUnits': require('./routes')
};

exports.start = (app, module) =>
{
  app.broker.subscribe('production.synced.**', (message) =>
  {
    if (_.includes(message.types, 'changeQuantitiesDone')
      || _.includes(message.types, 'changeShift'))
    {
      notifyReloadLine(message.prodLine);
    }
  });

  app.broker.subscribe('hourlyPlans.quantitiesPlanned', (message) =>
  {
    notifyReloadLine(message.prodLine);
  });

  app.broker.subscribe('heffLineStates.**', (message) =>
  {
    notifyReloadLine(message.model._id);
  });

  function notifyReloadLine(oldProdLine)
  {
    const newProdLine = app[module.config.orgUnitsId].fix.prodLine(oldProdLine);

    app.broker.publish(`heff.reload.${oldProdLine}`, {});

    if (newProdLine !== oldProdLine)
    {
      app.broker.publish(`heff.reload.${newProdLine}`, {});
    }
  }
};
