// h5-twilio/models/twilioRequest.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

exports.name = 'TwilioRequest';

exports.setUp = (app, mongoose) =>
{
  const twilioRequestSchema = new mongoose.Schema({
    _id: String,
    operation: String,
    options: {},
    status: String,
    createdAt: Date,
    updatedAt: Date
  }, {
    id: false
  });

  twilioRequestSchema.statics.updateStatus = function(id, newStatus, done)
  {
    this.updateOne(
      {_id: id},
      {$set: {status: newStatus, updatedAt: new Date()}},
      done || _.noop
    );
  };

  /**
   * @param {TwimlResponse} res
   * @returns {TwimlResponse}
   */
  twilioRequestSchema.methods.buildTwiml = function(res)
  {
    if (this.operation === 'say')
    {
      addSayResponse(this.options, res);
    }

    return res;
  };

  mongoose.model(exports.name, twilioRequestSchema);

  function addSayResponse(options, res)
  {
    res.say(options.message || 'Hello!', _.pick(options, ['voice', 'language', 'loop']));
  }
};
