// wmes-osh/routes/actions/add.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

module.exports = async (app, module, req, res, next) =>
{
  const {
    user,
    OshAction
  } = module;

  const input = _.pick(req.body, [
    'status',
    'subject',
    'userWorkplace',
    'userDivision',
    'kind',
    'activityKind',
    'workplace',
    'division',
    'building',
    'location',
    'station',
    'problem',
    'reason',
    'suggestion',
    'solution',
    'rootCauses',
    'participants',
    'implementers',
    'plannedAt',
    'attachments'
  ]);

  if (input.attachments && Array.isArray(input.attachments.added))
  {
    input.attachments = module.attachments.claimPendingUploads(input.attachments.added);
  }
  else
  {
    input.attachments = [];
  }

  try
  {
    const action = new OshAction({
      status: 'new',
      ...input,
      createdAt: new Date(),
      startedAt: null,
      implementedAt: null,
      finishedAt: null,
      duration: 0,
      creator: user.createUserInfo(req.session.user, req)
    });

    if (action.status === 'inProgress')
    {
      action.startedAt = action.createdAt;
    }
    else if (action.status === 'verify')
    {
      action.startedAt = action.createdAt;
      action.implementedAt = action.createdAt;
    }
    else if (action.status === 'finished')
    {
      action.startedAt = action.createdAt;
      action.implementedAt = action.createdAt;
      action.finishedAt = action.createdAt;
    }

    const [coordinators] = await Promise.all([
      OshAction.resolveCoordinators(action),
      OshAction.assignNextRid(action)
    ]);

    action.coordinators = coordinators;

    action.updateUsers([], action.creator);

    await action.save();

    app.broker.publish(`${OshAction.TOPIC_PREFIX}.added`, {
      user: action.creator,
      model: action
    });

    res.json(action);
  }
  catch (err)
  {
    next(err);
  }
};
