// prodShifts/routes.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const moment = require('moment');

module.exports = (app, module) =>
{
  const express = app[module.config.expressId];
  const settings = app[module.config.settingsId];
  const userModule = app[module.config.userId];
  const mongoose = app[module.config.mongooseId];
  const orgUnitsModule = app[module.config.orgUnitsId];
  const productionModule = app[module.config.productionId];
  const ProdShift = mongoose.model('ProdShift');

  const canView = userModule.auth('LOCAL', 'PROD_DATA:VIEW', 'PLANNING:VIEW');
  const canManage = userModule.auth(
    'PROD_DATA:MANAGE',
    'PROD_DATA:CHANGES:REQUEST',
    'PROD_DATA:CHANGES:MANAGE',
    'PROD_DATA:VIEW',
    'FN:master',
    'FN:leader'
  );

  express.get('/prodShifts', canView, express.crud.browseRoute.bind(null, app, ProdShift));

  express.post('/prodShifts', canManage, addProdShiftRoute);

  express.get(
    '/prodShifts;export.:format?',
    canView,
    function overrideFields(req, res, next)
    {
      req.rql.fields = {
        creator: 0,
        createdAt: 0
      };
      req.rql.sort = {};

      next();
    },
    express.crud.exportRoute.bind(null, app, {
      filename: 'WMES-SHIFTS',
      freezeRows: 1,
      freezeColumns: 3,
      columns: {
        date: 'date',
        shift: {type: 'integer', width: 5},
        efficiency: 'percent',
        h1: {type: 'integer', width: 5},
        h2: {type: 'integer', width: 5},
        h3: {type: 'integer', width: 5},
        h4: {type: 'integer', width: 5},
        h5: {type: 'integer', width: 5},
        h6: {type: 'integer', width: 5},
        h7: {type: 'integer', width: 5},
        h8: {type: 'integer', width: 5},
        division: 10,
        prodLine: 15
      },
      serializeRow: exportProdShift,
      model: ProdShift
    })
  );

  express.get('/prodShifts/:id', canView, findLatestProdLineShift, express.crud.readRoute.bind(null, app, ProdShift));

  express.put('/prodShifts/:id', canManage, editProdShiftRoute);

  express.delete('/prodShifts/:id', canManage, deleteProdShiftRoute);

  function exportProdShift(doc)
  {
    if (!Array.isArray(doc.orderMrp) || !doc.orderMrp.length)
    {
      return;
    }

    const subdivision = orgUnitsModule.getByTypeAndId('subdivision', doc.subdivision);
    const prodFlow = orgUnitsModule.getByTypeAndId('prodFlow', doc.prodFlow);

    return {
      date: doc.date,
      shift: doc.shift,
      prodLine: doc.prodLine,
      h1: doc.quantitiesDone[0].actual,
      h2: doc.quantitiesDone[1].actual,
      h3: doc.quantitiesDone[2].actual,
      h4: doc.quantitiesDone[3].actual,
      h5: doc.quantitiesDone[4].actual,
      h6: doc.quantitiesDone[5].actual,
      h7: doc.quantitiesDone[6].actual,
      h8: doc.quantitiesDone[7].actual,
      efficiency: doc.efficiency,
      master: doc.master ? doc.master.label : '',
      leader: doc.leader ? doc.leader.label : '',
      operator: doc.operator ? doc.operator.label : '',
      division: doc.division,
      subdivision: subdivision ? subdivision.name : doc.subdivision,
      lineMrp: Array.isArray(doc.mrpControllers) ? doc.mrpControllers.join(',') : '',
      orderMrp: Array.isArray(doc.orderMrp) ? doc.orderMrp.join(',') : '',
      prodFlow: prodFlow ? prodFlow.name : doc.prodFlow,
      workCenter: doc.workCenter,
      shiftId: doc._id
    };
  }

  function findLatestProdLineShift(req, res, next)
  {
    if (/^[0-9a-z-]{15,}$/.test(req.params.id))
    {
      return next();
    }

    step(
      function findDataStep()
      {
        settings.findById('factoryLayout.extendedDowntimeDelay', this.parallel());

        const conditions = {
          date: {
            $gt: moment().subtract(2, 'days')
          },
          prodLine: req.params.id
        };

        ProdShift.find(conditions).sort({date: -1}).limit(1).lean().exec(this.parallel());
      },
      function sendResultStep(err, extendedDowntimeDelay, prodShifts)
      {
        if (err)
        {
          return next(err);
        }

        const result = prodShifts.length ? prodShifts[0] : {
          _id: '?',
          prodLine: req.params.id
        };

        result.extendedDowntimeDelay = extendedDowntimeDelay ? extendedDowntimeDelay.value : 10;

        return res.json(result);
      }
    );
  }

  function addProdShiftRoute(req, res, next)
  {
    const user = req.session.user;
    const userInfo = userModule.createUserInfo(user, req);

    module.addProdShift(user, userInfo, req.body, (err, statusCode, logEntry) =>
    {
      if (statusCode)
      {
        res.statusCode = statusCode;
      }

      if (err)
      {
        return next(err);
      }

      if (statusCode)
      {
        return res.sendStatus(statusCode);
      }

      res.send(logEntry.data);
    });
  }

  function editProdShiftRoute(req, res, next)
  {
    const user = req.session.user;
    const userInfo = userModule.createUserInfo(user, req);

    module.editProdShift(user, userInfo, req.params.id, req.body, (err, statusCode, logEntry) =>
    {
      if (statusCode)
      {
        res.statusCode = statusCode;
      }

      if (err)
      {
        return next(err);
      }

      if (statusCode)
      {
        return res.sendStatus(statusCode);
      }

      const result = {
        _id: logEntry.prodShift,
        prodLine: logEntry.prodLine,
        ...logEntry.data
      };

      res.send(result);

      if (!productionModule.recreating)
      {
        app.broker.publish('production.edited.shift.' + logEntry.prodShift, result);
      }
    });
  }

  function deleteProdShiftRoute(req, res, next)
  {
    const user = req.session.user;
    const userInfo = userModule.createUserInfo(user, req);

    module.deleteProdShift(user, userInfo, req.params.id, req.body, function(err, statusCode, logEntry)
    {
      if (statusCode)
      {
        res.statusCode = statusCode;
      }

      if (err)
      {
        return next(err);
      }

      if (statusCode)
      {
        return res.sendStatus(statusCode);
      }

      res.send(logEntry.data);
    });
  }
};
