// wmes-wh/models/whSetCart.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'WhSetCart';

exports.setUp = (app, mongoose) =>
{
  const setCartSchema = new mongoose.Schema({
    status: {
      type: String,
      enum: ['completed', 'delivering', 'delivered'],
      required: true
    },
    kind: {
      type: String,
      enum: ['fifo', 'packaging'],
      required: true
    },
    set: {
      type: mongoose.Schema.Types.ObjectId,
      required: true
    },
    line: {
      type: String,
      required: true
    },
    cartNo: {
      type: Number,
      required: true
    },
    orderNo: {
      type: String,
      required: true
    },
    quantity: {
      type: Number,
      required: true
    }
  }, {
    id: false,
    minimize: false,
    versionKey: false
  });

  setCartSchema.statics.TOPIC_PREFIX = 'wh.setCarts';
  setCartSchema.statics.BROWSE_LIMIT = 100;

  setCartSchema.index({status: 1});
  setCartSchema.index({set: 1});
  setCartSchema.index({cartNo: 1});
  setCartSchema.index({orderNo: 1});

  mongoose.model(exports.name, setCartSchema);
};
