// wmes-wh/models/whSet.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'WhSet';

exports.setUp = (app, mongoose) =>
{
  const FUNCS = ['fmx', 'kitter', 'packer'];

  const setSchema = new mongoose.Schema({
    startedAt: Date,
    startedBy: {},
    finishedAt: Date,
    finishedBy: {},
    status: {
      type: String,
      enum: ['started', 'finished'],
      default: 'started'
    },
    mrp: String,
    line: String,
    lines: [String],
    redirLine: String,
    redirLines: [String],
    orders: [{
      _id: String,
      whOrder: mongoose.Schema.Types.ObjectId,
      sapOrder: String,
      qtyTodo: Number,
      qtyPickup: Number,
      pceTime: Number,
      funcs: [{
        _id: {
          type: String,
          enum: FUNCS
        }
      }]
    }]
  }, {
    id: false,
    minimize: false
  });

  setSchema.statics.TOPIC_PREFIX = 'wh.sets';
  setSchema.statics.BROWSE_LIMIT = 1000;

  setSchema.index({date: -1});
  setSchema.index({status: 1, date: -1});

  mongoose.model(exports.name, setSchema);
};
