// mrpControllers/models/mrpController.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'MrpController';

exports.setUp = (app, mongoose) =>
{
  const mrpControllerSchema = new mongoose.Schema({
    _id: {
      type: String,
      required: true,
      trim: true
    },
    subdivision: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Subdivision',
      required: true
    },
    description: {
      type: String,
      trim: true
    },
    deactivatedAt: {
      type: Date,
      default: null
    },
    replacedBy: {
      type: String,
      default: null
    },
    inout: {
      type: Number,
      default: 0
    }
  }, {
    id: false
  });

  mrpControllerSchema.statics.TOPIC_PREFIX = 'mrpControllers';
  mrpControllerSchema.statics.BROWSE_LIMIT = 1000;

  mrpControllerSchema.pre('remove', function(next)
  {
    mongoose.model('ProdFlow').findOne({mrpController: this._id}).select({_id: 1}).lean().exec((err, child) =>
    {
      if (err)
      {
        return next(err);
      }

      if (child)
      {
        return next(app.createError(`Org unit not empty: ${child._id}`, 'NOT_EMPTY', 400));
      }

      next();
    });
  });

  mongoose.model(exports.name, mrpControllerSchema);
};
